--- original/sound/soc/soc-dapm.c	2020-05-07 08:31:28.793760187 +0200
+++ changed/sound/soc/soc-dapm.c	2019-06-03 13:39:55.000000000 +0200
@@ -1068,6 +1068,9 @@ int snd_soc_dapm_dai_get_connected_widge
 	struct snd_soc_card *card = dai->card;
 	int paths;
 
+	if (dai->playback_widget == NULL || list == NULL)
+		return -EINVAL;
+
 	mutex_lock_nested(&card->dapm_mutex, SND_SOC_DAPM_CLASS_RUNTIME);
 	dapm_reset(card);
 
