--- original/sound/soc/codecs/msm8x16-wcd.h	2020-05-07 08:06:24.273163713 +0200
+++ changed/sound/soc/codecs/msm8x16-wcd.h	2019-06-03 13:39:55.000000000 +0200
@@ -231,6 +231,10 @@ struct msm8916_asoc_mach_data {
 	int mclk_freq;
 	int lb_mode;
 	int afe_clk_ver;
+#ifdef CONFIG_SND_EXT_MIC_BIAS
+	int mainmic_bias_gpio;
+	int submic_bias_gpio;
+#endif /* CONFIG_SND_EXT_MIC_BIAS */
 	u8 micbias1_cap_mode;
 	u8 micbias2_cap_mode;
 	atomic_t mclk_rsc_ref;
@@ -246,6 +250,25 @@ struct msm8916_asoc_mach_data {
 	void __iomem *vaddr_gpio_mux_quin_ctl;
 	void __iomem *vaddr_gpio_mux_pcm_ctl;
 	struct on_demand_supply wsa_switch_supply;
+#ifdef CONFIG_SND_SOC_MSM8X16_RT5659
+	struct snd_soc_codec *codec;
+	struct clk *rt5659_ext_clk;
+	int codec_irq_gpio;
+	int codec_ear_bias;
+#endif /* CONFIG_SND_SOC_MSM8X16_RT5659 */
+#ifdef CONFIG_SND_SOC_MSM8X16_RT5665
+	struct snd_soc_codec *codec;
+	struct clk *rt5665_ext_clk;
+	int codec_irq_gpio;
+	int codec_ear_bias;
+#endif /* CONFIG_SND_SOC_MSM8X16_RT5659 */
+#ifdef CONFIG_SAMSUNG_JACK
+	int mpp_ch_scale[3];
+#ifdef CONFIG_SEC_MPP_SHARE
+	bool mpp_standalone_mode;
+#endif /* CONFIG_SEC_MPP_SHARE */
+	int earjack_adc;
+#endif /* CONFIG_SAMSUNG_JACK */
 };
 
 struct msm8x16_wcd_pdata {
@@ -309,6 +332,9 @@ struct msm8x16_wcd_priv {
 	bool dec_active[NUM_DECIMATORS];
 	struct on_demand_supply on_demand_list[ON_DEMAND_SUPPLIES_MAX];
 	struct regulator *spkdrv_reg;
+#ifdef CONFIG_SAMSUNG_JACK
+	int micb_2_ref_cnt;
+#endif /* CONFIG_SAMSUNG_JACK */
 	/* mbhc module */
 	struct wcd_mbhc mbhc;
 	/* cal info for codec */
@@ -319,6 +345,12 @@ struct msm8x16_wcd_priv {
 	unsigned long status_mask;
 	struct wcd_imped_i_ref imped_i_ref;
 	enum wcd_mbhc_imp_det_pin imped_det_pin;
+	bool cfilt_ref_sel; /* 0x142, bit[1]:CFILT_REF_SEL */
+	bool tx2n_gnd_sel; /* 0x143, bit[2]:TX2_GND_SEL */
+	bool tx2n_int_pullup_en; /* 0x143, bit[3]:TX2_INT_PULLUP_EN */
+	bool tx2_int_rbias_en; /* 0x143, bit[4]:TX2_INT_RBIAS_EN */
+	bool tx1n_cfilt_ref_sel; /* 0x145, bit[0]:TX1N_CFILT_REF_SEL */
+	bool cdc_a_tx_1_2_atest_ctl_2;
 };
 
 extern int msm8x16_wcd_mclk_enable(struct snd_soc_codec *codec, int mclk_enable,
