--- original/sound/soc/msm/Kconfig	2020-05-07 08:31:28.676761240 +0200
+++ changed/sound/soc/msm/Kconfig	2019-06-03 13:39:55.000000000 +0200
@@ -197,21 +197,9 @@ config SND_SOC_MSM8X16
 	select SND_DYNAMIC_MINORS
 	select MSM_QDSP6_APRV2
 	select MSM_QDSP6V2_CODECS
-	select SND_SOC_WCD9330
-	select SND_SOC_WCD9335
 	select SND_SOC_MSM8X16_WCD
-	select SND_SOC_WSA881X_ANALOG
-	select SND_SOC_WSA881X
 	select MFD_CORE
 	select QTI_PP
-	select DTS_SRS_TM
-	select DOLBY_DAP
-	select DOLBY_DS2
-	select SND_SOC_CPE
-	select SND_SOC_WCD_CPE
-	select SND_SOC_MSM_HDMI_DBA_CODEC_RX
-	select SND_HWDEP
-        select DTS_EAGLE
 	help
 	 To add support for SoC audio on MSM8996.
 	 This will enable sound soc drivers which
@@ -219,6 +207,48 @@ config SND_SOC_MSM8X16
 	 the machine driver and the corresponding
 	 DAI-links
 
+config SND_SOC_MSM8X16_RT5659
+	tristate "SoC Machine driver for MSM8X16 RT5659 boards"
+	depends on ARCH_MSMTITANIUM || ARCH_MSM8916 || ARCH_MSM8937
+	select SND_SOC_QDSP6V2
+	select SND_SOC_MSM_STUB
+	select SND_SOC_MSM_HOSTLESS_PCM
+	select SND_DYNAMIC_MINORS
+	select MSM_QDSP6_APRV2
+	select MSM_QDSP6V2_CODECS
+	select SND_SOC_MSM8X16_WCD
+	select SND_SOC_RT5659
+	select AUDIO_EXT_CLK
+	select MFD_CORE
+	select QTI_PP
+	help
+	 To add support for SoC audio on MSM8X16 with RT5659 codec.
+	 This will enable sound soc drivers which
+	 interfaces with DSP, also it will enable
+	 the machine drivers and the corresponding
+	 DAI-links.
+
+config SND_SOC_MSM8X16_RT5665
+	tristate "SoC Machine driver for MSM8X16 RT5665 boards"
+	depends on ARCH_MSMTITANIUM || ARCH_MSM8916 || ARCH_MSM8937 || ARCH_MSM8953
+	select SND_SOC_QDSP6V2
+	select SND_SOC_MSM_STUB
+	select SND_SOC_MSM_HOSTLESS_PCM
+	select SND_DYNAMIC_MINORS
+	select MSM_QDSP6_APRV2
+	select MSM_QDSP6V2_CODECS
+	select SND_SOC_MSM8X16_WCD
+	select SND_SOC_RT5665
+	select AUDIO_EXT_CLK
+	select MFD_CORE
+	select QTI_PP
+	help
+	 To add support for SoC audio on MSM8X16 with RT5665 codec.
+	 This will enable sound soc drivers which
+	 interfaces with DSP, also it will enable
+	 the machine drivers and the corresponding
+	 DAI-links.
+
 config SND_SOC_MDM9650
         tristate "SoC Machine driver for MDM9650 or SDX20 boards"
         depends on ARCH_MDM9650 || ARCH_SDX20
@@ -302,4 +332,64 @@ config SND_SOC_MDM9640
 	 the machine drivers and the corresponding
 	 DAI-links.
 
+config SEC_SND_SOLUTION
+	tristate "Samsung sound solution"
+	default n
+	help
+		To add support for Samsung sound solution
+
+config SEC_VOC_SOLUTION
+	tristate "Samsung voice solution"
+	default n
+	help
+		To add support for Samsung voice solution
+
+config SND_SOC_JACK_AUDIO
+	tristate "Samsung Jack Audio"
+	default n
+	help
+		To add support for Samsung Jack Audio
+
+config AUDIO_SPEAKER_OUT_NXP_AMP_ENABLE
+	tristate "Samsung audio speaker NXP amp feature"
+	default n
+	help
+		To add support for speaker NXP amp
+
+config SND_SOC_SECONDARY_AUXPCM
+	tristate "SND_SOC_SECONDARY_AUXPCM feature"
+	default n
+	help
+		To add support for secondary aux pcm
+
+config SND_EXT_MIC_BIAS
+	tristate "SND_EXT_MIC_BIAS feature"
+	default n
+	help
+		To add support for external mic bias
+
+config SEC_SND_ADAPTATION
+	tristate "Samsung ADSP Adaptation"
+	default n
+	help
+		To add support for Samsung ADSP solution
+
+config SND_SOC_TDM_INTF
+	tristate "TDM interface support"
+	default n
+	help
+		To add support for TDM interface
+
+config SEC_SND_TFA98XX
+	tristate "NXP amp support"
+	default n
+	help
+		To add support for NXP amp
+
+config SEC_SND_DEBUG
+	tristate "Debug code for Samsung Audio"
+	default n
+	help
+		Debug code for Samsung Audio
+
 endmenu
