--- original/sound/soc/msm/Makefile	2020-05-07 08:31:28.677761231 +0200
+++ changed/sound/soc/msm/Makefile	2019-06-03 13:39:55.000000000 +0200
@@ -30,11 +30,24 @@ obj-$(CONFIG_SND_SOC_MSM8996) += snd-soc
 snd-soc-msm8996-vm-objs := apq8096-auto.o
 obj-$(CONFIG_SND_SOC_MSM8996_VM) += snd-soc-msm8996-vm.o
 
+# for MSM 8x16/8952 common obj for sound card driver
+snd-soc-msm8x16-audio-pinctl-objs := msm-audio-pinctrl.o
+obj-$(CONFIG_SND_SOC_MSM8X16) += snd-soc-msm8x16-audio-pinctl.o
+obj-$(CONFIG_SND_SOC_MSM8X16_RT5659) += snd-soc-msm8x16-audio-pinctl.o
+obj-$(CONFIG_SND_SOC_MSM8X16_RT5665) += snd-soc-msm8x16-audio-pinctl.o
+
 # for MSM 8x16/8952 sound card driver
-snd-soc-msm8x16-objs := msm8952.o msm-audio-pinctrl.o
-snd-soc-msm8x16-objs += msm8952-slimbus.o msm8952-dai-links.o
+snd-soc-msm8x16-objs := msm8952.o
 obj-$(CONFIG_SND_SOC_MSM8X16) += snd-soc-msm8x16.o
 
+# for MSM 8x16/8952 rt5659 sound card driver
+snd-soc-msm8x16-rt5659-objs := msm8952_rt5659.o
+obj-$(CONFIG_SND_SOC_MSM8X16_RT5659) += snd-soc-msm8x16-rt5659.o
+
+# for MSM 8x16/8952 rt5665 sound card driver
+snd-soc-msm8x16-rt5665-objs := msm8952_rt5665.o
+obj-$(CONFIG_SND_SOC_MSM8X16_RT5665) += snd-soc-msm8x16-rt5665.o
+
 # for MDM 9650 sound card driver
 snd-soc-mdm9650-objs := mdm9650.o msm-audio-pinctrl.o
 obj-$(CONFIG_SND_SOC_MDM9650) += snd-soc-mdm9650.o
