--- original/sound/soc/msm/msm8994.c	2020-05-07 08:06:24.482161659 +0200
+++ changed/sound/soc/msm/msm8994.c	2019-06-03 13:39:55.000000000 +0200
@@ -2572,7 +2572,7 @@ static struct snd_soc_dai_link msm8994_c
 		.name = "MSM8994 Compr8",
 		.stream_name = "COMPR8",
 		.cpu_dai_name = "MultiMedia8",
-		.platform_name = "msm-compr-dsp",
+		.platform_name = "msm-compress-dsp",
 		.dynamic = 1,
 		.trigger = {SND_SOC_DPCM_TRIGGER_POST,
 			 SND_SOC_DPCM_TRIGGER_POST},
