--- original/sound/soc/codecs/Kconfig	2020-05-07 08:31:28.588762032 +0200
+++ changed/sound/soc/codecs/Kconfig	2019-06-03 13:39:55.000000000 +0200
@@ -86,6 +86,8 @@ config SND_SOC_ALL_CODECS
 	select SND_SOC_RT5651 if I2C
 	select SND_SOC_RT5670 if I2C
 	select SND_SOC_RT5677 if I2C
+	select SND_SOC_RT5659 if I2C
+	select SND_SOC_RT5665 if I2C
 	select SND_SOC_SGTL5000 if I2C
 	select SND_SOC_SI476X if MFD_SI476X_CORE
 	select SND_SOC_SIRF_AUDIO_CODEC
@@ -476,6 +478,7 @@ config SND_SOC_RL6231
 	default y if SND_SOC_RT5651=y
 	default y if SND_SOC_RT5670=y
 	default y if SND_SOC_RT5677=y
+	default y if SND_SOC_RT5665=y
 	default m if SND_SOC_RT5640=m
 	default m if SND_SOC_RT5645=m
 	default m if SND_SOC_RT5651=m
@@ -504,6 +507,12 @@ config SND_SOC_RT5670
 config SND_SOC_RT5677
 	tristate
 
+config SND_SOC_RT5659
+	tristate
+
+config SND_SOC_RT5665
+	tristate
+
 #Freescale sgtl5000 codec
 config SND_SOC_SGTL5000
 	tristate "Freescale SGTL5000 CODEC"
@@ -675,7 +684,7 @@ config AUDIO_EXT_CLK
 
 config SND_SOC_WCD_MBHC
 	tristate
-	default y if (SND_SOC_MSM8909_WCD=y || SND_SOC_MSM8X16_WCD=y || SND_SOC_WCD9335=y) && SND_SOC_MDM9650!=y
+	default y if (SND_SOC_MSM8909_WCD=y || SND_SOC_WCD9335=y) && SND_SOC_MDM9650!=y
 
 config SND_SOC_WL1273
 	tristate
@@ -846,6 +855,9 @@ config SND_SOC_WM9712
 config SND_SOC_WM9713
 	tristate
 
+# DBMDx
+source "sound/soc/codecs/dbmdx/Kconfig"
+
 # Amp
 config SND_SOC_LM4857
 	tristate
@@ -879,4 +891,17 @@ config SND_SOC_MSM_HDMI_CODEC_RX
 config SND_SOC_MSM_HDMI_DBA_CODEC_RX
 	bool "HDMI DBA Audio Playback"
 	depends on MSM_DBA
+
+config SND_SOC_TFA9872
+	tristate "Support TFA9872"
+	default n
+	help
+	TFA9872 driver
+
+config SND_SOC_TFA9896
+	tristate "Support TFA9896"
+	default n
+	help
+	TFA9896 driver
+
 endmenu
