--- original/sound/soc/msm/msm-dai-fe.c	2020-05-07 08:31:28.694761078 +0200
+++ changed/sound/soc/msm/msm-dai-fe.c	2019-06-03 13:39:55.000000000 +0200
@@ -912,6 +912,36 @@ static struct snd_soc_dai_driver msm_fe_
 		.name = "QUAT_MI2S_RX_HOSTLESS",
 		.probe = fe_dai_probe,
 	},
+	{
+		.capture = {
+			.stream_name = "Quinary MI2S_TX Hostless Capture",
+			.aif_name = "QUIN_MI2S_UL_HL",
+			.rates = SNDRV_PCM_RATE_8000_48000,
+			.formats = SNDRV_PCM_FMTBIT_S16_LE,
+			.channels_min = 1,
+			.channels_max = 2,
+			.rate_min = 8000,
+			.rate_max = 48000,
+		},
+		.ops = &msm_fe_dai_ops,
+		.name = "QUIN_MI2S_TX_HOSTLESS",
+		.probe = fe_dai_probe,
+	},
+	{
+		.playback = {
+			.stream_name = "Quinary MI2S_RX Hostless Playback",
+			.aif_name = "QUIN_MI2S_DL_HL",
+			.rates = SNDRV_PCM_RATE_8000_48000,
+			.formats = SNDRV_PCM_FMTBIT_S16_LE,
+			.channels_min = 1,
+			.channels_max = 2,
+			.rate_min = 8000,
+			.rate_max =    48000,
+		},
+		.ops = &msm_fe_dai_ops,
+		.name = "QUIN_MI2S_RX_HOSTLESS",
+		.probe = fe_dai_probe,
+	},
 	/* TDM Hostless */
 	{
 		.capture = {
@@ -2544,6 +2574,20 @@ static struct snd_soc_dai_driver msm_fe_
 		.probe = fe_dai_probe,
 	},
 	{
+#ifdef CONFIG_SND_SOC_JACK_AUDIO
+		.playback = {
+			.stream_name = "MultiMedia17 Playback",
+			.aif_name = "MM_DL17",
+			.rates = (SNDRV_PCM_RATE_8000_192000 |
+				SNDRV_PCM_RATE_KNOT),
+			.formats = (SNDRV_PCM_FMTBIT_S16_LE |
+				SNDRV_PCM_FMTBIT_S24_LE),
+			.channels_min = 1,
+			.channels_max = 8,
+			.rate_min = 8000,
+			.rate_max = 192000,
+		},
+#endif
 		.capture = {
 			.stream_name = "MultiMedia17 Capture",
 			.aif_name = "MM_UL17",
@@ -2559,7 +2603,9 @@ static struct snd_soc_dai_driver msm_fe_
 			.rate_max =     192000,
 		},
 		.ops = &msm_fe_Multimedia_dai_ops,
+#ifndef CONFIG_SND_SOC_JACK_AUDIO
 		.compress_dai = 1,
+#endif
 		.name = "MultiMedia17",
 		.probe = fe_dai_probe,
 	},
