--- original/sound/soc/msm/qdsp6v2/q6adm.c	2020-05-07 11:40:16.709311736 +0200
+++ changed/sound/soc/msm/qdsp6v2/q6adm.c	2019-06-03 13:39:55.000000000 +0200
@@ -26,6 +26,10 @@
 #include "msm-dts-srs-tm-config.h"
 #include <sound/adsp_err.h>
 
+#ifdef CONFIG_SEC_SND_ADAPTATION
+#include <sound/sec_adaptation.h>
+#endif /* CONFIG_SEC_SND_ADAPTATION */
+
 #define TIMEOUT_MS 1000
 
 #define RESET_COPP_ID 99
@@ -2397,9 +2401,11 @@ int adm_open(int port_id, int path, int
 	int port_idx, copp_idx, flags;
 	int tmp_port = q6audio_get_port_id(port_id);
 
-	pr_debug("%s:port %#x path:%d rate:%d mode:%d perf_mode:%d,topo_id %d\n",
+	pr_info("%s:port %#x path:%d rate:%d mode:%d perf_mode:%d,topo_id %d\n",
 		 __func__, port_id, path, rate, channel_mode, perf_mode,
 		 topology);
+	pr_info("%s:bit_width:%d app_type:%#x acdb_id:%d\n",
+		__func__, bit_width, app_type, acdb_id);
 
 	port_id = q6audio_convert_virtual_to_portid(port_id);
 	port_idx = adm_validate_and_get_port_index(port_id);
@@ -2444,9 +2450,27 @@ int adm_open(int port_id, int path, int
 
 	if ((topology == VPM_TX_SM_ECNS_COPP_TOPOLOGY) ||
 	    (topology == VPM_TX_DM_FLUENCE_COPP_TOPOLOGY) ||
-	    (topology == VPM_TX_DM_RFECNS_COPP_TOPOLOGY))
+	    (topology == VPM_TX_DM_RFECNS_COPP_TOPOLOGY)
+	    || (topology == VPM_TX_SM_LVVEFQ_COPP_TOPOLOGY)
+	    || (topology == VPM_TX_DM_LVVEFQ_COPP_TOPOLOGY)
+	    || (topology == VPM_TX_SM_LVSAFQ_COPP_TOPOLOGY)
+	    || (topology == VOICE_TX_DIAMONDVOICE_FVSAM_DM)
+	    || (topology == VOICE_TX_DIAMONDVOICE_FVSAM_QM)
+#ifdef CONFIG_SEC_SND_ADAPTATION
+	    || (topology == VOICE_TX_DIAMONDVOICE_FRSAM_DM)
+#endif
+	    )
 		rate = 16000;
 
+#if defined(CONFIG_SND_SOC_MSM8X16_RT5659) || defined(CONFIG_SND_SOC_MSM8X16_RT5665)
+	/*
+	 * because ADSP cannot support 24bit recording,
+	 * fix to I2S tx bitwidth to 16bit
+	 */
+	if(port_id == AFE_PORT_ID_QUINARY_MI2S_TX)
+		bit_width = 16;
+#endif /* CONFIG_SND_SOC_MSM8X16_RT5659 */
+
 	copp_idx = adm_get_idx_if_copp_exists(port_idx, topology, perf_mode,
 					      rate, bit_width, app_type,
 					      session_type);
@@ -2917,7 +2941,7 @@ int adm_close(int port_id, int perf_mode
 	int ret = 0, port_idx;
 	int copp_id = RESET_COPP_ID;
 
-	pr_debug("%s: port_id=0x%x perf_mode: %d copp_idx: %d\n", __func__,
+	pr_info("%s: port_id=0x%x perf_mode: %d copp_idx: %d\n", __func__,
 		 port_id, perf_mode, copp_idx);
 
 	port_id = q6audio_convert_virtual_to_portid(port_id);
