--- original/Android.bp	2020-05-07 08:31:19.539843439 +0200
+++ changed/Android.bp	2019-06-03 13:39:47.000000000 +0200
@@ -16,12 +16,16 @@ gensrcs {
     tools: ["unifdef"],
     tool_files: ["scripts/headers_install.sh"],
     export_include_dirs: ["include/uapi"],
-    srcs: ["include/uapi/linux/qseecom.h"],
+    srcs: [
+          "include/uapi/linux/qseecom.h",
+          "include/uapi/linux/ion.h",
+    ],
     output_extension: "h",
 }
 
 cc_library_headers {
     name: "qseecom-kernel-headers",
     generated_headers: ["qseecom-kernel-includes"],
+    vendor_available: true,
     export_generated_headers: ["qseecom-kernel-includes"],
 }
