--- original/block/partition-generic.c	2020-05-07 08:31:21.165828811 +0200
+++ changed/block/partition-generic.c	2019-06-03 13:39:49.000000000 +0200
@@ -18,6 +18,12 @@
 #include <linux/genhd.h>
 #include <linux/blktrace_api.h>
 
+#ifdef CONFIG_BLOCK_SUPPORT_STLOG
+#include <linux/fslog.h>
+#else
+#define ST_LOG(fmt,...)
+#endif
+
 #include "partitions/check.h"
 
 #ifdef CONFIG_BLK_DEV_MD
@@ -254,6 +260,10 @@ void delete_partition(struct gendisk *di
 	struct disk_part_tbl *ptbl = disk->part_tbl;
 	struct hd_struct *part;
 
+#ifdef CONFIG_BLOCK_SUPPORT_STLOG
+	struct device *dev;
+#endif
+
 	if (partno >= ptbl->len)
 		return;
 
@@ -264,6 +274,11 @@ void delete_partition(struct gendisk *di
 	rcu_assign_pointer(ptbl->part[partno], NULL);
 	rcu_assign_pointer(ptbl->last_lookup, NULL);
 	kobject_put(part->holder_dir);
+#ifdef CONFIG_BLOCK_SUPPORT_STLOG
+	dev = part_to_dev(part);
+	ST_LOG("<%s> KOBJ_REMOVE %d:%d %s",
+	__func__,MAJOR(dev->devt),MINOR(dev->devt),dev->kobj.name);
+#endif	
 	device_del(part_to_dev(part));
 
 	hd_struct_put(part);
