--- original/block/blk-merge.c	2020-05-07 08:06:08.861315153 +0200
+++ changed/block/blk-merge.c	2019-06-03 13:39:49.000000000 +0200
@@ -665,6 +665,15 @@ bool blk_rq_merge_ok(struct request *rq,
 	    !blk_write_same_mergeable(rq->bio, bio))
 		return false;
 
+#ifdef CONFIG_JOURNAL_DATA_TAG
+	/* journal tagged bio can only be merged to REQ_META request */
+	if ((bio_flagged(bio, BIO_JMETA) || bio_flagged(bio, BIO_JOURNAL))
+			&& !(rq->cmd_flags & REQ_META))
+		return false;
+#endif
+	if (bio_flagged(bio, BIO_BYPASS) && !(rq->cmd_flags & REQ_BYPASS))
+		return false;
+
 	if (q->queue_flags & (1 << QUEUE_FLAG_SG_GAPS)) {
 		struct bio_vec *bprev;
 
