--- original/block/ioctl.c	2020-05-07 08:06:08.876315006 +0200
+++ changed/block/ioctl.c	2019-06-03 13:39:49.000000000 +0200
@@ -9,6 +9,12 @@
 #include <linux/blktrace_api.h>
 #include <asm/uaccess.h>
 
+#ifdef CONFIG_BLOCK_SUPPORT_STLOG
+#include <linux/fslog.h>
+#else
+#define ST_LOG(fmt,...)
+#endif
+
 static int blkpg_ioctl(struct block_device *bdev, struct blkpg_ioctl_arg __user *arg)
 {
 	struct block_device *bdevp;
@@ -182,6 +188,9 @@ static int blk_ioctl_discard(struct bloc
 		return -EINVAL;
 	if (secure)
 		flags |= BLKDEV_DISCARD_SECURE;
+	printk("%s %d:%d %llu %llu", secure?"SECDIS":"DIS", MAJOR(bdev->bd_dev), MINOR(bdev->bd_dev), 
+		(unsigned long long)start, (unsigned long long)len);
+	ST_LOG("%s %d:%d %lu %lu",secure?"SECDIS":"DIS",MAJOR(bdev->bd_dev),MINOR(bdev->bd_dev),start,len);
 	return blkdev_issue_discard(bdev, start, len, GFP_KERNEL, flags);
 }
 
