--- original/block/bio.c	2020-05-07 08:31:21.146828982 +0200
+++ changed/block/bio.c	2019-06-03 13:39:49.000000000 +0200
@@ -565,6 +565,10 @@ void __bio_clone_fast(struct bio *bio, s
 	 */
 	bio->bi_bdev = bio_src->bi_bdev;
 	bio->bi_flags |= 1 << BIO_CLONED;
+#ifdef CONFIG_JOURNAL_DATA_TAG
+	bio->bi_flags |= bio_src->bi_flags & BIO_JOURNAL_TAG_MASK;
+#endif
+	bio->bi_flags |= bio_src->bi_flags & 1UL << BIO_BYPASS;
 	bio->bi_rw = bio_src->bi_rw;
 	bio->bi_iter = bio_src->bi_iter;
 	bio->bi_io_vec = bio_src->bi_io_vec;
@@ -651,6 +655,10 @@ struct bio *bio_clone_bioset(struct bio
 	bio->bi_rw		= bio_src->bi_rw;
 	bio->bi_iter.bi_sector	= bio_src->bi_iter.bi_sector;
 	bio->bi_iter.bi_size	= bio_src->bi_iter.bi_size;
+#ifdef CONFIG_JOURNAL_DATA_TAG
+	bio->bi_flags |= bio_src->bi_flags & BIO_JOURNAL_TAG_MASK;
+#endif
+	bio->bi_flags |= bio_src->bi_flags & 1UL << BIO_BYPASS;
 
 	if (bio->bi_rw & REQ_DISCARD)
 		goto integrity_clone;
