--- original/security/selinux/exports.c	2020-05-07 08:06:23.817168193 +0200
+++ changed/security/selinux/exports.c	2019-06-03 13:39:54.000000000 +0200
@@ -15,9 +15,28 @@
 #include <linux/selinux.h>
 
 #include "security.h"
+#include "avc.h"
 
 bool selinux_is_enabled(void)
 {
+// [ SEC_SELINUX_PORTING_COMMON
+#ifdef CONFIG_ALWAYS_ENFORCE
+	return true;
+#else
 	return selinux_enabled;
+#endif
+// ] SEC_SELINUX_PORTING_COMMON
 }
 EXPORT_SYMBOL_GPL(selinux_is_enabled);
+
+bool selinux_is_enforcing(void)
+{
+// [ SEC_SELINUX_PORTING_COMMON
+#ifdef CONFIG_ALWAYS_ENFORCE
+	return true;
+#else
+	return selinux_enforcing;
+#endif
+// ] SEC_SELINUX_PORTING_COMMON
+}
+EXPORT_SYMBOL_GPL(selinux_is_enforcing);
