--- original/security/integrity/iint.c	2020-05-07 08:06:23.793168429 +0200
+++ changed/security/integrity/iint.c	2019-06-03 13:39:54.000000000 +0200
@@ -70,6 +70,13 @@ struct integrity_iint_cache *integrity_i
 
 static void iint_free(struct integrity_iint_cache *iint)
 {
+#ifdef CONFIG_FIVE
+	kfree(iint->five_label);
+	iint->five_label = NULL;
+	iint->five_flags = 0UL;
+	iint->five_status = FIVE_FILE_UNKNOWN;
+	iint->five_signing = false;
+#endif
 	kfree(iint->ima_hash);
 	iint->ima_hash = NULL;
 	iint->version = 0;
@@ -153,6 +160,11 @@ static void init_once(void *foo)
 
 	memset(iint, 0, sizeof(*iint));
 	iint->version = 0;
+#ifdef CONFIG_FIVE
+	iint->five_flags = 0UL;
+	iint->five_status = FIVE_FILE_UNKNOWN;
+	iint->five_signing = false;
+#endif
 	iint->flags = 0UL;
 	iint->ima_file_status = INTEGRITY_UNKNOWN;
 	iint->ima_mmap_status = INTEGRITY_UNKNOWN;
