--- original/security/selinux/Makefile	2020-05-07 08:06:23.815168213 +0200
+++ changed/security/selinux/Makefile	2019-06-03 13:39:54.000000000 +0200
@@ -2,6 +2,24 @@
 # Makefile for building the SELinux module as part of the kernel tree.
 #
 
+# [ SEC_SELINUX_PORTING_COMMON
+ifeq ($(TARGET_BUILD_VARIANT), eng)
+ifeq ($(SEC_BUILD_OPTION_PRODUCT_SHIP), false)
+ifeq ($(SEC_BUILD_OPTION_SELINUX_ENFORCE),true)
+EXTRA_CFLAGS += -DSEC_SELINUX_DEBUG
+endif
+endif
+endif
+
+ifeq ($(TARGET_BUILD_VARIANT), user)
+ifeq ($(SEC_BUILD_OPTION_PRODUCT_SHIP), true)
+ifeq ($(SEC_BUILD_OPTION_SELINUX_ENFORCE),true)
+EXTRA_CFLAGS += -DCONFIG_ALWAYS_ENFORCE=true
+endif
+endif
+endif
+# ] SEC_SELINUX_PORTING_COMMON
+
 obj-$(CONFIG_SECURITY_SELINUX) := selinux.o
 
 selinux-y := avc.o hooks.o selinuxfs.o netlink.o nlmsgtab.o netif.o \
