--- original/security/Makefile	2020-05-07 08:06:23.772168635 +0200
+++ changed/security/Makefile	2019-06-03 13:39:54.000000000 +0200
@@ -9,6 +9,8 @@ subdir-$(CONFIG_SECURITY_TOMOYO)
 subdir-$(CONFIG_SECURITY_APPARMOR)	+= apparmor
 subdir-$(CONFIG_SECURITY_YAMA)		+= yama
 subdir-$(CONFIG_ARCH_MSM)		+= pfe
+subdir-$(CONFIG_SECURITY_DEFEX)		+= samsung/defex_lsm
+subdir-$(CONFIG_SECURITY_DSMS)		+= samsung/dsms
 
 # always enable default capabilities
 obj-y					+= commoncap.o
@@ -26,7 +28,26 @@ obj-$(CONFIG_SECURITY_APPARMOR)		+= appa
 obj-$(CONFIG_SECURITY_YAMA)		+= yama/
 obj-$(CONFIG_ARCH_MSM)		        += pfe/
 obj-$(CONFIG_CGROUP_DEVICE)		+= device_cgroup.o
+obj-$(CONFIG_SECURITY_DEFEX)		+= samsung/defex_lsm/built-in.o
+obj-$(CONFIG_SECURITY_DSMS)		+= samsung/dsms/
 
 # Object integrity file lists
 subdir-$(CONFIG_INTEGRITY)		+= integrity
 obj-$(CONFIG_INTEGRITY)			+= integrity/
+
+# PROCA module
+subdir-$(CONFIG_PROCA)			+= proca
+obj-$(CONFIG_PROCA)			+= proca/
+
+# TIMA uevent
+obj-$(CONFIG_TIMA_UEVENT)     += tima_uevent/
+
+# Knox SDP
+obj-$(CONFIG_SDP) += sdp/
+obj-$(CONFIG_SDP) += sdp/built-in.o
+
+#TIMA ICCC
+obj-$(CONFIG_TZ_ICCC)     += tz_iccc/
+
+#KNOX TEST
+obj-$(CONFIG_KNOX_TEST)     += knox_test/
