--- original/kernel/time/timekeeping.c	2020-05-07 08:31:27.519771649 +0200
+++ changed/kernel/time/timekeeping.c	2019-06-03 13:39:54.000000000 +0200
@@ -1070,6 +1070,15 @@ static void __timekeeping_inject_sleepti
 	tk_xtime_add(tk, delta);
 	tk_set_wall_to_mono(tk, timespec64_sub(tk->wall_to_monotonic, *delta));
 	tk_update_sleep_time(tk, timespec64_to_ktime(*delta));
+#ifdef CONFIG_SEC_PM_DEBUG
+#ifdef CONFIG_ARCH_MSM8917
+	printk("Suspended for %llu.%03lu seconds\n",
+			delta->tv_sec, delta->tv_nsec / NSEC_PER_MSEC);
+#else
+	printk("Suspended for %lu.%03lu seconds\n",
+			delta->tv_sec, delta->tv_nsec / NSEC_PER_MSEC);
+#endif
+#endif
 	tk_debug_account_sleep_time(delta);
 }
 
