--- original/kernel/power/qos.c	2020-05-07 08:06:22.207184013 +0200
+++ changed/kernel/power/qos.c	2019-06-03 13:39:53.000000000 +0200
@@ -126,6 +126,19 @@ static struct pm_qos_object memory_bandw
 	.name = "memory_bandwidth",
 };
 
+static BLOCKING_NOTIFIER_HEAD(hyst_bias_notifier);
+static struct pm_qos_constraints hyst_bias_constraints = {
+	.list = PLIST_HEAD_INIT(hyst_bias_constraints.list),
+	.target_value = PM_QOS_HIST_BIAS_DEFAULT_VALUE,
+	.default_value = PM_QOS_HIST_BIAS_DEFAULT_VALUE,
+	.no_constraint_value = PM_QOS_HIST_BIAS_DEFAULT_VALUE,
+	.type = PM_QOS_MAX,
+	.notifiers = &hyst_bias_notifier,
+};
+static struct pm_qos_object hyst_bias_pm_qos = {
+	.constraints = &hyst_bias_constraints,
+	.name = "hyst_bias",
+};
 
 static struct pm_qos_object *pm_qos_array[] = {
 	&null_pm_qos,
@@ -133,6 +146,7 @@ static struct pm_qos_object *pm_qos_arra
 	&network_lat_pm_qos,
 	&network_throughput_pm_qos,
 	&memory_bandwidth_pm_qos,
+	&hyst_bias_pm_qos,
 };
 
 static ssize_t pm_qos_power_write(struct file *filp, const char __user *buf,
