--- original/kernel/power/Kconfig	2020-05-07 08:31:27.454772234 +0200
+++ changed/kernel/power/Kconfig	2019-06-03 13:39:53.000000000 +0200
@@ -27,6 +27,16 @@ config WAKELOCK
 
 	  Say Y if you are running an android userspace.
 
+config CPU_FREQ_LIMIT_USERSPACE
+	bool "User space cpufreq limit interface"
+	depends on CPU_FREQ_LIMIT
+	default n
+	help
+	  Add sysfs to limit CPU frequency for a userspace program.
+	  /sys/power/cpufreq_min_limit
+	  /sys/power/cpufreq_max_limit
+	  /sys/power/cpufreq_table
+
 config HIBERNATE_CALLBACKS
 	bool
 
@@ -329,3 +339,9 @@ config SUSPEND_TIME
 config DEDUCE_WAKEUP_REASONS
 	bool
 	default n
+
+config CPU_FREQ_LIMIT_SCHEDBOOST
+	bool
+	default n
+	help
+	  Sched booster will be active when cpu min limit is set through /sys/power/cpufreq_min_limit
