--- original/kernel/trace/blktrace.c	2020-05-07 08:06:22.384182274 +0200
+++ changed/kernel/trace/blktrace.c	2019-06-03 13:39:54.000000000 +0200
@@ -1855,6 +1855,8 @@ void blk_dump_cmd(char *buf, struct requ
 	}
 }
 
+SIO_PATCH_VERSION(ftrace_discard_bugfix, 1, 0, "");
+
 void blk_fill_rwbs(char *rwbs, u32 rw, int bytes)
 {
 	int i = 0;
@@ -1862,10 +1864,10 @@ void blk_fill_rwbs(char *rwbs, u32 rw, i
 	if (rw & REQ_FLUSH)
 		rwbs[i++] = 'F';
 
-	if (rw & WRITE)
-		rwbs[i++] = 'W';
-	else if (rw & REQ_DISCARD)
+	if (rw & REQ_DISCARD)
 		rwbs[i++] = 'D';
+        else if (rw & WRITE)
+                rwbs[i++] = 'W';
 	else if (bytes)
 		rwbs[i++] = 'R';
 	else
