--- original/kernel/time/hrtimer.c	2020-05-07 08:31:27.511771721 +0200
+++ changed/kernel/time/hrtimer.c	2019-06-03 13:39:54.000000000 +0200
@@ -56,6 +56,8 @@
 
 #include "timekeeping.h"
 
+#include <linux/sec_debug.h>
+
 /*
  * The timer bases:
  *
@@ -1224,7 +1226,9 @@ static void __run_hrtimer(struct hrtimer
 	 */
 	raw_spin_unlock(&cpu_base->lock);
 	trace_hrtimer_expire_entry(timer, now);
+	secdbg_msg("hrtimer %pS entry", fn);
 	restart = fn(timer);
+	secdbg_msg("hrtimer %pS exit", fn);
 	trace_hrtimer_expire_exit(timer);
 	raw_spin_lock(&cpu_base->lock);
 
