--- original/kernel/kallsyms.c	2020-05-07 08:06:22.135184721 +0200
+++ changed/kernel/kallsyms.c	2019-06-03 13:39:53.000000000 +0200
@@ -27,6 +27,9 @@
 
 #include <asm/sections.h>
 
+#include <linux/sec_debug.h>
+#include <linux/sec_debug_summary.h>
+
 #ifdef CONFIG_KALLSYMS_ALL
 #define all_var 1
 #else
@@ -52,6 +55,28 @@ extern const u16 kallsyms_token_index[]
 
 extern const unsigned long kallsyms_markers[] __weak;
 
+#ifdef CONFIG_SEC_DEBUG_SUMMARY
+void sec_debug_summary_set_kallsyms_info(
+		struct sec_debug_summary_data_apss *apss)
+{
+	apss->ksyms.addresses_pa = __pa(kallsyms_addresses);
+	apss->ksyms.names_pa = __pa(kallsyms_names);
+	apss->ksyms.num_syms = kallsyms_num_syms;
+	apss->ksyms.token_table_pa = __pa(kallsyms_token_table);
+	apss->ksyms.token_index_pa = __pa(kallsyms_token_index);
+	apss->ksyms.markers_pa = __pa(kallsyms_markers);
+
+	apss->ksyms.sect.sinittext = (uintptr_t)_sinittext;
+	apss->ksyms.sect.einittext = (uintptr_t)_einittext;
+	apss->ksyms.sect.stext = (uintptr_t)_stext;
+	apss->ksyms.sect.etext = (uintptr_t)_etext;
+	apss->ksyms.sect.end = (uintptr_t)_end;
+
+	apss->ksyms.kallsyms_all = all_var;
+	apss->ksyms.magic = SEC_DEBUG_SUMMARY_MAGIC1;
+}
+#endif
+
 static inline int is_kernel_inittext(unsigned long addr)
 {
 	if (addr >= (unsigned long)_sinittext
