--- original/kernel/irq/handle.c	2020-05-07 08:31:27.433772423 +0200
+++ changed/kernel/irq/handle.c	2019-06-03 13:39:53.000000000 +0200
@@ -15,11 +15,13 @@
 #include <linux/sched.h>
 #include <linux/interrupt.h>
 #include <linux/kernel_stat.h>
-
 #include <trace/events/irq.h>
 
 #include "internals.h"
 
+#include <linux/sec_debug.h>
+
+
 /**
  * handle_bad_irq - handle spurious and unhandled irqs
  * @irq:       the interrupt number
@@ -140,9 +142,11 @@ handle_irq_event_percpu(struct irq_desc
 	do {
 		irqreturn_t res;
 
+		sec_debug_irq_sched_log(irq, action->handler, (char *)action->name, IRQ_ENTRY);
 		trace_irq_handler_entry(irq, action);
 		res = action->handler(irq, action->dev_id);
 		trace_irq_handler_exit(irq, action, res);
+		sec_debug_irq_sched_log(irq, action->handler, (char *)action->name, IRQ_EXIT);
 
 		if (WARN_ONCE(!irqs_disabled(),"irq %u handler %pF enabled interrupts\n",
 			      irq, action->handler))
