--- original/kernel/sched/sched.h	2020-05-07 08:31:27.497771847 +0200
+++ changed/kernel/sched/sched.h	2019-06-03 13:39:54.000000000 +0200
@@ -1077,6 +1077,7 @@ static inline bool is_min_capacity_cpu(i
 	return cpu_max_possible_capacity(cpu) == min_max_possible_capacity;
 }
 
+
 /*
  * 'load' is in reference to "best cpu" at its best frequency.
  * Scale that in reference to a given cpu, accounting for how bad it is
