--- original/kernel/workqueue.c	2020-05-07 08:31:27.555771325 +0200
+++ changed/kernel/workqueue.c	2019-06-03 13:39:53.000000000 +0200
@@ -52,6 +52,9 @@
 
 #include "workqueue_internal.h"
 
+#include <linux/sec_debug.h>
+
+
 enum {
 	/*
 	 * worker_pool flags
@@ -2040,6 +2043,13 @@ __acquires(&pool->lock)
 	lock_map_acquire_read(&pwq->wq->lockdep_map);
 	lock_map_acquire(&lockdep_map);
 	trace_workqueue_execute_start(work);
+
+	if ((unsigned long)worker->current_func > PAGE_OFFSET)
+		secdbg_sched_msg("@%pS", worker->current_func);
+	else
+		secdbg_sched_msg("M:0x%lx",
+				(unsigned long)worker->current_func);
+
 	worker->current_func(work);
 	/*
 	 * While we must be careful to not use "work" after this, the trace
