--- original/kernel/exit.c	2020-05-07 08:31:27.423772512 +0200
+++ changed/kernel/exit.c	2019-06-03 13:39:53.000000000 +0200
@@ -60,6 +60,10 @@
 #include <asm/pgtable.h>
 #include <asm/mmu_context.h>
 
+#ifdef CONFIG_SECURITY_DEFEX
+#include <linux/defex.h>
+#endif
+
 static void exit_mm(struct task_struct *tsk);
 
 static void __unhash_process(struct task_struct *p, bool group_dead)
@@ -681,6 +685,10 @@ void do_exit(long code)
 	int group_dead;
 	TASKS_RCU(int tasks_rcu_i);
 
+#ifdef CONFIG_SECURITY_DEFEX
+	task_defex_zero_creds(current);
+#endif
+
 	profile_task_exit(tsk);
 	kcov_task_exit(tsk);
 
