--- original/fs/aio.c	2020-05-07 08:51:35.136255034 +0200
+++ changed/fs/aio.c	2019-06-03 13:39:53.000000000 +0200
@@ -175,6 +175,10 @@ static struct backing_dev_info aio_fs_ba
 	.capabilities   = BDI_CAP_NO_ACCT_AND_WRITEBACK | BDI_CAP_MAP_COPY,
 };
 
+#ifdef CONFIG_RKP_NS_PROT
+extern void rkp_set_mnt_flags(struct vfsmount *mnt, int flags);
+#endif
+
 static struct file *aio_private_file(struct kioctx *ctx, loff_t nr_pages)
 {
 	struct qstr this = QSTR_INIT("[aio]", 5);
@@ -230,8 +234,11 @@ static int __init aio_setup(void)
 	aio_mnt = kern_mount(&aio_fs);
 	if (IS_ERR(aio_mnt))
 		panic("Failed to create aio fs mount.");
+#ifdef CONFIG_RKP_NS_PROT
+	rkp_set_mnt_flags(aio_mnt, MNT_NOEXEC);
+#else
 	aio_mnt->mnt_flags |= MNT_NOEXEC;
-
+#endif
 	if (bdi_init(&aio_fs_backing_dev_info))
 		panic("Failed to init aio fs backing dev info.");
 
