--- original/fs/dcache.c	2020-05-07 08:51:35.178254622 +0200
+++ changed/fs/dcache.c	2019-06-03 13:39:53.000000000 +0200
@@ -3487,8 +3487,10 @@ EXPORT_SYMBOL(d_genocide);
 
 void __init vfs_caches_init_early(void)
 {
+	set_memsize_kernel_type(MEMSIZE_KERNEL_VFSHASH);
 	dcache_init_early();
 	inode_init_early();
+	set_memsize_kernel_type(MEMSIZE_KERNEL_OTHERS);
 }
 
 void __init vfs_caches_init(unsigned long mempages)
