--- original/fs/ecryptfs/Kconfig	2020-05-07 08:06:20.035205356 +0200
+++ changed/fs/ecryptfs/Kconfig	2019-06-03 13:39:53.000000000 +0200
@@ -4,6 +4,7 @@ config ECRYPT_FS
 	select CRYPTO_ECB
 	select CRYPTO_CBC
 	select CRYPTO_MD5
+	select CRYPTO_SHA256
 	help
 	  Encrypted filesystem that operates on the VFS layer.  See
 	  <file:Documentation/filesystems/ecryptfs.txt> to learn more about
@@ -20,3 +21,36 @@ config ECRYPT_FS_MESSAGING
 	  Enables the /dev/ecryptfs entry for use by ecryptfsd. This allows
 	  for userspace to wrap/unwrap file encryption keys by other
 	  backends, like OpenSSL.
+
+config WTL_ENCRYPTION_FILTER
+	bool "Enables filtering for some files not to encrypt on eCryptfs"
+	default n
+	depends on ECRYPT_FS
+	help
+	Modification of encrypted filesystem for SD card encryption
+
+config ECRYPTFS_FEK_INTEGRITY
+	bool "Enable check integrity of FEK(File encrypiton key) via HMAC"
+	default n
+	depends on ECRYPT_FS
+	help
+	MDFPP requirements: During file decryption of ecrypts, it should check integrity of FEK(File encrypiton key) via HMAC as per
+
+config SDP
+	bool "Enables SDP"
+	default n
+	depends on ECRYPT_FS
+	help
+	Sensitive Data Protection
+
+config SDP_KEY_DUMP
+	bool "SDP KEY DUMP"
+	default n
+	
+config DLP
+	bool "Enables DLP"
+	default n
+	depends on SDP
+	help
+	Data Loss Prevention	
+
