--- original/fs/ext4/Makefile	2020-05-07 08:31:26.358782093 +0200
+++ changed/fs/ext4/Makefile	2019-06-03 13:39:53.000000000 +0200
@@ -2,6 +2,10 @@
 # Makefile for the linux ext4-filesystem routines.
 #
 
+ifeq ($(SEC_BUILD_CONF_QUICK_DMVERITY),true)
+EXTRA_CFLAGS += -DVERIFY_META_ONLY=true
+endif
+
 obj-$(CONFIG_EXT4_FS) += ext4.o
 
 ext4-y	:= balloc.o bitmap.o dir.o file.o fsync.o ialloc.o inode.o page-io.o \
