--- original/fs/ext4/ext4_jbd2.c	2020-05-07 08:06:20.084204874 +0200
+++ changed/fs/ext4/ext4_jbd2.c	2019-06-03 13:39:53.000000000 +0200
@@ -43,7 +43,7 @@ static int ext4_journal_check_start(stru
 	journal_t *journal;
 
 	might_sleep();
-	if (sb->s_flags & MS_RDONLY)
+	if (sb->s_flags & MS_RDONLY && !ext4_journal_current_handle())
 		return -EROFS;
 	WARN_ON(sb->s_writers.frozen == SB_FREEZE_COMPLETE);
 	journal = EXT4_SB(sb)->s_journal;
@@ -258,7 +258,9 @@ int __ext4_handle_dirty_metadata(const c
 
 	might_sleep();
 
+#ifndef CONFIG_JOURNAL_DATA_TAG
 	set_buffer_meta(bh);
+#endif
 	set_buffer_prio(bh);
 	if (ext4_handle_valid(handle)) {
 		err = jbd2_journal_dirty_metadata(handle, bh);
@@ -315,6 +317,14 @@ int __ext4_handle_dirty_super(const char
 	struct buffer_head *bh = EXT4_SB(sb)->s_sbh;
 	int err = 0;
 
+	if (unlikely(le16_to_cpu(EXT4_SB(sb)->s_es->s_magic) !=
+			EXT4_SUPER_MAGIC)) {
+		print_bh(sb, bh, 0, EXT4_BLOCK_SIZE(sb));
+		if (test_opt(sb, ERRORS_PANIC))
+			panic("EXT4(Can not find EXT4_SUPER_MAGIC");
+		return -EIO;
+	}
+
 	ext4_superblock_csum_set(sb);
 	if (ext4_handle_valid(handle)) {
 		err = jbd2_journal_dirty_metadata(handle, bh);
