--- original/fs/file_table.c	2020-05-07 08:06:20.144204285 +0200
+++ changed/fs/file_table.c	2019-06-03 13:39:53.000000000 +0200
@@ -26,7 +26,7 @@
 #include <linux/hardirq.h>
 #include <linux/task_work.h>
 #include <linux/ima.h>
-
+#include <linux/task_integrity.h>
 #include <linux/atomic.h>
 
 #include "internal.h"
@@ -338,6 +338,7 @@ static void __fput(struct file *file)
 		if (file->f_op->fasync)
 			file->f_op->fasync(-1, file, 0);
 	}
+	five_file_free(file);
 	ima_file_free(file);
 	if (file->f_op->release)
 		file->f_op->release(inode, file);
