--- original/fs/fcntl.c	2020-05-07 08:31:26.461781167 +0200
+++ changed/fs/fcntl.c	2019-06-03 13:39:53.000000000 +0200
@@ -22,6 +22,8 @@
 #include <linux/pid_namespace.h>
 #include <linux/user_namespace.h>
 #include <linux/shmem_fs.h>
+#include <linux/task_integrity.h>
+#include <linux/proca.h>
 
 #include <asm/poll.h>
 #include <asm/siginfo.h>
@@ -334,6 +336,29 @@ static long do_fcntl(int fd, unsigned in
 	case F_GETPIPE_SZ:
 		err = pipe_fcntl(filp, cmd, arg);
 		break;
+#ifdef CONFIG_FIVE
+	case F_FIVE_SIGN:
+		err = five_fcntl_sign(filp,
+				(struct integrity_label __user *)arg);
+		break;
+	case F_FIVE_VERIFY_ASYNC:
+		err = five_fcntl_verify_async(filp);
+		break;
+	case F_FIVE_VERIFY_SYNC:
+		err = five_fcntl_verify_sync(filp);
+		break;
+#if defined(CONFIG_FIVE_PA_FEATURE) || defined(CONFIG_PROCA)
+	case F_FIVE_PA_SETXATTR:
+		err = proca_fcntl_setxattr(filp, (void __user *)arg);
+		break;
+#endif
+	case F_FIVE_EDIT:
+		err = five_fcntl_edit(filp);
+		break;
+	case F_FIVE_CLOSE:
+		err = five_fcntl_close(filp);
+		break;
+#endif
 	case F_ADD_SEALS:
 	case F_GET_SEALS:
 		err = shmem_fcntl(filp, cmd, arg);
