--- original/fs/f2fs/node.c	2020-05-07 08:51:35.255253868 +0200
+++ changed/fs/f2fs/node.c	2019-06-03 13:39:53.000000000 +0200
@@ -1234,6 +1234,8 @@ page_hit:
 			next_blkaddr_of_node(page));
 		err = -EINVAL;
 out_err:
+		if (PageUptodate(page))
+			print_block_data(sbi->sb, nid, page_address(page), 0, F2FS_BLKSIZE);
 		ClearPageUptodate(page);
 		f2fs_put_page(page, 1);
 		return ERR_PTR(err);
@@ -1990,7 +1992,11 @@ static void scan_nat_page(struct f2fs_sb
 			break;
 
 		blk_addr = le32_to_cpu(nat_blk->entries[i].block_addr);
-		f2fs_bug_on(sbi, blk_addr == NEW_ADDR);
+		if (unlikely(blk_addr == NEW_ADDR)) {
+			print_block_data(sbi->sb, current_nat_addr(sbi, start_nid),
+				page_address(nat_page), 0, F2FS_BLKSIZE);
+			f2fs_bug_on(sbi, 1);
+		}
 		if (blk_addr == NULL_ADDR) {
 			add_free_nid(sbi, start_nid, true, true);
 		} else {
