--- original/fs/fat/file.c	2020-05-07 08:06:20.138204344 +0200
+++ changed/fs/fat/file.c	2019-06-03 13:39:53.000000000 +0200
@@ -459,4 +459,10 @@ EXPORT_SYMBOL_GPL(fat_setattr);
 const struct inode_operations fat_file_inode_operations = {
 	.setattr	= fat_setattr,
 	.getattr	= fat_getattr,
+#ifdef CONFIG_FAT_VIRTUAL_XATTR
+	.setxattr	= fat_setxattr,
+	.getxattr	= fat_getxattr,
+	.listxattr	= fat_listxattr,
+	.removexattr	= fat_removexattr,
+#endif
 };
