--- original/fs/f2fs/sysfs.c	2020-05-07 08:51:35.277253652 +0200
+++ changed/fs/f2fs/sysfs.c	2019-06-03 13:39:53.000000000 +0200
@@ -225,7 +225,8 @@ out:
 	if (a->struct_type == RESERVED_BLOCKS) {
 		spin_lock(&sbi->stat_lock);
 		if (t > (unsigned long)(sbi->user_block_count -
-				F2FS_OPTION(sbi).root_reserved_blocks)) {
+				F2FS_OPTION(sbi).root_reserved_blocks -
+				F2FS_OPTION(sbi).core_reserved_blocks)) {
 			spin_unlock(&sbi->stat_lock);
 			return -EINVAL;
 		}
