--- original/fs/fat/Kconfig	2020-05-07 08:06:20.135204373 +0200
+++ changed/fs/fat/Kconfig	2019-06-03 13:39:53.000000000 +0200
@@ -98,3 +98,22 @@ config FAT_DEFAULT_IOCHARSET
 
 	  Enable any character sets you need in File Systems/Native Language
 	  Support.
+
+config FAT_VIRTUAL_XATTR
+	bool "Virtual xattr support for FAT filesystem"
+	default y
+	depends on FAT_FS
+	help
+	  Modification of FAT filesystem for virtual xattr
+
+config FAT_VIRTUAL_XATTR_SELINUX_LABEL
+	string "Default string for SELinux label"
+	depends on FAT_FS && FAT_VIRTUAL_XATTR
+	default "u:object_r:sdcard_external:s0"
+	help
+	  Set this to the default string for SELinux label.
+
+config FAT_SUPPORT_STLOG
+	bool "Enable storage log"
+	default y
+	depends on FAT_FS && PROC_STLOG
