--- original/fs/f2fs/data.c	2020-05-07 08:51:35.211254299 +0200
+++ changed/fs/f2fs/data.c	2019-06-03 13:39:53.000000000 +0200
@@ -170,6 +170,8 @@ static void f2fs_write_end_io(struct bio
 
 		if (unlikely(err)) {
 			set_bit(AS_EIO, &page->mapping->flags);
+			f2fs_bug_on(sbi, page->mapping == NODE_MAPPING(sbi) ||
+					 page->mapping == META_MAPPING(sbi));
 			if (type == F2FS_WB_CP_DATA)
 				f2fs_stop_checkpoint(sbi, true);
 		}
