--- original/fs/Kconfig	2020-05-07 08:31:26.177783722 +0200
+++ changed/fs/Kconfig	2019-06-03 13:39:53.000000000 +0200
@@ -8,11 +8,16 @@ menu "File systems"
 config DCACHE_WORD_ACCESS
        bool
 
+config INTERRUPTIBLE_SYNC
+	bool "Support interruptible sync for Samsung Mobile Device"
+	default y
+
 if BLOCK
 
 source "fs/ext2/Kconfig"
 source "fs/ext3/Kconfig"
 source "fs/ext4/Kconfig"
+source "fs/sdcardfs/Kconfig"
 
 config FS_XIP
 # execute in place
@@ -91,6 +96,7 @@ if BLOCK
 menu "DOS/FAT/NT Filesystems"
 
 source "fs/fat/Kconfig"
+source "fs/sdfat/Kconfig"
 source "fs/ntfs/Kconfig"
 
 endmenu
