--- original/fs/namei.c	2020-05-07 08:31:26.538780474 +0200
+++ changed/fs/namei.c	2019-06-03 13:39:53.000000000 +0200
@@ -3719,6 +3719,10 @@ retry:
 	if (error)
 		goto exit3;
 	error = vfs_rmdir2(nd.path.mnt, nd.path.dentry->d_inode, dentry);
+#ifdef CONFIG_PROC_DLOG
+	if (!error)
+		dlog_hook_rmdir(dentry, &nd.path);
+#endif
 exit3:
 	dput(dentry);
 exit2:
@@ -3847,6 +3851,10 @@ retry_deleg:
 		if (error)
 			goto exit2;
 		error = vfs_unlink2(nd.path.mnt, nd.path.dentry->d_inode, dentry, &delegated_inode);
+#ifdef CONFIG_PROC_DLOG
+		if (!error)
+			dlog_hook(dentry, inode, &nd.path);
+#endif
 exit2:
 		dput(dentry);
 	}
