--- original/fs/proc/Kconfig	2020-05-07 08:06:20.478201003 +0200
+++ changed/fs/proc/Kconfig	2019-06-03 13:39:53.000000000 +0200
@@ -71,3 +71,29 @@ config PROC_PAGE_MONITOR
 	  /proc/pid/smaps, /proc/pid/clear_refs, /proc/pid/pagemap,
 	  /proc/kpagecount, and /proc/kpageflags. Disabling these
           interfaces will reduce the size of the kernel by approximately 4kb.
+
+config PROC_AVC
+    bool "support (/proc/avc_msg)"
+    default n
+    help
+      support logging audit avc message
+
+config PROC_TASK_STATE_CHN_ORDER
+    bool "move Ngid to the last in task state"
+    default n
+    help
+      move Ngid to the last in task state for Chinese app
+
+
+config PROC_FSLOG
+	bool "Enable FS logs for improving filesystem debug"
+	depends on PROC_FS
+	select PROC_STLOG
+	select PROC_DLOG
+	default y
+
+config PROC_STLOG
+	bool "Enable Storage log"
+
+config PROC_DLOG
+	bool "Enable delete logs for improving filesystem debug"
