--- original/fs/pstore/Kconfig	2020-05-07 08:06:20.580200000 +0200
+++ changed/fs/pstore/Kconfig	2019-06-03 13:39:53.000000000 +0200
@@ -31,6 +31,21 @@ config PSTORE_PMSG
 
 	  If unsure, say N.
 
+config PSTORE_PMSG_SSPLOG
+	bool "Log user space messages for SS Platform log"
+	depends on PSTORE_PMSG
+	depends on SEC_DEBUG
+	select SEC_EVENT_LOG
+	help
+	  When the option is enabled, pmsg will save log at reserved
+	  dram area.
+
+	  If unsure, say N.
+
+config SEC_EVENT_LOG
+	bool "Enable Event log parser"
+	default n
+
 config PSTORE_FTRACE
 	bool "Persistent function tracer"
 	depends on PSTORE
