--- original/fs/quota/quota_v2.c	2020-05-07 08:06:20.664199175 +0200
+++ changed/fs/quota/quota_v2.c	2019-06-03 13:39:53.000000000 +0200
@@ -298,6 +298,11 @@ static int v2_free_file_info(struct supe
 	return 0;
 }
 
+static int v2_get_next_id(struct super_block *sb, struct kqid *qid)
+{
+	return qtree_get_next_id(sb_dqinfo(sb, qid->type)->dqi_priv, qid);
+}
+
 static const struct quota_format_ops v2_format_ops = {
 	.check_quota_file	= v2_check_quota_file,
 	.read_file_info		= v2_read_file_info,
@@ -306,6 +311,7 @@ static const struct quota_format_ops v2_
 	.read_dqblk		= v2_read_dquot,
 	.commit_dqblk		= v2_write_dquot,
 	.release_dqblk		= v2_release_dquot,
+	.get_next_id		= v2_get_next_id,
 };
 
 static struct quota_format_type v2r0_quota_format = {
