--- original/fs/sdcardfs/super.c	2020-05-07 08:31:26.663779349 +0200
+++ changed/fs/sdcardfs/super.c	2019-06-03 13:39:53.000000000 +0200
@@ -78,6 +78,11 @@ static int sdcardfs_statfs(struct dentry
 	err = vfs_statfs(&lower_path, buf);
 	sdcardfs_put_lower_path(dentry, &lower_path);
 
+	if (uid_eq(GLOBAL_ROOT_UID, current_fsuid()) ||
+			capable(CAP_SYS_RESOURCE) ||
+			in_group_p(AID_USE_ROOT_RESERVED))
+		goto out;
+
 	if (sbi->options.reserved_mb) {
 		/* Invalid statfs informations. */
 		if (buf->f_bsize == 0) {
@@ -96,7 +101,7 @@ static int sdcardfs_statfs(struct dentry
 		/* Make reserved blocks invisiable to media storage */
 		buf->f_bfree = buf->f_bavail;
 	}
-
+out:
 	/* set return buf to our f/s to avoid confusing user-level utils */
 	buf->f_type = SDCARDFS_SUPER_MAGIC;
 
@@ -144,7 +149,7 @@ static int sdcardfs_remount_fs2(struct v
 		pr_err("sdcardfs: remount flags 0x%x unsupported\n", *flags);
 		err = -EINVAL;
 	}
-	pr_info("Remount options were %s for vfsmnt %pK.\n", options, mnt);
+	pr_info("Remount options were %s.\n", options);
 	err = parse_options_remount(sb, options, *flags & ~MS_SILENT, mnt->data);
 
 
