--- original/fs/sdcardfs/dentry.c	2020-05-07 08:31:26.655779421 +0200
+++ changed/fs/sdcardfs/dentry.c	2019-06-03 13:39:53.000000000 +0200
@@ -180,11 +180,26 @@ static void sdcardfs_canonical_path(cons
 	sdcardfs_get_real_lower(path->dentry, actual_path);
 }
 
+/* P181109-04632 */
+static int sdcardfs_d_delete(const struct dentry *dentry)
+{
+	struct sdcardfs_dentry_info *info = SDCARDFS_D(dentry);
+	struct path *lower_path = &info->lower_path;
+	unsigned long lower_fs_magic = lower_path->mnt->mnt_sb->s_magic;
+
+	if (lower_fs_magic == EXT4_SUPER_MAGIC ||
+			lower_fs_magic == F2FS_SUPER_MAGIC)
+		return 0;
+
+	return 1;
+}	
+
 const struct dentry_operations sdcardfs_ci_dops = {
 	.d_revalidate	= sdcardfs_d_revalidate,
 	.d_release	= sdcardfs_d_release,
 	.d_hash	= sdcardfs_hash_ci,
 	.d_compare	= sdcardfs_cmp_ci,
+	.d_delete	= sdcardfs_d_delete,
 	.d_canonical_path = sdcardfs_canonical_path,
 };
 
