--- original/fs/super.c	2020-05-07 08:31:26.674779250 +0200
+++ changed/fs/super.c	2019-06-03 13:39:53.000000000 +0200
@@ -756,7 +756,14 @@ int do_remount_sb2(struct vfsmount *mnt,
 			     sb->s_type->name, retval);
 		}
 	}
+
+#ifdef CONFIG_FIVE
+	sb->s_flags = (sb->s_flags & ~MS_RMT_MASK) |
+				(flags & MS_RMT_MASK) | MS_I_VERSION;
+#else
 	sb->s_flags = (sb->s_flags & ~MS_RMT_MASK) | (flags & MS_RMT_MASK);
+#endif
+
 	/* Needs to be ordered wrt mnt_is_readonly() */
 	smp_wmb();
 	sb->s_readonly_remount = 0;
