--- original/lib/bug.c	2020-05-07 08:06:22.428181842 +0200
+++ changed/lib/bug.c	2019-06-03 13:39:53.000000000 +0200
@@ -45,6 +45,11 @@
 #include <linux/kernel.h>
 #include <linux/bug.h>
 #include <linux/sched.h>
+#include <linux/sec_debug.h>
+
+#ifdef CONFIG_USER_RESET_DEBUG
+#include <linux/user_reset/sec_debug_user_reset.h>
+#endif
 
 extern const struct bug_entry __start___bug_table[], __stop___bug_table[];
 
@@ -172,6 +177,28 @@ enum bug_trap_type report_bug(unsigned l
 		return BUG_TRAP_TYPE_WARN;
 	}
 
+#ifdef CONFIG_USER_RESET_DEBUG
+#define MAX_BUG_STRING_SIZE	56
+	if (file) {
+		unsigned int length;
+		char *new_file;
+		new_file = strstr(file, "kernel");
+		if (new_file == NULL) {
+			length = strlen(file);
+			if (length > MAX_BUG_STRING_SIZE)
+				sec_debug_store_bug_string("%s %u", &file[length - MAX_BUG_STRING_SIZE], line);
+			else
+				sec_debug_store_bug_string("%s %u", file, line);
+		} else {
+			length = strlen(new_file);
+			if (length > MAX_BUG_STRING_SIZE)
+				sec_debug_store_bug_string("%s %u", &new_file[length - MAX_BUG_STRING_SIZE], line);
+			else 
+				sec_debug_store_bug_string("%s %u", new_file, line);
+		}
+	}
+#endif
+
 	printk(KERN_DEFAULT "------------[ cut here ]------------\n");
 
 	if (file)
