--- original/mm/page_io.c	2020-05-07 08:06:22.568180466 +0200
+++ changed/mm/page_io.c	2019-06-03 13:39:53.000000000 +0200
@@ -63,11 +63,13 @@ void end_swap_bio_write(struct bio *bio,
 		 * Also clear PG_reclaim to avoid rotate_reclaimable_page()
 		 */
 		set_page_dirty(page);
+#ifndef CONFIG_VNSWAP
 		if (printk_timed_ratelimit(&error_time, ERROR_LOG_RATE_MS))
 			pr_info("Write-error on swap-device (%u:%u:%llu)\n",
 				imajor(bio->bi_bdev->bd_inode),
 				iminor(bio->bi_bdev->bd_inode),
 				(unsigned long long)bio->bi_iter.bi_sector);
+#endif
 		ClearPageReclaim(page);
 	}
 	end_page_writeback(page);
@@ -249,7 +251,13 @@ int swap_writepage(struct page *page, st
 		end_page_writeback(page);
 		goto out;
 	}
+#ifdef CONFIG_VNSWAP
+	set_page_dirty(page);
+	ClearPageReclaim(page);
+	unlock_page(page);
+#else
 	ret = __swap_writepage(page, wbc, end_swap_bio_write);
+#endif
 out:
 	return ret;
 }
