--- original/mm/process_vm_access.c	2020-05-07 08:06:22.572180427 +0200
+++ changed/mm/process_vm_access.c	2019-06-03 13:39:53.000000000 +0200
@@ -16,6 +16,7 @@
 #include <linux/ptrace.h>
 #include <linux/slab.h>
 #include <linux/syscalls.h>
+#include <linux/task_integrity.h>
 
 #ifdef CONFIG_COMPAT
 #include <linux/compat.h>
@@ -209,6 +210,10 @@ static ssize_t process_vm_rw_core(pid_t
 		goto put_task_struct;
 	}
 
+	rc = five_process_vm_rw(task, vm_write);
+	if (rc)
+		goto put_task_struct;
+
 	for (i = 0; i < riovcnt && iov_iter_count(iter) && !rc; i++)
 		rc = process_vm_rw_single_vec(
 			(unsigned long)rvec[i].iov_base, rvec[i].iov_len,
