--- original/mm/swap_state.c	2020-05-07 08:06:22.604180112 +0200
+++ changed/mm/swap_state.c	2019-06-03 13:39:54.000000000 +0200
@@ -393,6 +393,7 @@ struct page *read_swap_cache_async(swp_e
 	return found_page;
 }
 
+#ifdef CONFIG_SWAP_ENABLE_READAHEAD
 static unsigned long swapin_nr_pages(unsigned long offset)
 {
 	static unsigned long prev_offset;
@@ -436,6 +437,7 @@ static unsigned long swapin_nr_pages(uns
 
 	return pages;
 }
+#endif
 
 /**
  * swapin_readahead - swap in pages in hope we need them soon
@@ -459,6 +461,7 @@ static unsigned long swapin_nr_pages(uns
 struct page *swapin_readahead(swp_entry_t entry, gfp_t gfp_mask,
 			struct vm_area_struct *vma, unsigned long addr)
 {
+#ifdef CONFIG_SWAP_ENABLE_READAHEAD
 	struct page *page;
 	unsigned long entry_offset = swp_offset(entry);
 	unsigned long offset = entry_offset;
@@ -491,5 +494,6 @@ struct page *swapin_readahead(swp_entry_
 
 	lru_add_drain();	/* Push any new pages onto the LRU now */
 skip:
+#endif
 	return read_swap_cache_async(entry, gfp_mask, vma, addr);
 }
