--- original/drivers/spi/Kconfig	2020-05-07 08:06:17.704228260 +0200
+++ changed/drivers/spi/Kconfig	2019-06-03 13:39:52.000000000 +0200
@@ -642,6 +642,11 @@ config SPI_TLE62X0
 	  sysfs interface, with each line presented as a kind of GPIO
 	  exposing both switch control and diagnostic feedback.
 
+config SENSORS_FP_SPI_NUMBER
+	depends on SENSORS_FINGERPRINT
+	int "Fingerprint spi gpio number"
+	default 0
+
 #
 # Add new SPI protocol masters in alphabetical order above this line
 #
