--- original/drivers/of/fdt.c	2020-05-07 08:31:24.664797333 +0200
+++ changed/drivers/of/fdt.c	2019-06-03 13:39:51.000000000 +0200
@@ -575,6 +575,7 @@ void __init early_init_fdt_scan_reserved
 		if (!size)
 			break;
 		early_init_dt_reserve_memory_arch(base, size, 0);
+		record_memsize_reserved(NULL, base, size, 0, 0);
 	}
 
 	of_scan_flat_dt(__fdt_scan_reserved_mem, NULL);
@@ -978,6 +979,7 @@ int __init early_init_dt_scan_chosen(uns
 void __init __weak early_init_dt_add_memory_arch(u64 base, u64 size)
 {
 	const u64 phys_offset = __pa(PAGE_OFFSET);
+	u64 hole_s;
 
 	if (!PAGE_ALIGNED(base)) {
 		if (size < PAGE_SIZE - (base & ~PAGE_MASK)) {
@@ -1014,6 +1016,12 @@ void __init __weak early_init_dt_add_mem
 		base = phys_offset;
 	}
 	memblock_add(base, size);
+	hole_s = base % SZ_256M;
+	if (hole_s)
+		record_memsize_reserved(NULL, base - hole_s, hole_s, 1, 0);
+	hole_s = round_up(base + size, SZ_256M) - (base + size);
+	if (hole_s)
+		record_memsize_reserved(NULL, base + size, hole_s, 1, 0);
 }
 
 int __init __weak early_init_dt_reserve_memory_arch(phys_addr_t base,
