--- original/drivers/video/msm/mdss/mdss_dsi_cmd.h	2020-05-07 08:31:26.060784775 +0200
+++ changed/drivers/video/msm/mdss/mdss_dsi_cmd.h	2019-06-03 13:39:53.000000000 +0200
@@ -104,6 +104,11 @@ struct dsi_cmd_desc {
 #define CMD_REQ_LP_MODE 0x0010
 #define CMD_REQ_HS_MODE 0x0020
 
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+/* temp REQ to send packet broadcast */
+#define CMD_REQ_BROADCAST 0x0080
+#endif
+
 struct dcs_cmd_req {
 	struct dsi_cmd_desc *cmds;
 	int cmds_cnt;
