--- original/drivers/usb/misc/usb3503.c	2020-05-07 08:06:18.898216528 +0200
+++ changed/drivers/usb/misc/usb3503.c	2019-06-03 13:39:52.000000000 +0200
@@ -291,6 +291,8 @@ static int usb3503_probe(struct usb3503
 	if (gpio_is_valid(hub->gpio_reset)) {
 		err = devm_gpio_request_one(dev, hub->gpio_reset,
 				GPIOF_OUT_INIT_LOW, "usb3503 reset");
+		/* Datasheet defines a hardware reset to be at least 100us */
+		usleep_range(100, 10000);
 		if (err) {
 			dev_err(dev,
 				"unable to request GPIO %d as reset pin (%d)\n",
