--- original/drivers/platform/msm/msm_bus/msm_bus_arb_adhoc.c	2020-05-07 08:31:24.813795993 +0200
+++ changed/drivers/platform/msm/msm_bus/msm_bus_arb_adhoc.c	2019-06-03 13:39:52.000000000 +0200
@@ -1,4 +1,4 @@
-/* Copyright (c) 2014-2017, The Linux Foundation. All rights reserved.
+/* Copyright (c) 2014-2016, The Linux Foundation. All rights reserved.
  *
  * This program is Mree software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 and
@@ -887,7 +887,7 @@ static void unregister_client_adhoc(uint
 	}
 
 	curr = client->curr;
-	if ((curr < 0) || (curr >= pdata->num_usecases)) {
+	if (curr >= pdata->num_usecases) {
 		MSM_BUS_ERR("Invalid index Defaulting curr to 0");
 		curr = 0;
 	}
