--- original/drivers/usb/gadget/function/u_bam.c	2020-05-07 08:31:25.913786097 +0200
+++ changed/drivers/usb/gadget/function/u_bam.c	2019-06-03 13:39:52.000000000 +0200
@@ -1147,7 +1147,7 @@ static void gbam_notify(void *p, int eve
 		break;
 	case BAM_DMUX_TRANSMIT_SIZE:
 		d = &port->data_ch;
-		if (test_bit(BAM_CH_OPENED, &d->flags))
+		if (d && test_bit(BAM_CH_OPENED, &d->flags))
 			pr_warn("%s, BAM channel opened already", __func__);
 		bam_mux_rx_req_size = data;
 		pr_debug("%s rx_req_size: %lu", __func__, bam_mux_rx_req_size);
