--- original/drivers/pci/pci-sysfs.c	2020-05-07 08:51:34.967256689 +0200
+++ changed/drivers/pci/pci-sysfs.c	2019-06-03 13:39:51.000000000 +0200
@@ -539,7 +539,7 @@ static ssize_t driver_override_store(str
 		pdev->driver_override = NULL;
 	}
 	device_unlock(dev);
-
+	
 	kfree(old);
 
 	return count;
@@ -552,7 +552,7 @@ static ssize_t driver_override_show(stru
 	ssize_t len;
 
 	device_lock(dev);
-	len = snprintf(buf, PAGE_SIZE, "%s\n", pdev->driver_override);
+	len = sprintf(buf, "%s\n", pdev->driver_override);
 	device_unlock(dev);
 	return len;
 }
