--- original/drivers/video/msm/mdss/mdss_compat_utils.h	2020-05-07 08:31:26.053784837 +0200
+++ changed/drivers/video/msm/mdss/mdss_compat_utils.h	2019-06-03 13:39:53.000000000 +0200
@@ -304,15 +304,26 @@ struct mdp_qseed_cfg_data32 {
 	uint32_t block;
 	struct mdp_qseed_cfg32 qseed_data;
 };
-
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
 struct mdp_dither_cfg_data32 {
+	uint32_t version;
 	uint32_t block;
 	uint32_t flags;
+	uint32_t mode;
 	uint32_t g_y_depth;
 	uint32_t r_cr_depth;
 	uint32_t b_cb_depth;
+	compat_caddr_t cfg_payload;
 };
-
+#else
+struct mdp_dither_cfg_data32 {
+	uint32_t block;
+	uint32_t flags;
+	uint32_t g_y_depth;
+	uint32_t r_cr_depth;
+	uint32_t b_cb_depth;
+};
+#endif
 struct mdp_gamut_data_v1_7_32 {
 	uint32_t mode;
 	uint32_t tbl_size[MDP_GAMUT_TABLE_NUM_V1_7];
@@ -371,6 +382,9 @@ struct mdss_ad_init32 {
 	uint8_t logo_h;
 	uint32_t alpha;
 	uint32_t alpha_base;
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+	uint32_t al_thresh;
+#endif
 	uint32_t bl_lin_len;
 	uint32_t bl_att_len;
 	compat_caddr_t bl_lin;
