--- original/drivers/usb/gadget/udc/fsl_udc_core.c	2020-05-07 08:06:18.782217668 +0200
+++ changed/drivers/usb/gadget/udc/fsl_udc_core.c	2019-06-03 13:39:52.000000000 +0200
@@ -1305,7 +1305,7 @@ static void udc_reset_ep_queue(struct fs
 {
 	struct fsl_ep *ep = get_ep_by_pipe(udc, pipe);
 
-	if (ep->name)
+	if (ep->ep.name)
 		nuke(ep, -ESHUTDOWN);
 }
 
@@ -1693,7 +1693,7 @@ static void dtd_complete_irq(struct fsl_
 		curr_ep = get_ep_by_pipe(udc, i);
 
 		/* If the ep is configured */
-		if (curr_ep->name == NULL) {
+		if (!curr_ep->ep.name) {
 			WARNING("Invalid EP?");
 			continue;
 		}
