--- original/drivers/input/Makefile	2020-05-07 08:47:58.682375206 +0200
+++ changed/drivers/input/Makefile	2019-06-03 14:17:54.000000000 +0200
@@ -7,6 +7,11 @@
 obj-$(CONFIG_INPUT)		+= input-core.o
 input-core-y := input.o input-compat.o input-mt.o ff-core.o
 
+obj-$(CONFIG_SEC_DEBUG_TSP_LOG) += sec_tsp_log.o
+obj-$(CONFIG_INPUT_TOUCHSCREEN)	+= sec_cmd.o
+obj-$(CONFIG_INPUT_TOUCHSCREEN_TCLM)	+= sec_tclm.o
+obj-$(CONFIG_INPUT_TOUCHSCREEN_TCLMV2)	+= sec_tclm_v2.o
+
 obj-$(CONFIG_INPUT_FF_MEMLESS)	+= ff-memless.o
 obj-$(CONFIG_INPUT_POLLDEV)	+= input-polldev.o
 obj-$(CONFIG_INPUT_SPARSEKMAP)	+= sparse-keymap.o
@@ -29,3 +34,4 @@ obj-$(CONFIG_INPUT_KEYRESET)	+= keyreset
 obj-$(CONFIG_INPUT_KEYCOMBO)	+= keycombo.o
 obj-$(CONFIG_SENSORS_HALL)	+= hall_sensor.o
 obj-y	+= sensors/smi130/
+
