--- original/drivers/media/platform/msm/camera_v2/msm_buf_mgr/msm_generic_buf_mgr.c	2020-05-07 08:53:30.840121711 +0200
+++ changed/drivers/media/platform/msm/camera_v2/msm_buf_mgr/msm_generic_buf_mgr.c	2019-06-03 13:39:50.000000000 +0200
@@ -558,7 +558,6 @@ static long msm_buf_mngr_subdev_ioctl(st
 				}
 				k_ioctl.ioctl_ptr = (uintptr_t)&buf_info;
 			}
-
 			argp = &k_ioctl;
 			rc = msm_cam_buf_mgr_ops(cmd, argp);
 			}
