--- original/drivers/input/keyboard/Kconfig	2020-05-07 08:06:12.137282963 +0200
+++ changed/drivers/input/keyboard/Kconfig	2019-06-03 13:39:50.000000000 +0200
@@ -12,6 +12,30 @@ menuconfig INPUT_KEYBOARD
 
 if INPUT_KEYBOARD
 
+config KEYBOARD_ABOV_TOUCH_18XX
+	tristate "Abov FT18XX touchkey"
+	depends on I2C
+	help
+	  Say Y here if you want to use ABOV MC96FT18XX chip as touchkey.
+
+config KEYBOARD_ABOV_TOUCH_16XX
+	tristate "Abov FT16XX touchkey"
+	depends on I2C
+	help
+	  Say Y here if you want to use ABOV MC96FT16XX chip as touchkey.
+
+config KEYBOARD_ABOV_TOUCH_T316
+	tristate "Abov A96T316XX touchkey"
+	depends on I2C
+	help
+	  Say Y here if you want to use ABOV A96T316XX chip as touchkey.
+
+config KEYBOARD_TC3XXK
+	tristate "Coreriver TC3xxK touchkey series"
+	depends on I2C
+	help
+	 Say Y if you want use Coreriver TC3xxK chip series as touchkey.
+
 config KEYBOARD_ADP5520
 	tristate "Keypad Support for ADP5520 PMIC"
 	depends on PMIC_ADP5520
@@ -354,6 +378,16 @@ config KEYBOARD_LPC32XX
 
 	  To compile this driver as a module, choose M here: the
 	  module will be called lpc32xx-keys.
+	  
+config KEYBOARD_TC300K
+	tristate "Coreriver TC300K touchkey"
+	depends on I2C
+	help
+	  Say Y here if you want to use the keypad controller on
+	  TC300K I/O expander.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called tc300k Touchkey.
 
 config KEYBOARD_MAPLE
 	tristate "Maple bus keyboard"
@@ -567,6 +601,16 @@ config KEYBOARD_STMPE
 	  To compile this driver as a module, choose M here: the module will be
 	  called stmpe-keypad.
 
+config KEYBOARD_STMPE1801
+	tristate "STMPE keypad support"
+	select INPUT_MATRIXKMAP
+	help
+	  Say Y here if you want to use the keypad controller on STMPE I/O
+	  expanders.
+
+	  To compile this driver as a module, choose M here: the module will be
+	  called stmpe-keypad.
+
 config KEYBOARD_DAVINCI
 	tristate "TI DaVinci Key Scan"
 	depends on ARCH_DAVINCI_DM365
@@ -675,4 +719,44 @@ config KEYBOARD_CAP1106
 	  To compile this driver as a module, choose M here: the
 	  module will be called cap1106.
 
+config SENSORS_HALL
+	bool "HALL"
+	depends on KEYBOARD_GPIO
+	default n
+	help
+	Say Y here to enable the HALL
+
+config CERTIFY_HALL
+	tristate "Enable CERTIFY HALL"
+	default n
+	help
+	  Enable Certify Hall Feature
+	  
+config SENSORS_HALL_FOLDER
+	bool "HALL for FOLDER"
+	depends on SENSORS_HALL
+	default n
+	help
+	  Say Y here to use HALL IC for FOLDER.
+
+config TOUCHKEY_GRIP
+	tristate "touchkey grip for sar"
+	depends on I2C
+	help
+	 Say Y if you want use touchkey grip.
+	 
+config TOUCHKEY_LED
+	tristate "touchkey backlight led"
+	depends on I2C
+	help
+	 Say Y if you want use touchkey backlight led.
+	 
+config TOUCHKEY_LIGHT_EFS
+	tristate "touchkey led light with efs"
+	depends on I2C && TOUCHKEY_LED
+	help
+	  Say Y here if you want to use touchkey light efs.
+
+
+source "drivers/input/keyboard/melfas_mip4/Kconfig"
 endif
