--- original/drivers/media/platform/msm/camera_v2/sensor/flash/msm_led_flash.c	2020-05-07 08:31:23.682806167 +0200
+++ changed/drivers/media/platform/msm/camera_v2/sensor/flash/msm_led_flash.c	2019-06-03 13:39:50.000000000 +0200
@@ -37,13 +37,25 @@ static long msm_led_flash_subdev_ioctl(s
 	}
 	switch (cmd) {
 	case VIDIOC_MSM_SENSOR_GET_SUBDEV_ID:
-		return fctrl->func_tbl->flash_get_subdev_id(fctrl, argp);
+                if (fctrl->func_tbl->flash_get_subdev_id)
+			return fctrl->func_tbl->flash_get_subdev_id(fctrl, argp);
+		else
+			pr_err("NULL flash_get_subdev function\n");
+			return -EINVAL;
 	case VIDIOC_MSM_FLASH_LED_DATA_CFG:
-		return fctrl->func_tbl->flash_led_config(fctrl, argp);
+                if (fctrl->func_tbl->flash_led_config)
+			return fctrl->func_tbl->flash_led_config(fctrl, argp);
+		else
+			pr_err("NULL flash_led_config function\n");
+			return -EINVAL;
 	case MSM_SD_NOTIFY_FREEZE:
 		return 0;
 	case MSM_SD_SHUTDOWN:
-		return fctrl->func_tbl->flash_led_release(fctrl);
+                if (fctrl->func_tbl->flash_led_release)
+			return fctrl->func_tbl->flash_led_release(fctrl);
+		else
+			pr_err("NULL flash_led_release function\n");
+			return -EINVAL;
 	default:
 		pr_err_ratelimited("invalid cmd %d\n", cmd);
 		return -ENOIOCTLCMD;
