--- original/drivers/usb/gadget/function/f_gsi.c	2020-05-07 08:31:25.888786322 +0200
+++ changed/drivers/usb/gadget/function/f_gsi.c	2019-06-03 13:39:52.000000000 +0200
@@ -1459,14 +1459,23 @@ static ssize_t gsi_ctrl_dev_write(struct
 	struct gsi_ctrl_port *c_port = container_of(fp->private_data,
 						struct gsi_ctrl_port,
 						ctrl_device);
-	struct f_gsi *gsi = c_port_to_gsi(c_port);
-	struct usb_request *req = c_port->notify_req;
+	struct f_gsi *gsi;
+	struct usb_request *req;
+
+	if (!c_port) {
+		log_event_err("%s: c_port %p",
+			__func__, c_port);
+		return -ENODEV;
+	}
+
+	gsi = c_port_to_gsi(c_port);
+	req = c_port->notify_req;
 
 	log_event_dbg("Enter %zu", count);
 
-	if (!c_port || !req || !req->buf) {
-		log_event_err("%s: c_port %pK req %pK req->buf %pK",
-			__func__, c_port, req, req ? req->buf : req);
+	if (!req || !req->buf) {
+		log_event_err("%s: req %p req->buf %p",
+			__func__, req, req ? req->buf : req);
 		return -ENODEV;
 	}
 
@@ -3154,8 +3163,7 @@ static int gsi_bind(struct usb_configura
 
 		gsi_get_ether_addr(gsi_host_addr,
 				   gsi->d_port.ipa_init_params.host_ethaddr);
-
-		log_event_dbg("setting host_ethaddr=%pM, device_ethaddr = %pM",
+		log_event_dbg("setting host_ethaddr=%pKM, device_ethaddr = %pKM",
 				gsi->d_port.ipa_init_params.host_ethaddr,
 				gsi->d_port.ipa_init_params.device_ethaddr);
 
