--- original/drivers/net/wireless/Kconfig	2020-05-07 08:31:24.484798952 +0200
+++ changed/drivers/net/wireless/Kconfig	2019-06-03 13:39:51.000000000 +0200
@@ -285,6 +285,10 @@ config WCNSS_CORE_PRONTO
 	---help---
 	  Pronto Support for the Qualcomm WCNSS triple play connectivity subsystem
 
+config WLAN_USE_OLD_NV
+	int "use WCNSS_qcom_wlan_nv_old.bin for revision detecion"
+	default 0
+
 config WCNSS_REGISTER_DUMP_ON_BITE
 	bool "Enable/disable WCNSS register dump when there is a WCNSS bite"
 	depends on WCNSS_CORE_PRONTO
@@ -346,5 +350,6 @@ source "drivers/net/wireless/rsi/Kconfig
 source "drivers/net/wireless/cnss/Kconfig"
 source "drivers/net/wireless/cnss2/Kconfig"
 source "drivers/net/wireless/cnss_utils/Kconfig"
+source "drivers/net/wireless/prima/Kconfig"
 
 endif # WLAN
