--- original/drivers/usb/dwc3/dwc3-keystone.c	2020-05-07 08:06:18.706218415 +0200
+++ changed/drivers/usb/dwc3/dwc3-keystone.c	2019-06-03 13:39:52.000000000 +0200
@@ -117,6 +117,10 @@ static int kdwc3_probe(struct platform_d
 	dev->dma_mask = &kdwc3_dma_mask;
 
 	kdwc->clk = devm_clk_get(kdwc->dev, "usb");
+	if (IS_ERR(kdwc->clk)) {
+		dev_err(kdwc->dev, "unable to get usb clock\n");
+		return PTR_ERR(kdwc->clk);
+	}
 
 	error = clk_prepare_enable(kdwc->clk);
 	if (error < 0) {
