--- original/drivers/usb/gadget/ci13xxx_udc.h	2020-05-07 08:06:18.722218257 +0200
+++ changed/drivers/usb/gadget/ci13xxx_udc.h	2019-06-03 13:39:52.000000000 +0200
@@ -16,6 +16,16 @@
 #ifndef _CI13XXX_h_
 #define _CI13XXX_h_
 
+#ifdef CONFIG_USB_CHARGING_EVENT
+#if defined(CONFIG_BATTERY_SAMSUNG_V2)
+#include "../../battery_v2/include/sec_charging_common.h"
+#elif defined(CONFIG_BATTERY_SAMSUNG_V2_LEGACY)
+#include "../../battery_v2_legacy/include/sec_charging_common.h"
+#else
+#include <linux/battery/sec_charging_common.h>
+#endif
+#endif
+
 /******************************************************************************
  * DEFINE
  *****************************************************************************/
@@ -181,6 +191,11 @@ struct ci13xxx {
 	bool                      skip_flush; /* skip flushing remaining EP
 						upon flush timeout for the
 						first EP. */
+
+#if IS_ENABLED(CONFIG_USB_CHARGING_EVENT)
+	struct work_struct      set_vbus_current_work;
+	int			vbus_current;
+#endif
 };
 
 /******************************************************************************
