--- original/drivers/video/msm/mdss/Makefile	2020-05-07 08:31:26.043784928 +0200
+++ changed/drivers/video/msm/mdss/Makefile	2019-06-03 13:39:53.000000000 +0200
@@ -56,6 +56,7 @@ obj-$(CONFIG_FB_MSM_MDSS) += mdss_dba_ut
 obj-$(CONFIG_FB_MSM_MDSS_EDP_PANEL) += mdss_edp.o
 obj-$(CONFIG_FB_MSM_MDSS_EDP_PANEL) += mdss_edp_aux.o
 
+#obj-$(CONFIG_FB_MSM_MDSS) += mdss_io_util.o
 obj-$(CONFIG_FB_MSM_MDSS_HDMI_PANEL) += mdss_hdmi_tx.o
 obj-$(CONFIG_FB_MSM_MDSS_HDMI_PANEL) += mdss_hdmi_panel.o
 obj-$(CONFIG_FB_MSM_MDSS_HDMI_PANEL) += mdss_hdmi_hdcp.o
@@ -74,3 +75,6 @@ obj-$(CONFIG_FB_MSM_QPIC_ILI_QVGA_PANEL)
 
 obj-$(CONFIG_FB_MSM_MDSS) += mdss_fb.o mdss_util.o
 obj-$(CONFIG_COMPAT) += mdss_compat_utils.o
+
+# Samsung lcd common driver
+obj-$(CONFIG_FB_MSM_MDSS_SAMSUNG) += samsung/
