--- original/drivers/soc/qcom/smem.c	2020-05-07 08:06:17.677228526 +0200
+++ changed/drivers/soc/qcom/smem.c	2019-06-03 13:39:52.000000000 +0200
@@ -1312,6 +1312,15 @@ smem_targ_info_done:
 	}
 
 	smem_ram_base = ioremap_nocache(smem_ram_phys, smem_ram_size);
+#ifdef CONFIG_SEC_DEBUG_SUMMARY
+#ifdef CONFIG_ARM64
+	pr_info("%s: smem_ram_base=0x%lx  smem_ram_phys=0x%lx smem_ram_size=0x%lx \n",
+                        __func__,(long unsigned int)smem_ram_base,(long unsigned int)smem_ram_phys,(long unsigned int)smem_ram_size);
+#else
+	pr_info("%s: smem_ram_base=0x%x  smem_ram_phys=0x%x smem_ram_size=0x%x \n",
+                        __func__,(unsigned int)smem_ram_base,smem_ram_phys,smem_ram_size);
+#endif
+#endif
 
 	if (!smem_ram_base) {
 		LOG_ERR("%s: ioremap_nocache() of addr:%pa size: %pa\n",
