--- original/drivers/media/platform/msm/ais/isp/msm_isp47.c	2020-05-07 08:31:23.576807121 +0200
+++ changed/drivers/media/platform/msm/ais/isp/msm_isp47.c	2019-06-03 13:39:50.000000000 +0200
@@ -699,12 +699,6 @@ void msm_vfe47_reg_update(struct vfe_dev
 		vfe_dev->reg_update_requested;
 	if ((vfe_dev->is_split && vfe_dev->pdev->id == ISP_VFE1) &&
 		((frame_src == VFE_PIX_0) || (frame_src == VFE_SRC_MAX))) {
-		if (!vfe_dev->common_data->dual_vfe_res->vfe_base[ISP_VFE0]) {
-			pr_err("%s vfe_base for ISP_VFE0 is NULL\n", __func__);
-			spin_unlock_irqrestore(&vfe_dev->reg_update_lock,
-						flags);
-			return;
-		}
 		msm_camera_io_w_mb(update_mask,
 			vfe_dev->common_data->dual_vfe_res->
 			vfe_base[ISP_VFE0] + 0x4AC);
