--- original/drivers/media/v4l2-core/v4l2-compat-ioctl32.c	2020-05-07 08:31:23.818804944 +0200
+++ changed/drivers/media/v4l2-core/v4l2-compat-ioctl32.c	2019-06-03 13:39:50.000000000 +0200
@@ -29,7 +29,12 @@
 									\
 	get_user(__assign_tmp, from) || put_user(__assign_tmp, to);	\
 })
-
+#define convert_in_user(srcptr, dstptr)            \
+({                            \
+    typeof(*srcptr) val;                \
+                            \
+    get_user(val, srcptr) || put_user(val, dstptr);    \
+})
 static long native_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
 {
 	long ret = -ENOIOCTLCMD;
@@ -504,6 +509,14 @@ static int get_v4l2_buffer32(struct v4l2
 		    assign_in_user(&kp->timestamp.tv_usec,
 				   &up->timestamp.tv_usec))
 			return -EFAULT;
+    if (type == V4L2_BUF_TYPE_PRIVATE) {
+        compat_long_t tmp;
+
+        if (get_user(tmp, &up->m.userptr) ||
+                put_user((unsigned long) compat_ptr(tmp),
+                    &kp->m.userptr))
+            return -EFAULT;
+    }
 
 	if (V4L2_TYPE_IS_MULTIPLANAR(type)) {
 		u32 num_planes = length;
@@ -602,6 +615,9 @@ static int put_v4l2_buffer32(struct v4l2
 	    get_user(length, &kp->length) ||
 	    put_user(length, &up->length))
 		return -EFAULT;
+    if(type == V4L2_BUF_TYPE_PRIVATE)
+        if (convert_in_user(&kp->m.userptr, &up->m.userptr))
+            return -EFAULT;
 
 	if (V4L2_TYPE_IS_MULTIPLANAR(type)) {
 		u32 num_planes = length;
