--- original/drivers/thermal/msm-tsens.c	2020-05-07 08:31:25.702787996 +0200
+++ changed/drivers/thermal/msm-tsens.c	2019-06-03 13:39:52.000000000 +0200
@@ -5877,6 +5877,7 @@ static int tsens_thermal_zone_register(s
 	for (i = 0; i < tmdev->tsens_num_sensor; i++) {
 		char name[18];
 		if ((!strcmp(id->compatible, "qcom,mdm9640-tsens")) ||
+			(!strcmp(id->compatible, "qcom,msm8953-tsens")) ||
 			(!strcmp(id->compatible, "qcom,mdm9640v2-tsens")))
 			snprintf(name, sizeof(name), "tsens_tz_sensor%d",
 					tmdev->sensor[i].sensor_hw_num);
