--- original/drivers/video/msm/mdss/mdss_dsi_status.c	2020-05-07 08:31:26.069784694 +0200
+++ changed/drivers/video/msm/mdss/mdss_dsi_status.c	2019-06-03 13:39:53.000000000 +0200
@@ -165,6 +165,10 @@ static int fb_event_callback(struct noti
 		return NOTIFY_DONE;
 	}
 
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+	ctrl_pdata->panel_data.event_handler(&ctrl_pdata->panel_data, MDSS_SAMSUNG_EVENT_FB_EVENT_CALLBACK, &interval);
+#endif
+
 	pdata->mfd = evdata->info->par;
 	if (event == FB_EVENT_BLANK) {
 		int *blank = evdata->data;
@@ -174,15 +178,28 @@ static int fb_event_callback(struct noti
 
 		switch (*blank) {
 		case FB_BLANK_UNBLANK:
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+			if (!ctrl_pdata->status_mode == ESD_REG_IRQ)
+#endif
 			schedule_delayed_work(&pdata->check_status,
 				msecs_to_jiffies(interval));
 			break;
 		case FB_BLANK_VSYNC_SUSPEND:
 		case FB_BLANK_NORMAL:
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+			if (ctrl_pdata->status_mode == ESD_REG_IRQ)
+				cancel_work_sync(&pdata->check_status.work);
+			else
+#endif
 			pr_debug("%s : ESD thread running\n", __func__);
 			break;
 		case FB_BLANK_POWERDOWN:
 		case FB_BLANK_HSYNC_SUSPEND:
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+			if (ctrl_pdata->status_mode == ESD_REG_IRQ)
+				cancel_work_sync(&pdata->check_status.work);
+			else
+#endif
 			cancel_delayed_work(&pdata->check_status);
 			break;
 		default:
