--- original/drivers/clk/msm/clock-debug.c	2020-05-07 08:06:09.684307066 +0200
+++ changed/drivers/clk/msm/clock-debug.c	2019-06-03 13:39:49.000000000 +0200
@@ -34,7 +34,11 @@ static LIST_HEAD(clk_list);
 static DEFINE_MUTEX(clk_list_lock);
 
 static struct dentry *debugfs_base;
+#ifdef CONFIG_SEC_PM_DEBUG
+static u32 debug_suspend=1;
+#else
 static u32 debug_suspend;
+#endif
 
 static int clock_debug_rate_set(void *data, u64 val)
 {
