--- original/drivers/base/dma-removed.c	2020-05-07 08:06:09.290310938 +0200
+++ changed/drivers/base/dma-removed.c	2019-06-03 13:39:49.000000000 +0200
@@ -205,6 +205,7 @@ static void removed_region_fixup(struct
 	/* return remaining area to system */
 	fixup_size = (dma_mem->nr_pages - index) * PAGE_SIZE;
 	free_bootmem_late(dma_mem->base + index * PAGE_SIZE, fixup_size);
+	free_memsize_reserved(dma_mem->base + index * PAGE_SIZE, fixup_size);
 
 	/*
 	 * release freed resource region so as to show up under iomem resource
