--- original/drivers/staging/android/Kconfig	2020-05-07 08:31:25.441790343 +0200
+++ changed/drivers/staging/android/Kconfig	2019-06-03 13:39:52.000000000 +0200
@@ -106,6 +106,38 @@ config ANDROID_LOW_MEMORY_KILLER_AUTODET
 	  /sys/module/lowmemorykiller/parameters/adj and convert them
 	  to oom_score_adj values.
 
+config SEC_OOM_KILLER
+	bool "Android OOM Killer"
+	default n
+	help
+	 klaatu@samsung
+	 This enables 'Android-Style oom-killer'
+	 The way to select victim by oom-killer provided by
+	 linux kernel is totally different from android policy.
+	 Hence, it makes more sense that we select the oom victim
+	 as android does when LMK is invoked.
+
+config SEC_DEBUG_LMK_MEMINFO
+	bool "Show Meminfo when lmk invoked"
+	default n
+	help
+	  Show memory information when lmk kills process
+
+config SEC_DEBUG_LMK_COUNT_INFO
+	bool "Record LMK execution count"
+	default n
+	help
+	  Show LMK execution count information when lmk invoked
+
+config ANDROID_INTF_ALARM_DEV
+	bool "Android alarm driver"
+	depends on RTC_CLASS
+	default n
+	---help---
+	  Provides non-wakeup and rtc backed wakeup alarms based on rtc or
+	  elapsed realtime, and a non-wakeup alarm on the monotonic clock.
+	  Also exports the alarm interface to user-space.
+
 config SYNC
 	bool "Synchronization framework"
 	default n
