--- original/drivers/firmware/qcom/tz_log.c	2020-05-07 08:31:21.509825717 +0200
+++ changed/drivers/firmware/qcom/tz_log.c	2019-06-03 13:39:49.000000000 +0200
@@ -333,7 +333,7 @@ static int _disp_tz_general_stats(void)
 {
 	int len = 0;
 
-	len += snprintf(tzdbg.disp_buf + len, debug_rw_buf_size - 1,
+	len += scnprintf(tzdbg.disp_buf + len, debug_rw_buf_size - 1,
 			"   Version        : 0x%x\n"
 			"   Magic Number   : 0x%x\n"
 			"   Number of CPU  : %d\n",
@@ -358,7 +358,7 @@ static int _disp_tz_vmid_stats(void)
 
 	for (i = 0; i < num_vmid; i++) {
 		if (ptr->vmid < 0xFF) {
-			len += snprintf(tzdbg.disp_buf + len,
+			len += scnprintf(tzdbg.disp_buf + len,
 				(debug_rw_buf_size - 1) - len,
 				"   0x%x        %s\n",
 				(uint32_t)ptr->vmid, (uint8_t *)ptr->desc);
@@ -393,7 +393,7 @@ static int _disp_tz_boot_stats(void)
 
 	for (i = 0; i < tzdbg.diag_buf->cpu_count; i++) {
 		if (tzdbg.tz_version >= QSEE_VERSION_TZ_3_X) {
-			len += snprintf(tzdbg.disp_buf + len,
+			len += scnprintf(tzdbg.disp_buf + len,
 					(debug_rw_buf_size - 1) - len,
 					"  CPU #: %d\n"
 					"     Warmboot jump address : 0x%llx\n"
@@ -420,7 +420,7 @@ static int _disp_tz_boot_stats(void)
 			}
 			ptr_64++;
 		} else {
-			len += snprintf(tzdbg.disp_buf + len,
+			len += scnprintf(tzdbg.disp_buf + len,
 					(debug_rw_buf_size - 1) - len,
 					"  CPU #: %d\n"
 					"     Warmboot jump address     : 0x%x\n"
@@ -456,7 +456,7 @@ static int _disp_tz_reset_stats(void)
 					tzdbg.diag_buf->reset_info_off);
 
 	for (i = 0; i < tzdbg.diag_buf->cpu_count; i++) {
-		len += snprintf(tzdbg.disp_buf + len,
+		len += scnprintf(tzdbg.disp_buf + len,
 				(debug_rw_buf_size - 1) - len,
 				"  CPU #: %d\n"
 				"     Reset Type (reason)       : 0x%x\n"
@@ -496,7 +496,7 @@ static int _disp_tz_interrupt_stats(void
 	if (tzdbg.tz_version < QSEE_VERSION_TZ_4_X) {
 		for (i = 0; i < (*num_int); i++) {
 			tzdbg_ptr = (struct tzdbg_int_t *)ptr;
-			len += snprintf(tzdbg.disp_buf + len,
+			len += scnprintf(tzdbg.disp_buf + len,
 				(debug_rw_buf_size - 1) - len,
 				"     Interrupt Number          : 0x%x\n"
 				"     Type of Interrupt         : 0x%x\n"
@@ -505,13 +505,13 @@ static int _disp_tz_interrupt_stats(void
 				(uint32_t)tzdbg_ptr->int_info,
 				(uint8_t *)tzdbg_ptr->int_desc);
 			for (j = 0; j < tzdbg.diag_buf->cpu_count; j++) {
-				len += snprintf(tzdbg.disp_buf + len,
+				len += scnprintf(tzdbg.disp_buf + len,
 				(debug_rw_buf_size - 1) - len,
 				"     int_count on CPU # %d      : %u\n",
 				(uint32_t)j,
 				(uint32_t)tzdbg_ptr->int_count[j]);
 			}
-			len += snprintf(tzdbg.disp_buf + len,
+			len += scnprintf(tzdbg.disp_buf + len,
 					debug_rw_buf_size - 1, "\n");
 
 			if (len > (debug_rw_buf_size - 1)) {
@@ -524,7 +524,7 @@ static int _disp_tz_interrupt_stats(void
 	} else {
 		for (i = 0; i < (*num_int); i++) {
 			tzdbg_ptr_tz40 = (struct tzdbg_int_t_tz40 *)ptr;
-			len += snprintf(tzdbg.disp_buf + len,
+			len += scnprintf(tzdbg.disp_buf + len,
 				(debug_rw_buf_size - 1) - len,
 				"     Interrupt Number          : 0x%x\n"
 				"     Type of Interrupt         : 0x%x\n"
@@ -533,13 +533,13 @@ static int _disp_tz_interrupt_stats(void
 				(uint32_t)tzdbg_ptr_tz40->int_info,
 				(uint8_t *)tzdbg_ptr_tz40->int_desc);
 			for (j = 0; j < tzdbg.diag_buf->cpu_count; j++) {
-				len += snprintf(tzdbg.disp_buf + len,
+				len += scnprintf(tzdbg.disp_buf + len,
 				(debug_rw_buf_size - 1) - len,
 				"     int_count on CPU # %d      : %u\n",
 				(uint32_t)j,
 				(uint32_t)tzdbg_ptr_tz40->int_count[j]);
 			}
-			len += snprintf(tzdbg.disp_buf + len,
+			len += scnprintf(tzdbg.disp_buf + len,
 					debug_rw_buf_size - 1, "\n");
 
 			if (len > (debug_rw_buf_size - 1)) {
@@ -562,7 +562,7 @@ static int _disp_tz_log_stats_legacy(voi
 
 	ptr = (unsigned char *)tzdbg.diag_buf +
 					tzdbg.diag_buf->ring_off;
-	len += snprintf(tzdbg.disp_buf, (debug_rw_buf_size - 1) - len,
+	len += scnprintf(tzdbg.disp_buf, (debug_rw_buf_size - 1) - len,
 							"%s\n", ptr);
 
 	tzdbg.stat[TZDBG_LOG].data = tzdbg.disp_buf;
@@ -579,6 +579,7 @@ static int _disp_log_stats(struct tzdbg_
 	int max_len;
 	int len = 0;
 	int i = 0;
+	int retry = 2;
 
 	wrap_start = log_start->wrap;
 	wrap_end = log->log_pos.wrap;
@@ -612,7 +613,9 @@ static int _disp_log_stats(struct tzdbg_
 			/* Some event woke us up, so let's quit */
 			return 0;
 		}
-
+		retry--;
+        if (!retry)
+            return 0;
 		if (buf_idx == TZDBG_LOG)
 			memcpy_fromio((void *)tzdbg.diag_buf, tzdbg.virt_iobase,
 						debug_rw_buf_size);
@@ -750,7 +753,7 @@ static int _disp_hyp_general_stats(size_
 	int i;
 	struct hypdbg_boot_info_t *ptr = NULL;
 
-	len += snprintf((unsigned char *)tzdbg.disp_buf + len,
+	len += scnprintf((unsigned char *)tzdbg.disp_buf + len,
 			tzdbg.hyp_debug_rw_buf_size - 1,
 			"   Magic Number    : 0x%x\n"
 			"   CPU Count       : 0x%x\n"
@@ -761,7 +764,7 @@ static int _disp_hyp_general_stats(size_
 
 	ptr = tzdbg.hyp_diag_buf->boot_info;
 	for (i = 0; i < tzdbg.hyp_diag_buf->cpu_count; i++) {
-		len += snprintf((unsigned char *)tzdbg.disp_buf + len,
+		len += scnprintf((unsigned char *)tzdbg.disp_buf + len,
 				(tzdbg.hyp_debug_rw_buf_size - 1) - len,
 				"  CPU #: %d\n"
 				"     Warmboot entry CPU counter: 0x%x\n"
