--- original/drivers/gpio/Makefile	2020-05-07 11:02:09.579858150 +0200
+++ changed/drivers/gpio/Makefile	2019-06-03 13:39:49.000000000 +0200
@@ -108,3 +108,12 @@ obj-$(CONFIG_GPIO_XTENSA)	+= gpio-xtensa
 obj-$(CONFIG_GPIO_ZEVIO)	+= gpio-zevio.o
 obj-$(CONFIG_GPIO_ZYNQ)		+= gpio-zynq.o
 obj-$(CONFIG_MSM_SMP2P)         += gpio-msm-smp2p.o
+obj-$(CONFIG_GPIO_PCAL6416A)	+= gpio-pcal6416a.o
+
+# Debug and verify gpio configurations.
+obj-$(CONFIG_SEC_GPIO_DVS)      += secgpio_dvs.o
+obj-$(CONFIG_SEC_PM_DEBUG)      += sec-pinmux.o
+
+ifeq ($(CONFIG_SEC_FACTORY),y)
+    obj-$(CONFIG_SEC_CONNECTOR_DETECTOR)      += sec-detect-conn.o
+endif
