--- original/drivers/input/touchscreen/Makefile	2020-05-07 08:31:23.334809299 +0200
+++ changed/drivers/input/touchscreen/Makefile	2019-06-03 13:39:50.000000000 +0200
@@ -86,9 +86,19 @@ obj-$(CONFIG_TOUCHSCREEN_WM97XX_ZYLONITE
 obj-$(CONFIG_TOUCHSCREEN_W90X900)	+= w90p910_ts.o
 obj-$(CONFIG_TOUCHSCREEN_TPS6507X)	+= tps6507x-ts.o
 obj-$(CONFIG_TOUCHSCREEN_ZFORCE)	+= zforce_ts.o
+obj-$(CONFIG_TOUCHSCREEN_MSTAR21XX)	+= msg21xx_ts.o
 obj-$(CONFIG_TOUCHSCREEN_SYNAPTICS_I2C_RMI4)		+= synaptics_i2c_rmi4.o
 obj-$(CONFIG_TOUCHSCREEN_SYNAPTICS_DSX_RMI4_DEV)	+= synaptics_rmi_dev.o
 obj-$(CONFIG_TOUCHSCREEN_SYNAPTICS_DSX_FW_UPDATE) 	+= synaptics_fw_update.o
 obj-$(CONFIG_TOUCHSCREEN_GT9XX)		+= gt9xx/
+obj-$(CONFIG_TOUCHSCREEN_GT9XX)		+= gt9xx/
 obj-$(CONFIG_TOUCHSCREEN_HIMAX_CHIPSET)	+= hxchipset/
 obj-$(CONFIG_TOUCHSCREEN_FTS)		+= focaltech_touch/
+obj-$(CONFIG_TOUCHSCREEN_IST30XXH)	+= imagis_30xxh/
+obj-$(CONFIG_TOUCHSCREEN_SEC_TS)	+= sec_ts/
+obj-$(CONFIG_TOUCHSCREEN_SEC_INCELL_TS) += sec_incell_ts/
+obj-$(CONFIG_TOUCHSCREEN_FTS7)		+= stm/fts7/
+obj-$(CONFIG_TOUCHSCREEN_FTS8)		+= stm/fts8/
+obj-$(CONFIG_TOUCHSCREEN_FTS1BA90A)	+= stm/fts1ba90a/
+obj-$(CONFIG_TOUCHSCREEN_SYNAPTICS_INCELL)	+= synaptics_incell/
+obj-$(CONFIG_TOUCHSCREEN_ZINITIX_ZT75XX_TCLM) += zinitix/
