--- original/drivers/media/platform/msm/camera_v2/sensor/msm_sensor_init.h	2020-05-07 08:06:13.155272960 +0200
+++ changed/drivers/media/platform/msm/camera_v2/sensor/msm_sensor_init.h	2019-06-03 13:39:50.000000000 +0200
@@ -14,6 +14,9 @@
 #define MSM_SENSOR_INIT_H
 
 #include "msm_sensor.h"
+#include "msm_eeprom.h"
+
+#define CAMERA_INFO_MAXSIZE 150
 
 struct msm_sensor_init_t {
 	struct mutex imutex;
