--- original/drivers/video/msm/mdss/mdss_debug.h	2020-05-07 08:31:26.055784820 +0200
+++ changed/drivers/video/msm/mdss/mdss_debug.h	2019-06-03 13:39:53.000000000 +0200
@@ -104,12 +104,24 @@ struct mdss_debug_base {
 	u32 *reg_dump; /* address for the mem dump if no ranges used */
 };
 
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+struct debug_log {
+	struct dentry *xlog;
+	u32 xlog_enable;
+	u32 panic_on_err;
+	u32 enable_reg_dump;
+};
+#endif
+
 struct mdss_debug_data {
 	struct dentry *root;
 	struct dentry *perf;
 	struct dentry *bordercolor;
 	struct dentry *postproc;
 	struct list_head base_list;
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+	struct debug_log logd;
+#endif
 };
 
 struct dump_offset {
