--- original/drivers/usb/gadget/function/f_diag.c	2020-05-07 08:31:25.876786430 +0200
+++ changed/drivers/usb/gadget/function/f_diag.c	2019-06-03 13:39:52.000000000 +0200
@@ -36,8 +36,13 @@ static struct usb_interface_descriptor i
 	.bDescriptorType    =	USB_DT_INTERFACE,
 	.bNumEndpoints      =	2,
 	.bInterfaceClass    =	0xFF,
+#ifdef CONFIG_USB_ANDROID_SAMSUNG_COMPOSITE
+	.bInterfaceSubClass =	0x10,
+	.bInterfaceProtocol =	0x01,
+#else
 	.bInterfaceSubClass =	0xFF,
 	.bInterfaceProtocol =	0xFF,
+#endif
 };
 
 static struct usb_endpoint_descriptor hs_bulk_in_desc = {
