--- original/drivers/usb/serial/Kconfig	2020-05-07 08:06:19.073214808 +0200
+++ changed/drivers/usb/serial/Kconfig	2019-06-03 13:39:52.000000000 +0200
@@ -62,6 +62,7 @@ config USB_SERIAL_SIMPLE
 		- Fundamental Software dongle.
 		- HP4x calculators
 		- a number of Motorola phones
+		- Motorola Tetra devices
 		- Novatel Wireless GPS receivers
 		- Siemens USB/MPI adapter.
 		- ViVOtech ViVOpay USB device.
