--- original/drivers/media/platform/msm/camera_v2/Kconfig	2020-05-07 08:31:23.645806500 +0200
+++ changed/drivers/media/platform/msm/camera_v2/Kconfig	2019-06-03 13:39:50.000000000 +0200
@@ -9,6 +9,46 @@ config MSM_CAMERA_SENSOR
           data for camera support. The driver support V4L2
           subdev APIs.
 
+config MSM_FRONT_OTP
+        bool "Qualcomm MSM Camera ROM Interface for Calibration support"
+        depends on MSMB_CAMERA
+        ---help---
+          Enable support for ROM Interface for Calibration
+          Provides interface for reading the Claibration data.
+          and also provides support for writing data
+
+config MSM_OTP
+        bool "Qualcomm MSM Camera ROM Interface for Calibration support"
+        depends on MSMB_CAMERA
+        ---help---
+          Enable support for ROM Interface for Calibration
+          Provides interface for reading the Claibration data.
+          and also provides support for writing data
+
+config SR544
+    bool "Sensor for back camera"
+	depends on MSMB_CAMERA
+	default n
+	---help---
+	5MP sensor
+	Rear sensor
+
+config SR259
+    bool "Sensor for front camera"
+	depends on MSMB_CAMERA
+	default n
+	---help---
+	2MP sensor
+	Front sensor
+
+config S5K5E9YX
+        bool "Sensor for front camera"
+        depends on MSMB_CAMERA
+        default n
+        ---help---
+          5MP sensor
+          Front sensor
+
 config MSM_CPP
         bool "QTI MSM Camera Post Processing Engine support"
         depends on MSMB_CAMERA
@@ -95,6 +135,15 @@ config MSM_EEPROM
           and also provides support for writing data in case of FLASH ROM.
 	  Currently supports I2C, CCI and SPI protocol
 
+config MSM_FRONT_EEPROM
+        bool "Qualcomm MSM Camera ROM Interface for Calibration support"
+        depends on MSMB_CAMERA
+        ---help---
+          Enable support for ROM Interface for Calibration
+          Provides interface for reading the Claibration data.
+          and also provides support for writing data in case of FLASH ROM.
+          Currently supports I2C, CCI and SPI protocol
+
 config MSM_ISP_V1
         bool "QTI MSM Image Signal Processing interface support"
         depends on MSMB_CAMERA
@@ -203,6 +252,24 @@ config s5k4e1
 		and QSXGA snapshot at 15 fps.
 		This sensor driver does not support auto focus.
 
+config S5K5E3YX
+	bool "Sensor s5k5e3yx (BAYER 5MP)"
+	depends on MSMB_CAMERA
+	---help---
+		Samsung 5 MP Bayer Sensor. It uses 2 mipi lanes,
+
+config S5K3P8
+	bool "Sensor s5k3p8 (BAYER 16MP)"
+	depends on MSMB_CAMERA
+	---help---
+		Samsung 16 MP Bayer Sensor. It uses 4 mipi lanes,
+
+config S5K3L2XX
+	bool "Sensor s5k3l2xx (BAYER 5MP)"
+	depends on MSMB_CAMERA
+	---help---
+		Samsung 13 MP Bayer Sensor. It uses 4 mipi lanes,
+		
 config OV12830
 	bool "OmniVision OV12830 (BAYER 12MP)"
 	depends on MSMB_CAMERA
@@ -244,6 +311,12 @@ config MSM_FD
 	 ---help---
 	    Enables support for the MSM FD face detection engine.
 
+config CAMERA_SYSFS_V2
+	bool "CAMERA SYS"
+	depends on MSMB_CAMERA
+	---help---
+		CAMERA SYS.
+
 config MSM_JPEGDMA
 	tristate "QTI MSM Jpeg dma"
         depends on MSMB_CAMERA
@@ -262,3 +335,57 @@ config MSM_SEC_CCI_DEBUG
 	---help---
 	  Enables simulation of secure camera for Secure CCI Realy
 	  debugging.
+
+config SAMSUNG_MULTI_CAMERA
+	tristate "Multi camera support"
+	depends on MSMB_CAMERA
+	default n
+	---help---
+	  Enables support for multi camera.
+
+config GET_REAR_MODULE_ID
+        tristate "Rear Camera Module ID from 0xAE"
+        depends on MSMB_CAMERA
+        ---help---
+          Enable support to get module ID 10byte from eeprom 0xAE then 
+          set to sys_class_camera_rear_moduleid
+
+config GET_FRONT_MODULE_ID
+        tristate "Front Camera Module ID from 0x56"
+        depends on MSMB_CAMERA
+        ---help---
+          Enable support to get module ID 10byte from eeprom 0x56 then 
+          set to sys_class_camera_front_moduleid
+
+config GET_FRONT_MODULE_ID_OTP
+        tristate "Front Camera Module ID from OTP"
+        depends on MSMB_CAMERA
+        ---help---
+          Enable support to get module ID 10byte from OTP then 
+          set to sys_class_camera_front_moduleid
+
+config GET_REAR_SENSOR_ID
+        tristate "Rear Sensor ID from 0xB8"
+        depends on MSMB_CAMERA
+        ---help---
+          Enable support to get rear camera sensor ID 10byte from eeprom 0xB8 then
+          set to sys_class_camera_rear_sensorid_exif
+
+config GET_FRONT_SENSOR_ID
+        tristate "Front Sensor ID from 0x60"
+        depends on MSMB_CAMERA
+        ---help---
+          Enable support to get front camera sensor ID 10byte from eeprom 0x60 then
+          set to sys_class_camera_front_sensorid_exif
+
+config CAMERA_DUAL_REAR
+        tristate "Rear Multi camera support"
+        depends on MSMB_CAMERA
+        ---help---
+          Enables support for rear multi camera.
+
+config CAMERA_DUAL_FRONT
+        tristate "Front Multi camera support"
+        depends on MSMB_CAMERA
+        ---help---
+          Enables support for front multi camera.
