--- original/drivers/clk/msm/clock-gcc-8952.c	2020-05-07 08:31:21.387826814 +0200
+++ changed/drivers/clk/msm/clock-gcc-8952.c	2019-06-03 13:39:49.000000000 +0200
@@ -880,8 +880,10 @@ static struct clk_freq_tbl ftbl_gcc_blsp
 
 static struct clk_freq_tbl ftbl_gcc_blsp1_2_qup1_4_spi_apps_clk_8917[] = {
 	F( 960000,	xo,	10,	1,	2),
+	F( 4000000, gpll0, 10, 1, 20),
 	F( 4800000,	xo,	4,	0,	0),
 	F( 9600000,	xo,	2,	0,	0),
+	F( 12500000, gpll0, 16, 1, 4),
 	F( 16000000,	gpll0,	10,	1,	5),
 	F( 19200000,	xo,	1,	0,	0),
 	F( 25000000,	gpll0,	16,	1,	2),
@@ -1313,7 +1315,7 @@ static struct rcg_clk jpeg0_clk_src = {
 static struct clk_freq_tbl ftbl_gcc_camss_mclk0_2_clk[] = {
 	F( 19200000,	xo,	1,	0,	0),
 	F( 24000000,	gpll6,	1,	1,	45),
-	F( 66670000,	gpll0,	12,	0,	0),
+	F( 24000000,	gpll0,	12,	0,	0),
 	F_END
 };
 
@@ -1558,7 +1560,11 @@ static struct rcg_clk byte1_clk_src = {
 };
 
 static struct clk_freq_tbl ftbl_gcc_mdss_esc0_clk[] = {
+#if defined(CONFIG_SAMSUNG_12_8MHZ_ESC_CLOCK)
+	F( 12800000,	xo,	1.5,	0,	0),
+#else
 	F( 19200000,	xo,	1,	0,	0),
+#endif
 	F_END
 };
 
