--- original/drivers/net/wireless/cnss_prealloc/cnss_prealloc.c	2020-05-07 08:53:30.866121456 +0200
+++ changed/drivers/net/wireless/cnss_prealloc/cnss_prealloc.c	2019-06-03 13:39:51.000000000 +0200
@@ -25,7 +25,7 @@
 
 static DEFINE_SPINLOCK(alloc_lock);
 
-#ifdef CONFIG_SLUB_DEBUG
+#ifdef CONFIG_SLUB_DEBUG_ON
 #define WCNSS_MAX_STACK_TRACE			64
 #endif
 
@@ -38,7 +38,7 @@ struct wcnss_prealloc {
 	int occupied;
 	size_t size;
 	void *ptr;
-#ifdef CONFIG_SLUB_DEBUG
+#ifdef CONFIG_SLUB_DEBUG_ON
 	unsigned long stack_trace[WCNSS_MAX_STACK_TRACE];
 	struct stack_trace trace;
 #endif
@@ -192,7 +192,7 @@ void wcnss_prealloc_deinit(void)
 	cnss_skb_prealloc_deinit();
 }
 
-#ifdef CONFIG_SLUB_DEBUG
+#ifdef CONFIG_SLUB_DEBUG_ON
 static void wcnss_prealloc_save_stack_trace(struct wcnss_prealloc *entry)
 {
 	struct stack_trace *trace = &entry->trace;
@@ -308,7 +308,7 @@ int wcnss_skb_prealloc_put(struct sk_buf
 EXPORT_SYMBOL(wcnss_skb_prealloc_put);
 #endif
 
-#ifdef CONFIG_SLUB_DEBUG
+#ifdef CONFIG_SLUB_DEBUG_ON
 void wcnss_prealloc_check_memory_leak(void)
 {
 	int i;
@@ -331,7 +331,7 @@ void wcnss_prealloc_check_memory_leak(vo
 }
 #endif
 
-#if defined(CONFIG_WCNSS_SKB_PRE_ALLOC) && defined(CONFIG_SLUB_DEBUG)
+#if defined(CONFIG_WCNSS_SKB_PRE_ALLOC) && defined(CONFIG_SLUB_DEBUG_ON)
 /* Check memory leak for socket buffer pre-alloc memeory pool */
 void wcnss_skb_prealloc_check_memory_leak(void)
 {
