--- original/drivers/misc/qcom/qdsp6v2/audio_hwacc_effects.c	2020-05-07 08:31:23.885804341 +0200
+++ changed/drivers/misc/qcom/qdsp6v2/audio_hwacc_effects.c	2019-06-03 13:39:50.000000000 +0200
@@ -710,6 +710,7 @@ static int audio_effects_release(struct
 	q6asm_audio_client_free(effects->ac);
 
 	mutex_destroy(&effects->lock);
+
 	kfree(effects);
 
 	pr_debug("%s: close session success\n", __func__);
@@ -740,6 +741,7 @@ static int audio_effects_open(struct ino
 
 	init_waitqueue_head(&effects->read_wait);
 	init_waitqueue_head(&effects->write_wait);
+
 	mutex_init(&effects->lock);
 
 	effects->opened = 0;
