--- original/drivers/staging/android/ion/msm/msm_ion.c	2020-05-07 08:31:25.470790082 +0200
+++ changed/drivers/staging/android/ion/msm/msm_ion.c	2019-06-03 13:39:52.000000000 +0200
@@ -119,6 +119,28 @@ static struct ion_heap_desc ion_heap_met
 };
 #endif
 
+static int ion_system_heap_size_notifier(struct notifier_block *nb,
+					 unsigned long action, void *data)
+{
+	show_ion_system_heap_size((struct seq_file *)data);
+	return 0;
+}
+
+static struct notifier_block ion_system_heap_nb = {
+	.notifier_call = ion_system_heap_size_notifier,
+};
+
+static int ion_system_heap_pool_size_notifier(struct notifier_block *nb,
+					      unsigned long action, void *data)
+{
+	show_ion_system_heap_pool_size((struct seq_file *)data);
+	return 0;
+}
+
+static struct notifier_block ion_system_heap_pool_nb = {
+	.notifier_call = ion_system_heap_pool_size_notifier,
+};
+
 static int msm_ion_lowmem_notifier(struct notifier_block *nb,
 					unsigned long action, void *data)
 {
@@ -1048,6 +1070,9 @@ static int msm_ion_probe(struct platform
 	idev = new_dev;
 
 	show_mem_notifier_register(&msm_ion_nb);
+	show_mem_extra_notifier_register(&ion_system_heap_nb);
+	show_mem_extra_notifier_register(&ion_system_heap_pool_nb);
+
 	return 0;
 
 freeheaps:
@@ -1068,6 +1093,8 @@ static int msm_ion_remove(struct platfor
 
 	ion_device_destroy(idev);
 	kfree(heaps);
+	show_mem_extra_notifier_unregister(&ion_system_heap_nb);
+	show_mem_extra_notifier_unregister(&ion_system_heap_pool_nb);
 	return 0;
 }
 
