--- original/drivers/cpufreq/cpufreq_stats.c	2020-05-07 08:06:09.814305789 +0200
+++ changed/drivers/cpufreq/cpufreq_stats.c	2019-06-03 13:39:49.000000000 +0200
@@ -524,6 +524,12 @@ static void cpufreq_allstats_create(unsi
 	struct all_cpufreq_stats *all_stat;
 	bool sort_needed = false;
 
+#ifdef CONFIG_SEC_PM
+	if (!all_freq_table) {
+		pr_warn("all_freq_table is not initialized\n");
+		return;
+	}
+#endif
 	all_stat = kzalloc(sizeof(struct all_cpufreq_stats),
 			GFP_KERNEL);
 	if (!all_stat) {
