--- original/drivers/coresight/coresight-tmc.c	2020-05-07 08:06:09.802305907 +0200
+++ changed/drivers/coresight/coresight-tmc.c	2019-06-03 13:39:49.000000000 +0200
@@ -1834,6 +1834,8 @@ static ssize_t tmc_etr_byte_cntr_read(st
 
 	dev_dbg(drvdata->dev, "%s: %zu bytes copied, %d bytes left\n",
 		__func__, len, (int) (drvdata->size - *ppos));
+
+	mutex_lock(&drvdata->usb_lock);
 	return len;
 
 read_err0:
