--- original/drivers/Kconfig	2020-05-07 08:31:21.211828397 +0200
+++ changed/drivers/Kconfig	2019-06-03 13:39:49.000000000 +0200
@@ -20,6 +20,8 @@ source "drivers/block/Kconfig"
 
 source "drivers/misc/Kconfig"
 
+source "drivers/staging/samsung/sec_notifier/Kconfig"
+
 source "drivers/ide/Kconfig"
 
 source "drivers/scsi/Kconfig"
@@ -174,6 +176,12 @@ source "drivers/irqchip/Kconfig"
 
 source "drivers/ipack/Kconfig"
 
+source "drivers/battery/Kconfig"
+
+source "drivers/battery_v2/Kconfig"
+
+source "drivers/battery_v2_legacy/Kconfig"
+
 source "drivers/reset/Kconfig"
 
 source "drivers/fmc/Kconfig"
@@ -194,6 +202,26 @@ source "drivers/sensors/Kconfig"
 
 source "drivers/firmware/Kconfig"
 
+source "drivers/fingerprint/Kconfig"
+
 source "drivers/bif/Kconfig"
 
+source "drivers/muic/Kconfig"
+
+source "drivers/motor/Kconfig"
+
+source "drivers/muic/universal/Kconfig"
+
+source "drivers/nfc/Kconfig"
+
+source "drivers/ccic/Kconfig"
+
+source "drivers/gud/Kconfig"
+
+source "drivers/five/Kconfig"
+
+source "drivers/debug/Kconfig"
+
+source "drivers/icdrv/Kconfig"
+
 endmenu
