--- original/drivers/tty/serial/Makefile	2020-05-07 08:06:18.587219584 +0200
+++ changed/drivers/tty/serial/Makefile	2019-06-03 13:39:52.000000000 +0200
@@ -61,8 +61,18 @@ obj-$(CONFIG_SERIAL_SGI_IOC3) += ioc3_se
 obj-$(CONFIG_SERIAL_ATMEL) += atmel_serial.o
 obj-$(CONFIG_SERIAL_UARTLITE) += uartlite.o
 obj-$(CONFIG_SERIAL_MSM) += msm_serial.o
-obj-$(CONFIG_SERIAL_MSM_HS) += msm_serial_hs.o
-obj-$(CONFIG_SERIAL_MSM_HSL) += msm_serial_hs_lite.o
+
+ifeq ($(CONFIG_SERIAL_FACTORY_UART_HS),y)
+    ifeq ($(CONFIG_SEC_FACTORY),y)
+        obj-$(CONFIG_SERIAL_MSM_HS) += msm_serial_hs.o
+    else
+        obj-$(CONFIG_SERIAL_MSM_HSL) += msm_serial_hs_lite.o
+    endif
+else
+    obj-$(CONFIG_SERIAL_MSM_HS) += msm_serial_hs.o
+    obj-$(CONFIG_SERIAL_MSM_HSL) += msm_serial_hs_lite.o
+endif
+
 obj-$(CONFIG_SERIAL_NETX) += netx-serial.o
 obj-$(CONFIG_SERIAL_OF_PLATFORM) += of_serial.o
 obj-$(CONFIG_SERIAL_OF_PLATFORM_NWPSERIAL) += nwpserial.o
