--- original/drivers/video/msm/mdss/mdss_dsi_clk.c	2020-05-07 08:06:19.456211045 +0200
+++ changed/drivers/video/msm/mdss/mdss_dsi_clk.c	2019-06-03 13:39:53.000000000 +0200
@@ -798,7 +798,7 @@ int mdss_dsi_clk_req_state(void *client,
 					changed = true;
 					c->core_clk_state = MDSS_DSI_CLK_OFF;
 				} else {
-					pr_warn("Core refcount is zero for %s",
+					pr_warn("Core refcount is zero for %s\n",
 						c->name);
 				}
 			} else {
@@ -817,7 +817,7 @@ int mdss_dsi_clk_req_state(void *client,
 					changed = true;
 					c->link_clk_state = MDSS_DSI_CLK_OFF;
 				} else {
-					pr_warn("Link refcount is zero for %s",
+					pr_warn("Link refcount is zero for %s\n",
 						c->name);
 				}
 			} else {
