--- original/drivers/media/platform/msm/camera_v2/isp/msm_isp_axi_util_32.c	2020-05-07 11:18:28.041756297 +0200
+++ changed/drivers/media/platform/msm/camera_v2/isp/msm_isp_axi_util_32.c	2019-06-03 13:39:50.000000000 +0200
@@ -1,4 +1,4 @@
-/* Copyright (c) 2013-2018, The Linux Foundation. All rights reserved.
+/* Copyright (c) 2013-2017, The Linux Foundation. All rights reserved.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 and
@@ -1127,10 +1127,6 @@ static enum msm_isp_camif_update_state
 	cur_pix_stream_cnt =
 		axi_data->src_info[VFE_PIX_0].pix_stream_count +
 		axi_data->src_info[VFE_PIX_0].raw_stream_count;
-
-	if (stream_cfg_cmd->num_streams > VFE_AXI_SRC_MAX)
-		return NO_UPDATE;
-
 	for (i = 0; i < stream_cfg_cmd->num_streams; i++) {
 		stream_info =
 			&axi_data->stream_info[
@@ -1920,6 +1916,11 @@ int msm_isp_update_axi_stream(struct vfe
 			stream_info->output_format = update_info->output_format;
 			if (stream_info->state == ACTIVE) {
 				stream_info->state = PAUSE_PENDING;
+				trace_printk("%s:%d state %d src %d stream id %d\n",
+					__func__, __LINE__,
+					stream_info->state
+					stream_info->stream_src,
+					stream_info->stream_id);
 				msm_isp_axi_stream_enable_cfg(
 					vfe_dev, stream_info);
 				stream_info->state = PAUSING;
