--- original/drivers/pinctrl/qcom/Kconfig	2020-05-07 08:31:24.716796865 +0200
+++ changed/drivers/pinctrl/qcom/Kconfig	2019-06-03 13:39:51.000000000 +0200
@@ -143,3 +143,13 @@ config PINCTRL_MSM8909
 	  This is the pinctrl, pinmux, pinconf and gpiolib driver for the
 	  QTI TLMM block found in the QTI MSM8909 platform.
 endif
+
+config SENSORS_FP_SPI_GPIO_START
+	depends on SENSORS_FINGERPRINT
+	int "Fingerprint spi gpio number start"
+	default -1
+
+config SENSORS_FP_SPI_GPIO_END
+	depends on SENSORS_FINGERPRINT
+	int "Fingerprint spi gpio number end"
+	default -1
