--- original/drivers/rtc/interface.c	2020-05-07 08:51:34.989256473 +0200
+++ changed/drivers/rtc/interface.c	2019-06-03 13:39:52.000000000 +0200
@@ -451,7 +451,41 @@ int rtc_initialize_alarm(struct rtc_devi
 }
 EXPORT_SYMBOL_GPL(rtc_initialize_alarm);
 
+#ifdef CONFIG_RTC_AUTO_PWRON
+int rtc_set_bootalarm(struct rtc_device *rtc, struct rtc_wkalrm *alarm)
+{
+    int err;
+
+	if (!rtc->ops) {
+		dev_err(&rtc->dev, "ops not exist\n");
+		err = -ENODEV;
+	} else if (!rtc->ops->set_bootalarm) {
+		dev_err(&rtc->dev, "bootalarm func not exist\n");
+		err = -EINVAL;
+	} else
+		err = rtc->ops->set_bootalarm(rtc->dev.parent, alarm);
+
+	return err;
+}
+EXPORT_SYMBOL_GPL(rtc_set_bootalarm);
 
+int rtc_get_bootalarm(struct rtc_device *rtc, struct rtc_wkalrm *alarm)
+{
+    int err;
+
+	if (!rtc->ops) {
+		dev_err(&rtc->dev, "ops not exist\n");
+		err = -ENODEV;
+	} else if (!rtc->ops->read_bootalarm) {
+		dev_err(&rtc->dev, "bootalarm func not exist\n");
+		err = -EINVAL;
+	} else
+		err = rtc->ops->read_bootalarm(rtc->dev.parent, alarm);
+
+	return err;
+}
+EXPORT_SYMBOL_GPL(rtc_get_bootalarm);
+#endif /* CONFIG_AUTO_PWRON */
 
 int rtc_alarm_irq_enable(struct rtc_device *rtc, unsigned int enabled)
 {
