--- original/drivers/input/keyboard/Makefile	2020-05-07 08:06:12.137282963 +0200
+++ changed/drivers/input/keyboard/Makefile	2019-06-03 13:39:50.000000000 +0200
@@ -18,6 +18,8 @@ obj-$(CONFIG_KEYBOARD_DAVINCI)		+= davin
 obj-$(CONFIG_KEYBOARD_EP93XX)		+= ep93xx_keypad.o
 obj-$(CONFIG_KEYBOARD_GOLDFISH_EVENTS)	+= goldfish_events.o
 obj-$(CONFIG_KEYBOARD_GPIO)		+= gpio_keys.o
+obj-$(CONFIG_SENSORS_HALL)		+= hall_ic.o
+obj-$(CONFIG_CERTIFY_HALL)		+= certify_hall_ic.o
 obj-$(CONFIG_KEYBOARD_GPIO_POLLED)	+= gpio_keys_polled.o
 obj-$(CONFIG_KEYBOARD_TCA6416)		+= tca6416-keypad.o
 obj-$(CONFIG_KEYBOARD_TCA8418)		+= tca8418_keypad.o
@@ -51,6 +53,7 @@ obj-$(CONFIG_KEYBOARD_SAMSUNG)		+= samsu
 obj-$(CONFIG_KEYBOARD_SH_KEYSC)		+= sh_keysc.o
 obj-$(CONFIG_KEYBOARD_SPEAR)		+= spear-keyboard.o
 obj-$(CONFIG_KEYBOARD_STMPE)		+= stmpe-keypad.o
+obj-$(CONFIG_KEYBOARD_STMPE1801)	+= stmpe1801_key.o
 obj-$(CONFIG_KEYBOARD_STOWAWAY)		+= stowaway.o
 obj-$(CONFIG_KEYBOARD_ST_KEYSCAN)	+= st-keyscan.o
 obj-$(CONFIG_KEYBOARD_SUNKBD)		+= sunkbd.o
@@ -59,3 +62,9 @@ obj-$(CONFIG_KEYBOARD_TEGRA)		+= tegra-k
 obj-$(CONFIG_KEYBOARD_TWL4030)		+= twl4030_keypad.o
 obj-$(CONFIG_KEYBOARD_XTKBD)		+= xtkbd.o
 obj-$(CONFIG_KEYBOARD_W90P910)		+= w90p910_keypad.o
+obj-$(CONFIG_KEYBOARD_ABOV_TOUCH_18XX)	+= abov_touchkey_18xx.o
+obj-$(CONFIG_KEYBOARD_ABOV_TOUCH_16XX)	+= abov_touchkey_16xx.o
+obj-$(CONFIG_KEYBOARD_ABOV_TOUCH_T316)	+= abov_tk_t316.o
+obj-$(CONFIG_KEYBOARD_TC3XXK)   += tc3xxk.o
+obj-$(CONFIG_KEYBOARD_MELFAS_MHS204)	+= melfas_mip4/
+obj-$(CONFIG_KEYBOARD_TC300K) 	+= tc300k.o
\ Geen regeleindeteken (LF) aan einde van bestand
