--- original/drivers/media/platform/msm/camera_v2/fd/msm_fd_dev.c	2020-05-07 08:31:23.647806482 +0200
+++ changed/drivers/media/platform/msm/camera_v2/fd/msm_fd_dev.c	2019-06-03 13:39:50.000000000 +0200
@@ -1,4 +1,4 @@
-/* Copyright (c) 2014-2017, The Linux Foundation. All rights reserved.
+/* Copyright (c) 2014-2018, The Linux Foundation. All rights reserved.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 and
@@ -440,7 +440,7 @@ static int msm_fd_open(struct file *file
 	}
 
 	ctx->mem_pool.fd_device = ctx->fd_device;
-	ctx->stats = vmalloc(sizeof(*ctx->stats) * MSM_FD_MAX_RESULT_BUFS);
+	ctx->stats = vzalloc(sizeof(*ctx->stats) * MSM_FD_MAX_RESULT_BUFS);
 	if (!ctx->stats) {
 		dev_err(device->dev, "No memory for face statistics\n");
 		ret = -ENOMEM;
