--- original/drivers/usb/gadget/function/storage_common.h	2020-05-07 08:06:18.751217972 +0200
+++ changed/drivers/usb/gadget/function/storage_common.h	2019-06-03 13:39:52.000000000 +0200
@@ -68,6 +68,11 @@ do {									\
 /* Length of a SCSI Command Data Block */
 #define MAX_COMMAND_SIZE	16
 
+#ifdef CONFIG_USB_ANDROID_SAMSUNG_COMPOSITE
+/* SCSI commands that we recognize */
+#define READ_CD					0xbe
+#endif
+
 /* SCSI Sense Key/Additional Sense Code/ASC Qualifier values */
 #define SS_NO_SENSE				0
 #define SS_COMMUNICATION_FAILURE		0x040800
