--- original/drivers/regulator/Makefile	2020-05-07 08:31:24.880795390 +0200
+++ changed/drivers/regulator/Makefile	2019-06-03 13:39:52.000000000 +0200
@@ -111,4 +111,6 @@ obj-$(CONFIG_REGULATOR_QPNP_LABIBB) += q
 obj-$(CONFIG_REGULATOR_STUB) += stub-regulator.o
 obj-$(CONFIG_REGULATOR_KRYO) += kryo-regulator.o
 obj-$(CONFIG_REGULATOR_CPR2_GFX) += cpr2-gfx-regulator.o
+obj-$(CONFIG_REGULATOR_SM5703) += sm5703_regulator.o
+obj-$(CONFIG_REGULATOR_SM5705) += sm5705-regulator.o
 ccflags-$(CONFIG_REGULATOR_DEBUG) += -DDEBUG
