--- original/drivers/media/platform/msm/vidc/msm_v4l2_vidc.c	2020-05-07 08:31:23.713805889 +0200
+++ changed/drivers/media/platform/msm/vidc/msm_v4l2_vidc.c	2019-06-03 13:39:50.000000000 +0200
@@ -36,7 +36,11 @@
 
 struct msm_vidc_drv *vidc_driver;
 
+#if defined(CONFIG_SEC_J4CORELTE_PROJECT)
+uint32_t msm_vidc_pwr_collapse_delay = 5000;
+#else
 uint32_t msm_vidc_pwr_collapse_delay = 2000;
+#endif
 
 static inline struct msm_vidc_inst *get_vidc_inst(struct file *filp, void *fh)
 {
