--- original/drivers/misc/Kconfig	2020-05-07 08:31:23.859804575 +0200
+++ changed/drivers/misc/Kconfig	2019-06-03 13:39:50.000000000 +0200
@@ -304,6 +304,17 @@ config APDS9802ALS
 	  This driver can also be built as a module.  If so, the module
 	  will be called apds9802als.
 
+config SAMSUNG_JACK
+	bool "3.5 PI Jack driver for Samsung devices"
+	depends on INPUT
+	default n
+	help
+	 This is 3.5 PI JACK driver for Samsung devies.
+	 sec_jack from samsung driver used for samsung devices.
+
+	 To samsung jack is reviewed and supported in all the
+	 samsung devices.
+
 config ISL29003
 	tristate "Intersil ISL29003 ambient light sensor"
 	depends on I2C && SYSFS
@@ -324,6 +335,12 @@ config ISL29020
 	  This driver can also be built as a module.  If so, the module
 	  will be called isl29020.
 
+config KNOX_KAP
+       bool "Enable KNOX KAP mode"
+       default n
+       help 
+              Knox Active Protection mode
+
 config SENSORS_TSL2550
 	tristate "Taos TSL2550 ambient light sensor"
 	depends on I2C && SYSFS
@@ -365,6 +382,18 @@ config SENSORS_APDS990X
 	   To compile this driver as a module, choose M here: the
 	   module will be called apds990x. If unsure, say N here.
 
+config TIMA_LOG
+	tristate "Support for dumping TIMA log"
+	default n
+	help
+	  This option enables support for dumping TIMA log
+
+config TIMA_TEST
+	tristate "Tima Test Suite"
+	default n
+	help
+	  This option enables kernel component of tima test suite
+
 config APDS9930
 	 tristate "Avago APDS9930 combined als and proximity sensors"
 	 depends on I2C
@@ -528,6 +557,12 @@ config QSEECOM
           Manager (SCM) interface. It exposes APIs for both userspace and
           kernel clients.
 
+config QSEECOM_DEBUG
+        bool "QTI Secure Execution Communicator driver debug"
+	default	n
+        help
+          QTI Secure Execution Communicator driver debug		  
+
 config HDCP_QSEECOM
 	tristate "QTI High-Bandwidth Digital Content Protection Module"
 	help
@@ -600,6 +635,18 @@ config QPNP_MISC
 	  and the driver provides an API to check if this interrupt
 	  is available on the current PMIC chip.
 
+config SEC_MPP_SHARE
+	bool "Samsung MPP share"
+	help
+	  Say Y to include support for mpp share
+
+config SEC_MPP_SHARE_LOW_TEMP_WA
+	bool "Samsung MPP share Low Temp WA"
+	help
+	  Say Y to include support for MUX share WA.
+	  When EAR_ADC_3.5 and BAT_THM use same MUX Switch IC,
+	  connect some non-official earphone, will cause BAT_THM Low Temp.
+
 config UID_SYS_STATS_DEBUG
 	bool "Per-TASK statistics"
 	depends on UID_SYS_STATS
@@ -627,4 +674,6 @@ source "drivers/misc/mic/Kconfig"
 source "drivers/misc/genwqe/Kconfig"
 source "drivers/misc/echo/Kconfig"
 source "drivers/misc/cxl/Kconfig"
+source "drivers/misc/bhy/Kconfig"
+source "drivers/misc/dev_ril_bridge/Kconfig"
 endmenu
