--- original/drivers/tty/sysrq.c	2020-05-07 08:31:25.798787132 +0200
+++ changed/drivers/tty/sysrq.c	2019-06-03 13:39:52.000000000 +0200
@@ -51,6 +51,8 @@
 #include <asm/ptrace.h>
 #include <asm/irq_regs.h>
 
+#include <linux/notifier.h>
+
 /* Whether we react on sysrq keys or just ignore them */
 static int __read_mostly sysrq_enabled = CONFIG_MAGIC_SYSRQ_DEFAULT_ENABLE;
 static bool __read_mostly sysrq_always_enabled;
@@ -298,6 +300,9 @@ static struct sysrq_key_op sysrq_showsta
 static void sysrq_handle_showstate_blocked(int key)
 {
 	show_state_filter(TASK_UNINTERRUPTIBLE);
+	show_mem(0);
+	show_mem_extra_call_notifiers();
+	dump_tasks(NULL, NULL);
 }
 static struct sysrq_key_op sysrq_showstate_blocked_op = {
 	.handler	= sysrq_handle_showstate_blocked,
