--- original/drivers/thermal/Kconfig	2020-05-07 08:06:18.527220174 +0200
+++ changed/drivers/thermal/Kconfig	2019-06-03 13:39:52.000000000 +0200
@@ -202,6 +202,13 @@ config THERMAL_MONITOR
 	  entity starts running in the userspace. Monitors TSENS temperature
 	  and limits the max frequency of the cores.
 
+config SS_THERMAL_CPU_WEAR_LEVEL
+	bool "Monitor thermal state and limit CPU Frequency"
+	depends on THERMAL
+	default n
+	help
+	  Controlling CPU status for heat mitigation.
+
 config IMX_THERMAL
 	tristate "Temperature sensor driver for Freescale i.MX SoCs"
 	depends on CPU_THERMAL
