--- original/drivers/media/platform/msm/Kconfig	2020-05-07 08:31:23.554807319 +0200
+++ changed/drivers/media/platform/msm/Kconfig	2019-06-03 13:39:50.000000000 +0200
@@ -29,6 +29,13 @@ config MSM_CAMERA_AUTOMOTIVE
 	---help---
 	  Enables support for automotive targets
 
+config SR259
+	bool "Sensor SR259 (BAYER 2M)"
+	depends on MSMB_CAMERA
+	---help---
+		Siliconfile 2 MP Bayer Sensor,
+		Front Sensor.
+
 menuconfig MSMB_CAMERA
 	bool "QTI MSM camera and video capture 2.0 support"
 	depends on ARCH_MSM && VIDEO_V4L2 && I2C
