--- original/drivers/video/msm/mdss/mdss_mdp.c	2020-05-07 08:31:26.089784514 +0200
+++ changed/drivers/video/msm/mdss/mdss_mdp.c	2019-06-03 13:39:53.000000000 +0200
@@ -58,6 +58,10 @@
 
 #include "mdss_mdp_trace.h"
 
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+#include "samsung/ss_dsi_panel_common.h"
+#endif
+
 #define AXI_HALT_TIMEOUT_US	0x4000
 #define AUTOSUSPEND_TIMEOUT_MS	200
 #define DEFAULT_MDP_PIPE_WIDTH	2048
@@ -99,7 +103,11 @@ static struct mdss_panel_intf pan_types[
 	{"edp", MDSS_PANEL_INTF_EDP},
 	{"hdmi", MDSS_PANEL_INTF_HDMI},
 };
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+char mdss_mdp_panel[MDSS_MAX_PANEL_LEN];
+#else
 static char mdss_mdp_panel[MDSS_MAX_PANEL_LEN];
+#endif
 
 struct mdss_hw mdss_mdp_hw = {
 	.hw_ndx = MDSS_HW_MDP,
@@ -2184,7 +2192,7 @@ static int mdss_mdp_get_pan_cfg(struct m
 	/* point to the start of panel name */
 	t = t + 1;
 	strlcpy(&pan_cfg->arg_cfg[0], t, sizeof(pan_cfg->arg_cfg));
-	pr_debug("%d: t=[%s] panel name=[%s]\n", __LINE__,
+	pr_info("%d: t=[%s] panel name=[%s]\n", __LINE__,
 		t, pan_cfg->arg_cfg);
 
 	panel_len = strlen(pan_cfg->arg_cfg);
@@ -2787,6 +2795,11 @@ static int mdss_mdp_probe(struct platfor
 	if (mdss_res->mdp_irq_mask == NULL)
 		return -ENOMEM;
 
+#if defined(CONFIG_FB_MSM_MDSS_SAMSUNG)
+	if (!mdss_panel_attached(DISPLAY_1) && !mdss_panel_attached(DISPLAY_2))
+		mdata->handoff_pending = false;
+#endif
+
 	pr_info("mdss version = 0x%x, bootloader display is %s, num %d, intf_sel=0x%08x\n",
 		mdata->mdp_rev, num_of_display_on ? "on" : "off",
 		num_of_display_on, intf_sel);
