--- original/drivers/input/evdev.c	2020-05-07 08:31:23.241810135 +0200
+++ changed/drivers/input/evdev.c	2019-06-03 13:39:50.000000000 +0200
@@ -190,6 +190,11 @@ static void evdev_pass_values(struct evd
 		__pass_event(client, &event);
 		if (v->type == EV_SYN && v->code == SYN_REPORT)
 			wakeup = true;
+#ifdef CONFIG_USB_HMT_SAMSUNG_INPUT
+		if (v->type== EV_KEY && v->code >= KEY_HMT_CMD_START)
+			pr_info("%s type:KEY code:0x%x value:%x\n", __func__,
+					v->code, v->value);
+#endif
 	}
 
 	spin_unlock(&client->buffer_lock);
