--- original/drivers/base/power/main.c	2020-05-07 08:31:21.275827822 +0200
+++ changed/drivers/base/power/main.c	2019-06-03 13:39:49.000000000 +0200
@@ -59,6 +59,10 @@ struct suspend_stats suspend_stats;
 static DEFINE_MUTEX(dpm_list_mtx);
 static pm_message_t pm_transition;
 
+#ifdef CONFIG_SEC_PM
+extern int wakeup_gpio_irq_flag;
+#endif
+
 static int async_error;
 
 static char *pm_verb(int event)
@@ -621,6 +625,10 @@ static int device_resume_early(struct de
 	TRACE_DEVICE(dev);
 	TRACE_RESUME(0);
 
+#ifdef CONFIG_SEC_PM
+	wakeup_gpio_irq_flag = 0;
+#endif
+
 	if (dev->power.syscore || dev->power.direct_complete)
 		goto Out;
 
