--- original/drivers/hid/Kconfig	2020-05-07 08:06:11.550288731 +0200
+++ changed/drivers/hid/Kconfig	2019-06-03 13:39:50.000000000 +0200
@@ -745,6 +745,13 @@ config SMARTJOYPLUS_FF
 	---help---
 	Say Y here if you have a SmartJoy PLUS PS2/USB adapter and want to
 	enable force feedback support for it.
+	
+config HID_SYNAPTICS_BT
+	tristate "HID Synaptics"
+	depends on HID
+	default n
+	---help---
+	For a Bluetooth HID touchpad
 
 config HID_TIVO
 	tristate "TiVo Slide Bluetooth remote control support"
@@ -865,6 +872,12 @@ config HID_SENSOR_HUB
 	  for events and handle data streams. Each sensor driver can format
 	  data and present to user mode using input or IIO interface.
 
+config HID_MADCATZ
+	tristate "Madcatz wireless game controller support"
+	depends on (USB_HID || HID)
+	---help---
+	Support for Madcatz wireless game controller support.	
+
 endmenu
 
 endif # HID
