--- original/drivers/staging/android/ion/ion_system_secure_heap.c	2020-05-07 08:31:25.468790100 +0200
+++ changed/drivers/staging/android/ion/ion_system_secure_heap.c	2019-06-03 13:39:52.000000000 +0200
@@ -155,8 +155,8 @@ static void ion_system_secure_heap_prefe
 		ret = sys_heap->ops->allocate(sys_heap, buffer, size,
 						PAGE_SIZE, 0);
 		if (ret) {
-			pr_debug("%s: Failed to get %zx allocation for %s, ret = %d\n",
-				__func__, info->size, secure_heap->heap.name,
+			pr_debug("%s: Failed to get allocation for %s, ret = %d\n",
+				__func__, secure_heap->heap.name,
 				ret);
 			spin_lock_irqsave(&secure_heap->work_lock, flags);
 			continue;
