--- original/drivers/char/mem.c	2020-05-07 08:31:21.370826967 +0200
+++ changed/drivers/char/mem.c	2019-06-03 13:39:49.000000000 +0200
@@ -34,6 +34,10 @@
 # include <linux/efi.h>
 #endif
 
+#ifdef CONFIG_KNOX_KAP
+#include <linux/knox_kap.h>
+#endif
+
 #define DEVPORT_MINOR	4
 
 static inline unsigned long size_inside_page(unsigned long start,
@@ -841,6 +845,9 @@ static const struct memdev {
 #ifdef CONFIG_PRINTK
 	[11] = { "kmsg", 0644, &kmsg_fops, NULL },
 #endif
+#ifdef CONFIG_KNOX_KAP
+	[13] = { "knox_kap", 0666, &knox_kap_fops, NULL },
+#endif
 };
 
 static int memory_open(struct inode *inode, struct file *filp)
