--- original/drivers/hid/Makefile	2020-05-07 08:06:11.551288721 +0200
+++ changed/drivers/hid/Makefile	2019-06-03 13:39:50.000000000 +0200
@@ -118,12 +118,14 @@ obj-$(CONFIG_HID_UCLOGIC)	+= hid-uclogic
 obj-$(CONFIG_HID_XINMO)		+= hid-xinmo.o
 obj-$(CONFIG_HID_ZEROPLUS)	+= hid-zpff.o
 obj-$(CONFIG_HID_ZYDACRON)	+= hid-zydacron.o
+obj-$(CONFIG_HID_SYNAPTICS_BT)  += hid-synaptics-bt.o
 
 wacom-objs			:= wacom_wac.o wacom_sys.o
 obj-$(CONFIG_HID_WACOM)		+= wacom.o
 obj-$(CONFIG_HID_WALTOP)	+= hid-waltop.o
 obj-$(CONFIG_HID_WIIMOTE)	+= hid-wiimote.o
 obj-$(CONFIG_HID_SENSOR_HUB)	+= hid-sensor-hub.o
+obj-$(CONFIG_HID_MADCATZ)	+= hid-madcatz.o
 
 obj-$(CONFIG_USB_HID)		+= usbhid/
 obj-$(CONFIG_USB_MOUSE)		+= usbhid/
