--- original/drivers/gpu/msm/adreno_ringbuffer.h	2020-05-07 08:06:11.522289006 +0200
+++ changed/drivers/gpu/msm/adreno_ringbuffer.h	2019-06-03 13:39:50.000000000 +0200
@@ -1,4 +1,4 @@
-/* Copyright (c) 2002,2007-2016, The Linux Foundation. All rights reserved.
+/* Copyright (c) 2002,2007-2016,2018 The Linux Foundation. All rights reserved.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 and
@@ -105,6 +105,7 @@ struct adreno_ringbuffer_pagetable_info
  * or how long it has been scheduled for after preempting in
  * @starve_timer_state: Indicates the state of the wait.
  * @preempt_lock: Lock to protect the wptr pointer while it is being updated
+ * @timer: Starvation timer for this ringbuffer
  */
 struct adreno_ringbuffer {
 	uint32_t flags;
@@ -127,6 +128,7 @@ struct adreno_ringbuffer {
 	unsigned long sched_timer;
 	enum adreno_dispatcher_starve_timer_states starve_timer_state;
 	spinlock_t preempt_lock;
+	struct timer_list timer;
 };
 
 /* Returns the current ringbuffer */
