--- original/drivers/crypto/msm/ice.c	2020-05-07 11:40:16.447314239 +0200
+++ changed/drivers/crypto/msm/ice.c	2019-06-03 13:39:49.000000000 +0200
@@ -120,11 +120,14 @@ static int qti_ice_setting_config(struct
 		memcpy(&setting->crypto_data, crypto_data,
 				sizeof(setting->crypto_data));
 
-		if (rq_data_dir(req) == WRITE)
+		if (unlikely(req->cmd_flags & REQ_BYPASS)) {
+			setting->encr_bypass = true;
+			setting->decr_bypass = true;
+		} else if (rq_data_dir(req) == WRITE) {
 			setting->encr_bypass = false;
-		else if (rq_data_dir(req) == READ)
+		} else if (rq_data_dir(req) == READ) {
 			setting->decr_bypass = false;
-		else {
+		} else {
 			/* Should I say BUG_ON */
 			setting->encr_bypass = true;
 			setting->decr_bypass = true;
@@ -1664,9 +1667,6 @@ int qcom_ice_setup_ice_hw(const char *st
 	struct ice_device *ice_dev = NULL;
 
 	ice_dev = get_ice_device_from_storage_type(storage_type);
-	if (ice_dev == ERR_PTR(-EPROBE_DEFER))
-		return -EPROBE_DEFER;
-
 	if (!ice_dev)
 		return ret;
 
