--- original/drivers/usb/host/xhci.c	2020-05-07 08:31:25.961785665 +0200
+++ changed/drivers/usb/host/xhci.c	2019-06-03 13:39:52.000000000 +0200
@@ -4184,6 +4184,11 @@ int xhci_set_usb2_hardware_lpm(struct us
 	if (udev->usb2_hw_lpm_capable != 1)
 		return -EPERM;
 
+	/* some USB3.0 memory stick doesn't support L1 mode,
+	  * so we add XHCI_LPM_L1_DISABLE quirks for disabling L1 mode */
+	if (!(xhci->quirks & XHCI_LPM_L1_SUPPORT))
+		return -EPERM;
+
 	spin_lock_irqsave(&xhci->lock, flags);
 
 	port_array = xhci->usb2_ports;
