--- original/drivers/media/platform/msm/camera_v2/sensor/ois/msm_ois.c	2020-05-07 08:31:23.700806006 +0200
+++ changed/drivers/media/platform/msm/camera_v2/sensor/ois/msm_ois.c	2019-06-03 13:39:50.000000000 +0200
@@ -615,13 +615,11 @@ static long msm_ois_subdev_ioctl(struct
 			pr_err("o_ctrl->i2c_client.i2c_func_tbl NULL\n");
 			return -EINVAL;
 		}
-		mutex_lock(o_ctrl->ois_mutex);
 		rc = msm_ois_power_down(o_ctrl);
 		if (rc < 0) {
 			pr_err("%s:%d OIS Power down failed\n",
 				__func__, __LINE__);
 		}
-		mutex_unlock(o_ctrl->ois_mutex);
 		return msm_ois_close(sd, NULL);
 	default:
 		return -ENOIOCTLCMD;
