--- original/drivers/gpio/Kconfig	2020-05-07 08:31:21.511825699 +0200
+++ changed/drivers/gpio/Kconfig	2019-06-03 13:39:49.000000000 +0200
@@ -719,6 +719,12 @@ config GPIO_ADNP
 	  enough to represent all pins, but the driver will assume a
 	  register layout for 64 pins (8 registers).
 
+config GPIO_PCAL6416A
+	bool "PCAL6416A GPIO expander"
+	depends on I2C
+	help
+	  GPIO driver for PCAL6416A I2C-based GPIO expander.
+
 comment "PCI GPIO expanders:"
 
 config GPIO_CS5535
@@ -913,6 +919,12 @@ config GPIO_BCM_KONA
 	help
 	  Turn on GPIO support for Broadcom "Kona" chips.
 
+config SEC_GPIO_DVS
+	tristate "setting Samsung GPIO debugging and verification system"
+	help
+	  To verify gpio configurations of devices, set this feature.
+	  This feature should be enabled for user-debug mode.
+
 comment "USB GPIO expanders:"
 
 config GPIO_VIPERBOARD
@@ -926,4 +938,11 @@ config GPIO_VIPERBOARD
           River Tech's viperboard.h for detailed meaning
           of the module parameters.
 
+config SEC_CONNECTOR_DETECTOR
+	tristate "Setting Samsung Connector Detector feature"
+	help
+	  To detect connection of connector, set this feature.
+	  This feature should be enabled for factory mode.
+
+
 endif
