--- original/drivers/usb/gadget/Kconfig	2020-05-07 08:06:18.717218307 +0200
+++ changed/drivers/usb/gadget/Kconfig	2019-06-03 13:39:52.000000000 +0200
@@ -432,12 +432,83 @@ config USB_G_ANDROID
 	  Each function can be configured and enabled/disabled
 	  dynamically from userspace through a sysfs interface.
 
+config USB_ANDROID_SAMSUNG_COMPOSITE
+	boolean "Samsung Composite function"
+	depends on USB_G_ANDROID
+	help
+		Provides SAMSUNG composite driver.
+		It also provides KIES connection and
+		Multi Configuration.
+		If you enable this option, android composite will be changed.
+
+config USB_DUN_SUPPORT
+	boolean "DUN support function"
+	depends on USB_G_ANDROID
+	help
+	  Provides USB modem serial driver.
+	  This function makes connection to acm from data router.
+	  It uses misc register.
+	  Support fops : open, close, release, read, poll, llseek, ioctl
+
+config USB_ANDROID_SAMSUNG_MTP
+	boolean "Samsung MTP function"
+	depends on USB_G_ANDROID && !SEC_FACTORY
+	help
+		Provides Media Transfer Protocol (MTP) support
+		for samsung gadget driver.
+		If you enable this option,
+		google mtp will be changed to samsung mtp.
+
+config USB_ANDROID_SAMSUNG_SIDESYNC
+	boolean "Samsung SideSync function"
+	depends on USB_G_ANDROID
+	help
+	  Provides SideSync support
+	  for samsung gadget driver.
+	  If you enable this option,
+	  SideSync will be able to be used.
+
+config USB_LOCK_SUPPORT_FOR_MDM
+	boolean "Samsung USB_LOCK function for MDM"
+	depends on USB_G_ANDROID
+	help
+	  Support USB lock for SDS MDM function.
+	  If you enable this option,
+	  USB is not connected to the work in office.
+	  Of employees who have installed MDM(SDS).
+
+config USB_RNDIS_VZW_REQ
+	boolean "Samsung USB RNDIS 5.x function"
+	depends on USB_G_ANDROID
+	help
+	  Provides USB RNDIS version 5.x support
+	  for samsung gadget driver.
+	  If you enable this option,
+	  you can connect via RNDIS 5.x.
+	  It's VZW carrier concept.
+
+config USB_NCM_SUPPORT_MTU_CHANGE
+	boolean "Samsung NCM MTU Change function"
+	depends on USB_G_ANDROID
+	help
+	   Provides NCM MTU Change support
+	   for samsung gadget driver.
+	   If you enable this option,
+	   NCM MTU Change will be able to be used.
+
 config USB_ANDROID_RNDIS_DWORD_ALIGNED
 	boolean "Use double word aligned"
 	depends on USB_G_ANDROID
 	help
 		Provides dword aligned for DMA controller.
 
+config USB_CHARGING_EVENT
+	bool "Set usb charging current "
+	depends on USB_G_ANDROID
+	help
+	  For satisfy BC1.2 specification
+	  In unconfigured state or suspend state, Set vbus current to 100mA
+
 source "drivers/usb/gadget/legacy/Kconfig"
 
 endchoice
