--- original/drivers/input/touchscreen/Kconfig	2020-05-07 08:31:23.334809299 +0200
+++ changed/drivers/input/touchscreen/Kconfig	2019-06-03 13:39:50.000000000 +0200
@@ -11,11 +11,8 @@ menuconfig INPUT_TOUCHSCREEN
 
 if INPUT_TOUCHSCREEN
 
-source "drivers/input/touchscreen/synaptics_dsx/Kconfig"
-source "drivers/input/touchscreen/synaptics_dsx_2.6/Kconfig"
-
 config OF_TOUCHSCREEN
-	def_tristate INPUT
+	#def_tristate INPUT
 	depends on INPUT && OF
 
 config TOUCHSCREEN_88PM860X
@@ -995,36 +992,25 @@ config TOUCHSCREEN_FT5X06_GESTURE
 
          If unsure, say N.
 
-config TOUCHSCREEN_SYNAPTICS_DSX_FW_UPDATE_EXTRA_SYSFS
-	bool "Synaptics DSX firmware update extra sysfs attributes"
-	depends on TOUCHSCREEN_SYNAPTICS_DSX_FW_UPDATE
+config TOUCHSCREEN_MSTAR21XX
+	tristate "mstar touchscreens"
+	depends on I2C
 	help
-	  Say Y here to enable support for extra sysfs attributes
-	  supporting firmware update in a development environment.
-	  This does not affect the core or other subsystem attributes.
+	 Say Y here if you have a mstar touchscreen.
 
-	  If unsure, say N.
+	 If unsure, say N.
+
+	 To compile this driver as a module, choose M here: the
+	 module will be called msg21xx_ts.
 
 config SECURE_TOUCH
 	bool "Secure Touch"
-	depends on (TOUCHSCREEN_SYNAPTICS_I2C_RMI4 || \
-	  TOUCHSCREEN_SYNAPTICS_DSX_I2C_v21 || TOUCHSCREEN_ATMEL_MXT || \
-	  TOUCHSCREEN_ATMEL_MAXTOUCH_TS)
+	depends on I2C
 	help
 	  Say Y here to enable Secure Touch in supported drivers.
 
 	  If unsure, say N.
 
-config TOUCHSCREEN_SYNAPTICS_DSX_FW_UPDATE_EXTRA_SYSFS
-	bool "Synaptics DSX firmware update extra sysfs attributes"
-	depends on TOUCHSCREEN_SYNAPTICS_DSX_FW_UPDATE
-	help
-	  Say Y here to enable support for extra sysfs attributes
-	  supporting firmware update in a development environment.
-	  This does not affect the core or other subsystem attributes.
-
-	  If unsure, say N.
-
 config TOUCHSCREEN_IT7260_I2C
 	tristate "IT7260 Touchscreen Driver"
 	depends on I2C
@@ -1135,9 +1121,6 @@ config TOUCHSCREEN_MAXIM_STI
 	  To compile this driver as a module, choose M here: the
 	  module will be called maxim_sti.
 
-source "drivers/input/touchscreen/gt9xx/Kconfig"
-source "drivers/input/touchscreen/focaltech_touch/Kconfig"
-
 config TOUCHSCREEN_HIMAX_CHIPSET
 	bool "Himax touchpanel CHIPSET"
 	depends on I2C
@@ -1148,6 +1131,23 @@ config TOUCHSCREEN_HIMAX_CHIPSET
 
           If unsure, say N.
 
-source "drivers/input/touchscreen/hxchipset/Kconfig"
+config TOUCHSCREEN_DUMP_MODE
+	bool "touchscreen dump mode"
+	depends on I2C
+	help
+	  Say Y here if you need to get raw data.
+
+	  If unsure, say N.
 
+source "drivers/input/touchscreen/focaltech_touch/Kconfig"
+source "drivers/input/touchscreen/hxchipset/Kconfig"
+source "drivers/input/touchscreen/imagis_30xxh/Kconfig"
+source "drivers/input/touchscreen/sec_ts/Kconfig"
+source "drivers/input/touchscreen/sec_incell_ts/Kconfig"
+source "drivers/input/touchscreen/imagis_30xxh/Kconfig"
+source "drivers/input/touchscreen/stm/fts7/Kconfig"
+source "drivers/input/touchscreen/stm/fts8/Kconfig"
+source "drivers/input/touchscreen/stm/fts1ba90a/Kconfig"
+source "drivers/input/touchscreen/zinitix/Kconfig"
+source "drivers/input/touchscreen/synaptics_incell/Kconfig"
 endif
