--- original/drivers/bluetooth/btfm_slim.c	2020-05-07 08:06:09.435309513 +0200
+++ changed/drivers/bluetooth/btfm_slim.c	2019-06-03 13:39:49.000000000 +0200
@@ -1,4 +1,4 @@
-/* Copyright (c) 2017, The Linux Foundation. All rights reserved.
+/* Copyright (c) 2016, The Linux Foundation. All rights reserved.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 and
@@ -509,6 +509,7 @@ static int btfm_slim_remove(struct slim_
 	BTFMSLIM_DBG("");
 	mutex_destroy(&btfm_slim->io_lock);
 	mutex_destroy(&btfm_slim->xfer_lock);
+	kfree(btfm_slim);
 	snd_soc_unregister_codec(&slim->dev);
 
 	BTFMSLIM_DBG("slim_remove_device() - btfm_slim->slim_ifd");
@@ -516,8 +517,6 @@ static int btfm_slim_remove(struct slim_
 
 	BTFMSLIM_DBG("slim_remove_device() - btfm_slim->slim_pgd");
 	slim_remove_device(slim);
-
-	kfree(btfm_slim);
 	return 0;
 }
 
