--- original/drivers/clk/msm/clock.c	2020-05-07 08:06:09.701306899 +0200
+++ changed/drivers/clk/msm/clock.c	2019-06-03 13:39:49.000000000 +0200
@@ -28,6 +28,7 @@
 #include <linux/clk/msm-clk-provider.h>
 #include <linux/of_platform.h>
 #include <linux/pm_opp.h>
+#include <linux/sec_debug.h>
 
 #include <trace/events/power.h>
 #include "clock.h"
@@ -672,6 +673,7 @@ int clk_set_rate(struct clk *clk, unsign
 	}
 
 	trace_clock_set_rate(name, rate, raw_smp_processor_id());
+	sec_debug_clock_rate_log(name, rate, raw_smp_processor_id());
 
 	start_rate = clk->rate;
 
@@ -707,6 +709,7 @@ int clk_set_rate(struct clk *clk, unsign
 		__clk_notify(clk, POST_RATE_CHANGE, start_rate, clk->rate);
 
 	trace_clock_set_rate_complete(name, clk->rate, raw_smp_processor_id());
+	sec_debug_clock_rate_complete_log(name, clk->rate, raw_smp_processor_id());
 out:
 	mutex_unlock(&clk->prepare_lock);
 	return rc;
