--- original/drivers/nfc/Makefile	2020-05-07 08:06:16.077244248 +0200
+++ changed/drivers/nfc/Makefile	2019-06-03 13:39:51.000000000 +0200
@@ -1,10 +1,11 @@
 #
 # Makefile for nfc devices
 #
+ifneq ($(SEC_BUILD_CONF_USE_ESE_TZ), false)
+	ccflags-y += -DENABLE_ESE_SPI_SECURED
+endif
 
-obj-$(CONFIG_NFC_PN544)		+= pn544/
 obj-$(CONFIG_NFC_MICROREAD)	+= microread/
-obj-$(CONFIG_NFC_PN533)		+= pn533.o
 obj-$(CONFIG_NFC_WILINK)	+= nfcwilink.o
 obj-$(CONFIG_NFC_MEI_PHY)	+= mei_phy.o
 obj-$(CONFIG_NFC_SIM)		+= nfcsim.o
@@ -14,5 +15,9 @@ obj-$(CONFIG_NFC_TRF7970A)	+= trf7970a.o
 obj-$(CONFIG_NFC_ST21NFCA)  	+= st21nfca/
 obj-$(CONFIG_NFC_ST21NFCB)	+= st21nfcb/
 obj-$(CONFIG_NFC_NQ)		+= nq-nci.o
+obj-$(CONFIG_SEC_NFC)           += sec_nfc.o
+obj-$(CONFIG_ESE_P3)           += ese_p3.o
+obj-$(CONFIG_NFC_PN547_ESE_SUPPORT)        += p61.o
+obj-$(CONFIG_NFC_PN547)        += pn547.o
+obj-y		+= nfc_logger/
 
-ccflags-$(CONFIG_NFC_DEBUG) := -DDEBUG
