--- original/drivers/hwmon/Makefile	2020-05-07 08:06:11.661287640 +0200
+++ changed/drivers/hwmon/Makefile	2019-06-03 13:39:50.000000000 +0200
@@ -9,6 +9,9 @@ obj-$(CONFIG_HWMON_VID)		+= hwmon-vid.o
 obj-$(CONFIG_SENSORS_ACPI_POWER) += acpi_power_meter.o
 obj-$(CONFIG_SENSORS_ATK0110)	+= asus_atk0110.o
 
+# Thermistor
+obj-$(CONFIG_SEC_THERMISTOR)	+= sec_thermistor.o
+
 # Native drivers
 # asb100, then w83781d go first, as they can override other drivers' addresses.
 obj-$(CONFIG_SENSORS_ASB100)	+= asb100.o
