--- original/drivers/gpu/msm/adreno.h	2020-05-07 08:31:22.947812780 +0200
+++ changed/drivers/gpu/msm/adreno.h	2019-06-03 13:39:50.000000000 +0200
@@ -231,6 +231,7 @@ enum adreno_preempt_states {
  * @work: A work struct for the preemption worker (for 5XX)
  * @token_submit: Indicates if a preempt token has been submitted in
  * current ringbuffer (for 4XX)
+ * @starved: A bitmask of ringbuffers that are starved for attention
  */
 struct adreno_preemption {
 	atomic_t state;
@@ -238,6 +239,7 @@ struct adreno_preemption {
 	struct timer_list timer;
 	struct work_struct work;
 	bool token_submit;
+	unsigned long starved;
 };
 
 
