--- original/drivers/md/dm-verity-fec.c	2020-05-07 08:31:23.457808192 +0200
+++ changed/drivers/md/dm-verity-fec.c	2019-06-03 13:39:50.000000000 +0200
@@ -266,6 +266,12 @@ static int fec_read_bufs(struct dm_verit
 			continue;
 		}
 
+		/* assumes block0's first 1024 bytes were all zeroes when encoding.*/
+		if(block == 0 && bufio == v->fec->data_bufio){
+			memset(bbuf, 0, 1024);
+			goto skip_erasure;
+		}
+
 		/* locate erasures if the block is on the data device */
 		if (bufio == v->fec->data_bufio &&
 		    verity_hash_for_block(v, io, block, want_digest,
@@ -283,6 +289,7 @@ static int fec_read_bufs(struct dm_verit
 				fio->erasures[(*neras)++] = i;
 		}
 
+skip_erasure :
 		/*
 		 * deinterleave and copy the bytes that fit into bufs,
 		 * starting from block_offset
@@ -439,6 +446,9 @@ int verity_fec_decode(struct dm_verity *
 	struct dm_verity_fec_io *fio = fec_io(io);
 	u64 offset, res, rsb;
 
+	if (block == 0)
+		return -1;
+
 	if (!verity_fec_is_enabled(v))
 		return -EOPNOTSUPP;
 
@@ -449,8 +459,6 @@ int verity_fec_decode(struct dm_verity *
 
 	fio->level++;
 
-	if (type == DM_VERITY_BLOCK_TYPE_METADATA)
-		block += v->data_blocks;
 
 	/*
 	 * For RS(M, N), the continuous FEC data is divided into blocks of N
