--- original/drivers/net/ethernet/broadcom/Makefile	2020-05-07 08:06:14.227262426 +0200
+++ changed/drivers/net/ethernet/broadcom/Makefile	2019-06-03 14:17:54.000000000 +0200
@@ -9,6 +9,5 @@ obj-$(CONFIG_BNX2) += bnx2.o
 obj-$(CONFIG_CNIC) += cnic.o
 obj-$(CONFIG_BNX2X) += bnx2x/
 obj-$(CONFIG_SB1250_MAC) += sb1250-mac.o
-obj-$(CONFIG_TIGON3) += tg3.o
 obj-$(CONFIG_BGMAC) += bgmac.o
 obj-$(CONFIG_SYSTEMPORT) += bcmsysport.o
