--- original/arch/arm/Kconfig	2020-05-07 08:31:19.791841172 +0200
+++ changed/arch/arm/Kconfig	2019-06-03 13:39:48.000000000 +0200
@@ -294,6 +294,9 @@ config PHYS_OFFSET
 	default 0xd0000000 if EP93XX_SDCE1_PHYS_OFFSET
 	default 0xe0000000 if EP93XX_SDCE2_PHYS_OFFSET
 	default 0xf0000000 if EP93XX_SDCE3_ASYNC_PHYS_OFFSET
+	default 0x80000000 if ARCH_MSM8917 && !3G_DDR_SUPPORT
+	default 0x20000000 if ARCH_MSM8917 && 3G_DDR_SUPPORT && !SEC_J6PRIMELTE_PROJECT
+	default 0x40000000 if ARCH_MSM8917 && 3G_DDR_SUPPORT && SEC_J6PRIMELTE_PROJECT
 	help
 	  Please provide the physical address corresponding to the
 	  location of main memory in your system.
@@ -2282,6 +2285,33 @@ config ARCH_HIBERNATION_POSSIBLE
 
 endmenu
 
+config TIMA_LKMAUTH
+       bool "Enable LKM authentication TIMA feature"
+       default n
+       help
+         TIMA authenticates loaded kernel modules. Disable it if you don't
+         want TIMA enabled.
+
+config TIMA_LKMAUTH_CODE_PROT
+       bool "Support kernel Module"
+       depends on MODULES
+       default n
+       help
+         When enabled, this feature will mark all Executable sections in an
+         LKM as RO
+
+config TIMA_TEST_INFRA
+       tristate "Testing infrastructure for TIMA"
+       default n
+       help
+         Kernel bits of the TIMA test infrastructure
+
+config TIMA_UEVENT
+       bool "Uevent message propagation for TIMA"
+       default n
+       help
+         Uevent messages for TIMA PKM/LKM
+
 source "net/Kconfig"
 
 source "drivers/Kconfig"
@@ -2300,3 +2330,5 @@ endif
 source "lib/Kconfig"
 
 source "arch/arm/kvm/Kconfig"
+
+source "arch/arm/Kconfig.sec"
