--- original/arch/arm64/Kconfig	2020-05-07 08:31:20.334836287 +0200
+++ changed/arch/arm64/Kconfig	2019-06-03 13:39:48.000000000 +0200
@@ -58,11 +58,8 @@ config ARM64
 	select HAVE_DMA_API_DEBUG
 	select HAVE_DMA_ATTRS
 	select HAVE_DMA_CONTIGUOUS
-	select HAVE_DYNAMIC_FTRACE
 	select HAVE_EFFICIENT_UNALIGNED_ACCESS
-	select HAVE_FTRACE_MCOUNT_RECORD
 	select HAVE_FUNCTION_TRACER
-	select HAVE_FUNCTION_GRAPH_TRACER
 	select HAVE_GENERIC_DMA_COHERENT
 	select HAVE_HW_BREAKPOINT if PERF_EVENTS
 	select HAVE_IRQ_TIME_ACCOUNTING
@@ -364,6 +361,10 @@ config ARCH_MSM8994_V1_TLBI_WA
 
 endmenu
 
+config SAMSUNG_PRODUCT_SHIP
+        bool "set up for product shipping"
+        default n
+        
 menu "Bus support"
 
 config ARM_AMBA
@@ -1119,6 +1120,43 @@ config ARM64_ERRATUM_843419
 
 endmenu
 
+config TIMA_GET_QSEE_LOGS_USING_RDX
+	bool "Enable QSEE Log Extraction using RDX tool"
+	default n
+
+config TIMA_LKMAUTH
+       bool "Enable LKM authentication TIMA feature"
+       default n
+       help
+         TIMA authenticates loaded kernel modules. Disable it if you don't
+         want TIMA enabled.
+
+config TIMA_LKMAUTH_CODE_PROT
+       bool "Support kernel Module"
+       depends on MODULES
+       default n
+       help
+         When enabled, this feature will mark all Executable sections in an
+         LKM as RO
+
+config TIMA_TEST_INFRA
+       tristate "Testing infrastructure for TIMA"
+       default n
+       help
+         Kernel bits of the TIMA test infrastructure
+
+config TIMA_UEVENT
+       bool "Uevent message propagation for TIMA"
+       default n
+       help
+         Uevent messages for TIMA PKM/LKM
+
+config RELOCATABLE_KERNEL 
+	bool "set kernel to relocatable in both virtual and physical address"
+	default n 
+	help 	
+	  This configuare will allow you to config the kernel into position independent kernel, with '--PIE' insert into LD script.
+
 source "net/Kconfig"
 
 source "drivers/Kconfig"
@@ -1141,3 +1179,13 @@ endif
 source "lib/Kconfig"
 
 source "arch/arm64/mm/Kconfig"
+
+config RTC_AUTO_PWRON
+        bool "RTC Auto Power on PMICs"
+        depends on ARCH_MSM8953
+        default n
+        help
+          Support for the auto power on alarm on the PMIC.
+          Say Y if you want to enable this feature.
+
+source "arch/arm64/Kconfig.sec"
