--- original/arch/arm/boot/dts/qcom/msm8917-pm.dtsi	2020-05-07 08:06:05.309350055 +0200
+++ changed/arch/arm/boot/dts/qcom/msm8917-pm.dtsi	2019-06-03 13:39:48.000000000 +0200
@@ -181,6 +181,7 @@
 			<0xff 126>, /* smmu_bus_intr[39] */
 			<0xff 128>, /* blsp1_peripheral_irq[3] */
 			<0xff 129>, /* blsp1_peripheral_irq[4] */
+			<0xff 130>, /* rotate sensor */
 			<0xff 131>, /* qup_irq */
 			<0xff 136>, /* smmu_bus_intr[43] */
 			<0xff 137>, /* smmu_intr_bus[44] */
@@ -241,6 +242,7 @@
 			<0xff 263>, /* ipa_bam_irq[2] */
 			<0xff 269>, /* rpm_wdog_expired_irq */
 			<0xff 270>, /* blsp1_bam_irq[0] */
+			<0xff 271>, /* i2c dma */
 			<0xff 272>, /* smmu_intr_bus[17] */
 			<0xff 273>, /* smmu_bus_intr[18] */
 			<0xff 274>, /* smmu_bus_intr[19] */
@@ -254,7 +256,8 @@
 			<0xff 322>, /* q6ss_irq_out(5) */
 			<0xff 323>, /* q6ss_irq_out(6) */
 			<0xff 325>, /* q6ss_wdog_exp_irq */
-			<0xff 344>; /* sdcc1ice */
+			<0xff 344>, /* sdcc1ice */
+			<0xff 351>; /* trustonic */
 
 		qcom,gpio-parent = <&tlmm>;
 		qcom,gpio-map = <3  38 >,
