--- original/arch/arm64/Makefile	2020-05-07 08:31:20.335836278 +0200
+++ changed/arch/arm64/Makefile	2019-06-03 13:39:48.000000000 +0200
@@ -10,7 +10,12 @@
 #
 # Copyright (C) 1995-2001 by Russell King
 
+ifeq ($(CONFIG_RELOCATABLE_KERNEL), y) 
+LDFLAGS_vmlinux	:=-p --no-undefined -X -pie -Bsymbolic
+else 
 LDFLAGS_vmlinux	:=-p --no-undefined -X
+endif 
+
 CPPFLAGS_vmlinux.lds = -DTEXT_OFFSET=$(TEXT_OFFSET)
 OBJCOPYFLAGS	:=-O binary -R .note -R .note.gnu.build-id -R .comment -S
 GZFLAGS		:=-9
@@ -92,7 +97,19 @@ zinstall install: vmlinux
 PHONY += dtbs dtbs_install
 
 dtbs: prepare scripts
-	$(Q)$(MAKE) $(build)=$(boot)/dts
+#	$(Q)$(MAKE) $(build)=$(boot)/dts
+# MSM8953
+ifeq ($(CONFIG_ARCH_SDM450),y)
+	$(Q)$(MAKE) $(build)=$(boot)/dts/samsung/sdm450
+# MSM8953
+else ifeq ($(CONFIG_ARCH_MSM8953),y)
+	$(Q)$(MAKE) $(build)=$(boot)/dts/samsung/msm8953
+# MSM8937
+else ifeq ($(CONFIG_ARCH_MSM8937),y)
+	$(Q)$(MAKE) $(build)=$(boot)/dts/samsung/msm8937
+else
+	$(Q)$(MAKE) $(build)=$(boot)/dts/qcom
+endif
 
 dtbs_install:
 	$(Q)$(MAKE) $(dtbinst)=$(boot)/dts
