--- original/arch/arm64/mm/mmu.c	2020-05-07 08:31:20.420835513 +0200
+++ changed/arch/arm64/mm/mmu.c	2019-06-03 13:39:48.000000000 +0200
@@ -725,6 +725,7 @@ core_initcall(map_entry_trampoline);
  */
 void __init paging_init(void)
 {
+	set_memsize_kernel_type(MEMSIZE_KERNEL_PAGING);
 	map_mem();
 	dma_contiguous_remap();
 	remap_pages();
@@ -741,6 +742,7 @@ void __init paging_init(void)
 	set_kernel_text_ro();
 	local_flush_tlb_all();
 	cpu_set_default_tcr_t0sz();
+	set_memsize_kernel_type(MEMSIZE_KERNEL_OTHERS);
 }
 
 /*
