--- original/arch/arm/crypto/sha2-ce-glue.c	2020-05-07 08:31:20.155837897 +0200
+++ changed/arch/arm/crypto/sha2-ce-glue.c	2019-06-03 13:39:48.000000000 +0200
@@ -180,7 +180,7 @@ static struct shash_alg algs[] = { {
 	.base			= {
 		.cra_name		= "sha256",
 		.cra_driver_name	= "sha256-ce",
-		.cra_priority		= 300,
+		.cra_priority		= 500,
 		.cra_flags		= CRYPTO_ALG_TYPE_SHASH,
 		.cra_blocksize		= SHA256_BLOCK_SIZE,
 		.cra_module		= THIS_MODULE,
