--- original/arch/arm64/kernel/setup.c	2020-05-07 08:31:20.397835720 +0200
+++ changed/arch/arm64/kernel/setup.c	2019-06-03 13:39:48.000000000 +0200
@@ -63,6 +63,20 @@
 #include <asm/efi.h>
 #include <asm/system_misc.h>
 
+#ifdef CONFIG_SEC_DEBUG
+#include <linux/sec_debug.h>
+#endif
+
+// [ SEC_SELINUX_PORTING_QUALCOMM
+#ifdef CONFIG_PROC_AVC
+#include <linux/proc_avc.h>
+#endif
+// ] SEC_SELINUX_PORTING_QUALCOMM
+
+
+unsigned int system_rev;
+EXPORT_SYMBOL(system_rev);
+
 unsigned int boot_reason;
 EXPORT_SYMBOL(boot_reason);
 
@@ -215,6 +229,14 @@ static void __init setup_machine_fdt(phy
 
 }
 
+static int __init msm_hw_rev_setup(char *p)
+{
+	system_rev = memparse(p, NULL);
+	printk("androidboot.revision %x", system_rev);
+	return 0;
+}
+early_param("androidboot.revision", msm_hw_rev_setup);
+
 static void __init request_standard_resources(void)
 {
 	struct memblock_region *region;
@@ -373,17 +395,27 @@ void __init setup_arch(char **cmdline_p)
 #endif
 #endif
 	init_random_pool();
-
+#ifndef CONFIG_RELOCATABLE_KERNEL
 	if (boot_args[1] || boot_args[2] || boot_args[3]) {
 		pr_err("WARNING: x1-x3 nonzero in violation of boot protocol:\n"
 			"\tx1: %016llx\n\tx2: %016llx\n\tx3: %016llx\n"
 			"This indicates a broken bootloader or old kernel\n",
 			boot_args[1], boot_args[2], boot_args[3]);
 	}
+#endif
 }
 
 static int __init arm64_device_init(void)
 {
+#ifdef CONFIG_SEC_DEBUG
+	sec_debug_init();
+#endif
+// [ SEC_SELINUX_PORTING_QUALCOMM
+#ifdef CONFIG_PROC_AVC
+        sec_avc_log_init();
+#endif
+// ] SEC_SELINUX_PORTING_QUALCOMM
+
 	of_platform_populate(NULL, of_default_bus_match_table, NULL, NULL);
 	return 0;
 }
