--- original/arch/arm/kernel/irq.c	2020-05-07 08:06:05.565347540 +0200
+++ changed/arch/arm/kernel/irq.c	2019-06-03 13:39:48.000000000 +0200
@@ -43,6 +43,10 @@
 #include <asm/mach/irq.h>
 #include <asm/mach/time.h>
 
+#ifdef CONFIG_SEC_DEBUG
+#include <linux/sec_debug.h>
+#endif
+
 unsigned long irq_err_count;
 
 int arch_show_interrupts(struct seq_file *p, int prec)
@@ -65,7 +69,14 @@ int arch_show_interrupts(struct seq_file
  */
 void handle_IRQ(unsigned int irq, struct pt_regs *regs)
 {
+#ifdef CONFIG_SEC_DEBUG
+	int cpu = smp_processor_id();
+	unsigned long long start_time = cpu_clock(cpu);
+#endif
 	__handle_domain_irq(NULL, irq, false, regs);
+#ifdef CONFIG_SEC_DEBUG
+	sec_debug_irq_enterexit_log(irq, start_time);
+#endif
 }
 
 /*
