--- original/arch/arm/boot/dts/qcom/msm8917-pinctrl.dtsi	2020-05-07 08:31:19.990839382 +0200
+++ changed/arch/arm/boot/dts/qcom/msm8917-pinctrl.dtsi	2019-06-03 13:39:48.000000000 +0200
@@ -615,7 +615,7 @@
 				config {
 					pins = "gpio67";
 					drive-strength = <2>;
-					bias-pull-up;
+					bias-disable;
 				};
 			};
 
@@ -967,24 +967,24 @@
 		pri-tlmm-lines {
 			pri_tlmm_lines_act: pri_tlmm_lines_act {
 				mux {
-					pins = "gpio85", "gpio88";
+                    pins = "gpio85", "gpio88", "gpio86";
 					function = "pri_mi2s";
 				};
 
 				config {
-					pins = "gpio85", "gpio88";
+                    pins = "gpio85", "gpio88", "gpio86";
 					drive-strength = <8>;
 				};
 			};
 
 			pri_tlmm_lines_sus: pri_tlmm_lines_sus {
 				mux {
-					pins = "gpio85", "gpio88";
+                    pins = "gpio85", "gpio88", "gpio86";
 					function = "pri_mi2s";
 				};
 
 				config {
-					pins = "gpio85", "gpio88";
+                    pins = "gpio85", "gpio88", "gpio86";
 					drive-strength = <2>;
 					bias-pull-down;
 				};
