--- original/arch/arm64/kernel/smp.c	2020-05-07 08:51:34.478261479 +0200
+++ changed/arch/arm64/kernel/smp.c	2019-06-03 13:39:48.000000000 +0200
@@ -56,6 +56,8 @@
 #define CREATE_TRACE_POINTS
 #include <trace/events/ipi.h>
 
+#include <linux/sec_debug.h>
+
 /*
  * as from 2.5, kernels no longer have an init_tasks structure
  * so we need some other way of telling a new secondary core
@@ -596,6 +598,7 @@ static void ipi_cpu_stop(unsigned int cp
 		show_regs(regs);
 		dump_stack();
 		arm64_check_cache_ecc(NULL);
+		sec_debug_save_context();
 		raw_spin_unlock(&stop_lock);
 	}
 
