--- original/scripts/kconfig/confdata.c	2020-05-07 08:06:23.734169009 +0200
+++ changed/scripts/kconfig/confdata.c	2019-06-03 13:39:54.000000000 +0200
@@ -242,7 +242,7 @@ e_out:
 	return -1;
 }
 
-int conf_read_simple(const char *name, int def)
+int conf_read_simple(const char *name, int def, int sym_init)
 {
 	FILE *in = NULL;
 	char   *line = NULL;
@@ -290,6 +290,8 @@ load:
 	conf_unsaved = 0;
 
 	def_flags = SYMBOL_DEF << def;
+	if (!sym_init)
+		goto readsym;
 	for_all_symbols(i, sym) {
 		sym->flags |= SYMBOL_CHANGED;
 		sym->flags &= ~(def_flags|SYMBOL_VALID);
@@ -307,7 +309,7 @@ load:
 			sym->def[def].tri = no;
 		}
 	}
-
+readsym:
 	while (compat_getline(&line, &line_asize, in) != -1) {
 		conf_lineno++;
 		sym = NULL;
@@ -411,7 +413,7 @@ int conf_read(const char *name)
 
 	sym_set_change_count(0);
 
-	if (conf_read_simple(name, S_DEF_USER))
+	if (conf_read_simple(name, S_DEF_USER, true))
 		return 1;
 
 	for_all_symbols(i, sym) {
@@ -842,7 +844,7 @@ static int conf_split_config(void)
 	int res, i, fd;
 
 	name = conf_get_autoconfig_name();
-	conf_read_simple(name, S_DEF_AUTO);
+	conf_read_simple(name, S_DEF_AUTO, true);
 
 	if (chdir("include/config"))
 		return 1;
