--- original/scripts/sortextable.c	2020-05-07 08:06:23.768168675 +0200
+++ changed/scripts/sortextable.c	2019-06-03 13:39:54.000000000 +0200
@@ -262,9 +262,9 @@ do_file(char const *const fname)
 		break;
 	}  /* end switch */
 	if (memcmp(ELFMAG, ehdr->e_ident, SELFMAG) != 0
-	||  r2(&ehdr->e_type) != ET_EXEC
+	|| ( r2(&ehdr->e_type) != ET_EXEC && r2(&ehdr->e_type) != ET_DYN )
 	||  ehdr->e_ident[EI_VERSION] != EV_CURRENT) {
-		fprintf(stderr, "unrecognized ET_EXEC file %s\n", fname);
+		fprintf(stderr, "unrecognized ET_EXEC/ET_DYN file %s\n", fname);
 		fail_file();
 	}
 
