--- original/include/trace/events/trace_msm_low_power.h	2020-05-07 08:31:27.247774096 +0200
+++ changed/include/trace/events/trace_msm_low_power.h	2019-06-03 13:39:53.000000000 +0200
@@ -122,7 +122,7 @@ DECLARE_EVENT_CLASS(cluster,
 	TP_ARGS(name, index, sync_cpus, child_cpus, from_idle),
 
 	TP_STRUCT__entry(
-		__field(const char *, name)
+		__string(       name,           name            )
 		__field(int, index)
 		__field(unsigned long, sync_cpus)
 		__field(unsigned long, child_cpus)
@@ -130,7 +130,7 @@ DECLARE_EVENT_CLASS(cluster,
 	),
 
 	TP_fast_assign(
-		__entry->name = name;
+		__assign_str(name, name);
 		__entry->index = index;
 		__entry->sync_cpus = sync_cpus;
 		__entry->child_cpus = child_cpus;
@@ -138,7 +138,7 @@ DECLARE_EVENT_CLASS(cluster,
 	),
 
 	TP_printk("cluster_name:%s idx:%d sync:0x%lx child:0x%lx idle:%d",
-		__entry->name,
+		__get_str(name),
 		__entry->index,
 		__entry->sync_cpus,
 		__entry->child_cpus,
