--- original/include/linux/mmc/pm.h	2020-05-07 08:06:21.345192483 +0200
+++ changed/include/linux/mmc/pm.h	2019-06-03 13:39:53.000000000 +0200
@@ -27,5 +27,6 @@ typedef unsigned int mmc_pm_flag_t;
 #define MMC_PM_KEEP_POWER	(1 << 0)	/* preserve card power during suspend */
 #define MMC_PM_WAKE_SDIO_IRQ	(1 << 1)	/* wake up host system on SDIO IRQ assertion */
 #define MMC_PM_IGNORE_PM_NOTIFY	(1 << 2)	/* ignore mmc pm notify */
+#define MMC_PM_SKIP_RESUME_INIT	(1 << 3)	/* skip eMMC init during resume */
 
 #endif /* LINUX_MMC_PM_H */
