--- original/include/net/tcp.h	2020-05-07 08:51:35.383252614 +0200
+++ changed/include/net/tcp.h	2019-06-03 13:39:53.000000000 +0200
@@ -280,6 +280,10 @@ extern unsigned int sysctl_tcp_notsent_l
 extern int sysctl_tcp_min_tso_segs;
 extern int sysctl_tcp_autocorking;
 extern int sysctl_tcp_default_init_rwnd;
+#ifdef CONFIG_NETPM
+extern int sysctl_tcp_netpm[4];
+extern struct net_device *ip6_dev_find(struct net *net, const struct in6_addr *addr);
+#endif
 
 extern atomic_long_t tcp_memory_allocated;
 
@@ -1087,7 +1091,6 @@ static inline void tcp_prequeue_init(str
 }
 
 bool tcp_prequeue(struct sock *sk, struct sk_buff *skb);
-int tcp_filter(struct sock *sk, struct sk_buff *skb);
 
 #undef STATE_TRACE
 
@@ -1117,6 +1120,16 @@ void tcp_select_initial_window(int __spa
 			       __u32 *window_clamp, int wscale_ok,
 			       __u8 *rcv_wscale, __u32 init_rcv_wnd);
 
+#ifdef CONFIG_NETPM
+static inline int tcp_space_from_win(int win)
+{
+	return sysctl_tcp_adv_win_scale <= 0 ?
+			(win<<(-sysctl_tcp_adv_win_scale)) :
+			(win<<sysctl_tcp_adv_win_scale)/
+				((1<<sysctl_tcp_adv_win_scale)-1);
+}
+#endif
+
 static inline int tcp_win_from_space(int space)
 {
 	int tcp_adv_win_scale = sysctl_tcp_adv_win_scale;
