--- original/include/linux/spi/spi.h	2020-05-07 08:06:21.529190675 +0200
+++ changed/include/linux/spi/spi.h	2019-06-03 13:39:53.000000000 +0200
@@ -774,6 +774,17 @@ extern int spi_sync_locked(struct spi_de
 extern int spi_bus_lock(struct spi_master *master);
 extern int spi_bus_unlock(struct spi_master *master);
 
+#ifdef ENABLE_SENSORS_FPRINT_SECURE
+extern int fp_spi_clock_set_rate(struct spi_device *spidev);
+extern int fp_spi_clock_enable(struct spi_device *spidev);
+extern int fp_spi_clock_disable(struct spi_device *spidev);
+extern int fp_spi_request_gpios(struct spi_device *spidev);
+#endif
+extern int ese_spi_clock_set_rate(struct spi_device *spidev);
+extern int ese_spi_clock_enable(struct spi_device *spidev);
+extern int ese_spi_clock_disable(struct spi_device *spidev);
+extern int ese_spi_request_gpios(struct spi_device *spidev);
+
 /**
  * spi_write - SPI synchronous write
  * @spi: device to which data will be written
