--- original/include/linux/init_task.h	2020-05-07 08:31:26.879777407 +0200
+++ changed/include/linux/init_task.h	2019-06-03 13:39:53.000000000 +0200
@@ -14,6 +14,7 @@
 #include <linux/rbtree.h>
 #include <net/net_namespace.h>
 #include <linux/sched/rt.h>
+#include <linux/task_integrity.h>
 
 #ifdef CONFIG_SMP
 # define INIT_PUSHABLE_TASKS(tsk)					\
@@ -156,6 +157,23 @@ extern struct task_group root_task_group
 # define INIT_VTIME(tsk)
 #endif
 
+#ifdef CONFIG_FIVE
+# define INIT_TASK_INTEGRITY(integrity) {				\
+	.user_value = INTEGRITY_NONE,					\
+	.value = INTEGRITY_NONE,					\
+	.usage_count = ATOMIC_INIT(1),					\
+	.value_lock = __SPIN_LOCK_UNLOCKED(integrity.value_lock),	\
+	.list_lock = __SPIN_LOCK_UNLOCKED(integrity.list_lock),		\
+	.events = { .list = LIST_HEAD_INIT(integrity.events.list),},	\
+}
+
+# define INIT_INTEGRITY(tsk)						\
+	.integrity = &init_integrity,
+#else
+# define INIT_INTEGRITY(tsk)
+# define INIT_TASK_INTEGRITY(integrity)
+#endif
+
 #define INIT_TASK_COMM "swapper"
 
 #ifdef CONFIG_RT_MUTEXES
@@ -248,6 +266,7 @@ extern struct task_group root_task_group
 	INIT_RT_MUTEXES(tsk)						\
 	INIT_VTIME(tsk)							\
 	INIT_KASAN(tsk)							\
+	INIT_INTEGRITY(tsk)						\
 }
 
 
