--- original/include/linux/sched.h	2020-05-07 08:31:27.034776012 +0200
+++ changed/include/linux/sched.h	2019-06-03 13:39:53.000000000 +0200
@@ -1194,6 +1194,10 @@ struct ravg {
 	u64 mark_start;
 	u32 sum, demand;
 	u32 sum_history[RAVG_HIST_SIZE_MAX];
+#ifdef CONFIG_SEC_ADAPTIVE_LOAD_TRACKING
+	u32 global_average;
+	u32 nr_global_average;
+#endif
 #ifdef CONFIG_SCHED_FREQ_INPUT
 	u32 curr_window, prev_window;
 	u16 active_windows;
@@ -1310,6 +1314,10 @@ union rcu_special {
 };
 struct rcu_node;
 
+#ifdef CONFIG_FIVE
+struct task_integrity;
+#endif
+
 enum perf_event_task_context {
 	perf_invalid_context = -1,
 	perf_hw_context = 0,
@@ -1782,6 +1790,13 @@ struct task_struct {
 	unsigned int	sequential_io;
 	unsigned int	sequential_io_avg;
 #endif
+#ifdef CONFIG_SDP
+	unsigned int sensitive;
+#endif
+
+#ifdef CONFIG_FIVE
+	struct task_integrity *integrity;
+#endif
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
