--- original/include/linux/mmc/sdhci.h	2020-05-07 08:06:21.346192474 +0200
+++ changed/include/linux/mmc/sdhci.h	2019-06-03 13:39:53.000000000 +0200
@@ -304,6 +304,12 @@ struct sdhci_host {
 	ktime_t reset_wa_t; /* time when the reset workaround is applied */
 	int reset_wa_cnt; /* total number of times workaround is used */
 
+	u32 send_cmd_idx;
+	s64 send_cmd_timestamp;
+	u32 irq_cmd_idx;
+	s64 irq_timestamp;
+	s64 finish_tasklet_timestamp;
+
 	unsigned long private[0] ____cacheline_aligned;
 };
 #endif /* LINUX_MMC_SDHCI_H */
