--- original/include/linux/tcp.h	2020-05-07 08:31:27.068775706 +0200
+++ changed/include/linux/tcp.h	2019-06-03 13:39:53.000000000 +0200
@@ -283,6 +283,21 @@ struct tcp_sock {
 
 	int			linger2;
 
+/* Network Pacemaker */
+#ifdef CONFIG_NETPM
+	u8 netpm_netif;
+	u8 netpm_rbuf_flag;
+	u32 netpm_rtt_min;
+	u32 netpm_srtt;
+	u32 netpm_rttvar;
+	int netpm_cwnd_est;
+	int netpm_tcp_rmem_max;
+	int netpm_tcp_rmem_max_base;
+	int netpm_rwnd_max_adjust;
+	int netpm_max_tput;
+	int netpm_rmem_max_curbdp[2];
+#endif
+
 /* Receiver side RTT estimation */
 	struct {
 		u32	rtt;
