--- original/include/soc/qcom/boot_stats.h	2020-05-07 08:31:27.195774563 +0200
+++ changed/include/soc/qcom/boot_stats.h	2019-06-03 13:39:53.000000000 +0200
@@ -10,6 +10,8 @@
  * GNU General Public License for more details.
  */
 
+#ifndef __BOOT_STATS__
+#define __BOOT_STATS__
 #ifdef CONFIG_MSM_BOOT_STATS
 
 #define TIMER_KHZ 32768
@@ -44,3 +46,4 @@ void place_marker(const char *name);
 inline void place_marker(char *name);
 static inline int boot_marker_enabled(void) { return 0; }
 #endif
+#endif /*__BOOT_STATS__ */
\ Geen regeleindeteken (LF) aan einde van bestand
