--- original/include/trace/events/ext4.h	2020-05-07 08:06:21.800188013 +0200
+++ changed/include/trace/events/ext4.h	2019-06-03 13:39:53.000000000 +0200
@@ -839,6 +839,7 @@ TRACE_EVENT(ext4_sync_file_enter,
 		__field(	ino_t,	ino			)
 		__field(	ino_t,	parent			)
 		__field(	int,	datasync		)
+		__array(unsigned char,	d_name,	EXT4_NAME_LEN)
 	),
 
 	TP_fast_assign(
@@ -848,12 +849,14 @@ TRACE_EVENT(ext4_sync_file_enter,
 		__entry->ino		= dentry->d_inode->i_ino;
 		__entry->datasync	= datasync;
 		__entry->parent		= dentry->d_parent->d_inode->i_ino;
+		memcpy(__entry->d_name, dentry->d_name.name, EXT4_NAME_LEN);
 	),
 
-	TP_printk("dev %d,%d ino %lu parent %lu datasync %d ",
+	TP_printk("dev %d,%d ino %lu parent %lu datasync %d d_name %s",
 		  MAJOR(__entry->dev), MINOR(__entry->dev),
 		  (unsigned long) __entry->ino,
-		  (unsigned long) __entry->parent, __entry->datasync)
+		  (unsigned long) __entry->parent, __entry->datasync,
+		  __entry->d_name)
 );
 
 TRACE_EVENT(ext4_sync_file_exit,
@@ -1422,6 +1425,7 @@ TRACE_EVENT(ext4_unlink_enter,
 		__field(	ino_t,	ino			)
 		__field(	ino_t,	parent			)
 		__field(	loff_t,	size			)
+		__array(unsigned char,	d_name,	EXT4_NAME_LEN)
 	),
 
 	TP_fast_assign(
@@ -1429,12 +1433,13 @@ TRACE_EVENT(ext4_unlink_enter,
 		__entry->ino		= dentry->d_inode->i_ino;
 		__entry->parent		= parent->i_ino;
 		__entry->size		= dentry->d_inode->i_size;
+		memcpy(__entry->d_name, dentry->d_name.name, EXT4_NAME_LEN);
 	),
 
-	TP_printk("dev %d,%d ino %lu size %lld parent %lu",
+	TP_printk("dev %d,%d ino %lu size %lld parent %lu d_name %s",
 		  MAJOR(__entry->dev), MINOR(__entry->dev),
 		  (unsigned long) __entry->ino, __entry->size,
-		  (unsigned long) __entry->parent)
+		  (unsigned long) __entry->parent, __entry->d_name)
 );
 
 TRACE_EVENT(ext4_unlink_exit,
@@ -1446,18 +1451,20 @@ TRACE_EVENT(ext4_unlink_exit,
 		__field(	dev_t,	dev			)
 		__field(	ino_t,	ino			)
 		__field(	int,	ret			)
+		__array(unsigned char,	d_name,	EXT4_NAME_LEN)
 	),
 
 	TP_fast_assign(
 		__entry->dev		= dentry->d_inode->i_sb->s_dev;
 		__entry->ino		= dentry->d_inode->i_ino;
 		__entry->ret		= ret;
+		memcpy(__entry->d_name, dentry->d_name.name, EXT4_NAME_LEN);
 	),
 
-	TP_printk("dev %d,%d ino %lu ret %d",
+	TP_printk("dev %d,%d ino %lu ret %d d_name %s",
 		  MAJOR(__entry->dev), MINOR(__entry->dev),
 		  (unsigned long) __entry->ino,
-		  __entry->ret)
+		  __entry->ret, __entry->d_name)
 );
 
 DECLARE_EVENT_CLASS(ext4__truncate,
