--- original/include/soc/qcom/camera2.h	2020-05-07 08:31:27.196774555 +0200
+++ changed/include/soc/qcom/camera2.h	2019-06-03 13:39:53.000000000 +0200
@@ -20,6 +20,7 @@
 #include <linux/of_device.h>
 #include <linux/of.h>
 
+#define MAX_SPECIAL_SUPPORT_SIZE 10
 
 enum msm_camera_device_type_t {
 	MSM_CAMERA_I2C_DEVICE,
@@ -150,6 +151,8 @@ struct msm_camera_sensor_board_info {
 	const char *actuator_name;
 	const char *flash_name;
 	const char *ois_name;
+	const char *special_support_sensors[MAX_SPECIAL_SUPPORT_SIZE];
+	int32_t special_support_size;
 	struct msm_camera_slave_info *slave_info;
 	struct msm_camera_csi_lane_params *csi_lane_params;
 	struct msm_camera_sensor_strobe_flash_data *strobe_flash_data;
