--- original/include/sound/q6asm-v2.h	2020-05-07 08:31:27.227774276 +0200
+++ changed/include/sound/q6asm-v2.h	2019-06-03 13:39:53.000000000 +0200
@@ -274,7 +274,7 @@ int q6asm_open_read_v3(struct audio_clie
 		       uint16_t bits_per_sample);
 
 int q6asm_open_read_v4(struct audio_client *ac, uint32_t format,
-			uint16_t bits_per_sample);
+			uint16_t bits_per_sample, bool ts_mode);
 
 int q6asm_open_write(struct audio_client *ac, uint32_t format
 		/*, uint16_t bits_per_sample*/);
@@ -639,4 +639,12 @@ int q6asm_send_mtmx_strtr_window(struct
 /* Retrieve the current DSP path delay */
 int q6asm_get_path_delay(struct audio_client *ac);
 
+#ifdef CONFIG_SEC_SND_SOLUTION
+int q6asm_set_sa(struct audio_client *ac, long *param);
+int q6asm_set_vsp(struct audio_client *ac, long *param);
+int q6asm_set_dha(struct audio_client *ac, long *param);
+int q6asm_set_lrsm(struct audio_client *ac, long *param);
+int q6asm_set_msp(struct audio_client *ac, long *param);
+int q6asm_set_sb(struct audio_client *ac, long *param);
+#endif /* CONFIG_SEC_SND_SOLUTION */
 #endif /* __Q6_ASM_H__ */
