--- original/net/ipv4/af_inet.c	2020-05-07 08:31:27.855768626 +0200
+++ changed/net/ipv4/af_inet.c	2019-06-03 13:39:54.000000000 +0200
@@ -405,7 +405,6 @@ out_rcu_unlock:
 	goto out;
 }
 
-
 /*
  *	The peer socket should always be NULL (or else). When we call this
  *	function we are destroying the object and from then on nobody
@@ -748,6 +747,7 @@ int inet_sendmsg(struct kiocb *iocb, str
 		 size_t size)
 {
 	struct sock *sk = sock->sk;
+	int err;
 
 	sock_rps_record_flow(sk);
 
@@ -756,7 +756,9 @@ int inet_sendmsg(struct kiocb *iocb, str
 	    inet_autobind(sk))
 		return -EAGAIN;
 
-	return sk->sk_prot->sendmsg(iocb, sk, msg, size);
+    err = sk->sk_prot->sendmsg(iocb, sk, msg, size);
+
+	return err;
 }
 EXPORT_SYMBOL(inet_sendmsg);
 
@@ -789,8 +791,9 @@ int inet_recvmsg(struct kiocb *iocb, str
 
 	err = sk->sk_prot->recvmsg(iocb, sk, msg, size, flags & MSG_DONTWAIT,
 				   flags & ~MSG_DONTWAIT, &addr_len);
-	if (err >= 0)
+	if (err >= 0) {
 		msg->msg_namelen = addr_len;
+	}
 	return err;
 }
 EXPORT_SYMBOL(inet_recvmsg);
