--- original/net/rose/rose_in.c	2020-05-07 08:06:23.465171652 +0200
+++ changed/net/rose/rose_in.c	2019-06-03 13:39:54.000000000 +0200
@@ -164,8 +164,7 @@ static int rose_state3_machine(struct so
 		rose_frames_acked(sk, nr);
 		if (ns == rose->vr) {
 			rose_start_idletimer(sk);
-			if (sk_filter_trim_cap(sk, skb, ROSE_MIN_LEN) == 0 &&
-			    sock_queue_rcv_skb(sk, skb) == 0) {
+			if (sock_queue_rcv_skb(sk, skb) == 0) {
 				rose->vr = (rose->vr + 1) % ROSE_MODULUS;
 				queued = 1;
 			} else {
