--- original/net/Kconfig	2020-05-07 08:06:22.675179415 +0200
+++ changed/net/Kconfig	2019-06-03 13:39:54.000000000 +0200
@@ -253,6 +253,7 @@ source "net/netlink/Kconfig"
 source "net/mpls/Kconfig"
 source "net/hsr/Kconfig"
 source "net/rmnet_data/Kconfig"
+source "net/ncm/Kconfig"
 
 config RPS
 	boolean "RPS"
@@ -410,3 +411,9 @@ endif   # if NET
 # Used by archs to tell that they support BPF_JIT
 config HAVE_BPF_JIT
 	bool
+
+# Network Pacemaker
+config NETPM
+	bool "Network Pacemaker"
+	---help---
+	This feature implements Network Pacemaker (NETPM)
