--- original/net/ipv6/netfilter/ip6_tables.c	2020-05-07 11:40:16.672312089 +0200
+++ changed/net/ipv6/netfilter/ip6_tables.c	2019-06-03 13:39:54.000000000 +0200
@@ -1495,8 +1495,11 @@ check_compat_entry_size_and_hooks(struct
 		return -EINVAL;
 	}
 
-	/* For purposes of check_entry casting the compat entry is fine */
-	ret = check_entry((struct ip6t_entry *)e);
+	if (!ip6_checkentry(&e->ipv6))
+		return -EINVAL;
+
+	ret = xt_compat_check_entry_offsets(e, e->elems,
+					    e->target_offset, e->next_offset);
 	if (ret)
 		return ret;
 
