--- original/net/ipv4/netfilter/arp_tables.c	2020-05-07 11:40:16.666312147 +0200
+++ changed/net/ipv4/netfilter/arp_tables.c	2019-06-03 13:39:54.000000000 +0200
@@ -1228,8 +1228,11 @@ check_compat_entry_size_and_hooks(struct
 		return -EINVAL;
 	}
 
-	/* For purposes of check_entry casting the compat entry is fine */
-	ret = check_entry((struct arpt_entry *)e);
+	if (!arp_checkentry(&e->arp))
+		return -EINVAL;
+
+	ret = xt_compat_check_entry_offsets(e, e->elems, e->target_offset,
+					    e->next_offset);
 	if (ret)
 		return ret;
 
