--- original/net/sunrpc/sched.c	2020-05-07 08:31:28.316764479 +0200
+++ changed/net/sunrpc/sched.c	2019-06-03 13:39:54.000000000 +0200
@@ -250,11 +250,11 @@ void rpc_destroy_wait_queue(struct rpc_w
 }
 EXPORT_SYMBOL_GPL(rpc_destroy_wait_queue);
 
-static int rpc_wait_bit_killable(struct wait_bit_key *key)
+static int rpc_wait_bit_killable(struct wait_bit_key *key, int mode)
 {
-	if (fatal_signal_pending(current))
-		return -ERESTARTSYS;
 	freezable_schedule_unsafe();
+	if (signal_pending_state(mode, current))
+		return -ERESTARTSYS;
 	return 0;
 }
 
