--- original/net/netlink/af_netlink.c	2020-05-07 08:31:28.214765396 +0200
+++ changed/net/netlink/af_netlink.c	2019-06-03 13:39:54.000000000 +0200
@@ -269,6 +269,11 @@ static int __netlink_deliver_tap_skb(str
 	struct sk_buff *nskb;
 	struct sock *sk = skb->sk;
 	int ret = -ENOMEM;
+	
+	if (!net_eq(dev_net(dev), sock_net(sk)) &&
+		!net_eq(dev_net(dev), &init_net)) {
+		return 0;
+	}
 
 	if (!net_eq(dev_net(dev), sock_net(sk)))
 		return 0;
